{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The contents include:\n",
    "\n",
    "* 1. Querying the SRA database\n",
    "* 2. Downloading data from the SRA database\n",
    "* 3. Reading FASTQ files in R\n",
    "* 4. Reading alignment data\n",
    "* 5. Preprocessing the raw NGS data\n",
    "\n",
    "## 1. Querying the SRA database\n",
    "\n",
    "Most DNA sequencing data are stored [Sequence Read Archive (SRA) database](http://www.ncbi.nlm.nih.gov/sra) in terms of short reads generated by high-throughput sequencing obtained via different platforms. The sequences are usually less than 1,000 base pairs in length. At first, we need to know how to find sequences what we need. This part will introduce steps to query and access the data residing in the SRA database from within R. <br>\n",
    "\n",
    "The following steps will help us run the desired queries on the SRA database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Bioconductor version 3.9 (BiocManager 1.30.10), R 3.6.2 (2019-12-12)\n",
      "\n",
      "Installing package(s) 'SRAdb'\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "package 'SRAdb' successfully unpacked and MD5 sums checked\n",
      "\n",
      "The downloaded binary packages are in\n",
      "\tC:\\Users\\Master\\AppData\\Local\\Temp\\RtmpCEfOQw\\downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Old packages: 'BH', 'bit', 'caTools', 'cli', 'FactoMineR', 'fansi', 'farver',\n",
      "  'GetoptLong', 'ggridges', 'gplots', 'hms', 'knitr', 'leaps', 'multcomp',\n",
      "  'mvtnorm', 'PMA', 'precrec', 'prettyunits', 'pROC', 'RCurl', 'RSQLite',\n",
      "  'Rttf2pt1', 'shinyjs', 'stringi', 'tinytex', 'xfun', 'XML', 'xts', 'zoo'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "install.packages(\"BiocManager\")\n",
    "BiocManager::install(\"SRAdb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Unzipping...\n",
      "\n",
      "\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error: database disk image is malformed\n",
     "output_type": "error",
     "traceback": [
      "Error: database disk image is malformed\nTraceback:\n",
      "1. getSRAdbFile()",
      "2. dbGetQuery(con, \"SELECT * FROM metaInfo\")",
      "3. dbGetQuery(con, \"SELECT * FROM metaInfo\")",
      "4. .local(conn, statement, ...)",
      "5. dbSendQuery(conn, statement, ...)",
      "6. dbSendQuery(conn, statement, ...)",
      "7. .local(conn, statement, ...)",
      "8. new(\"SQLiteResult\", sql = statement, ptr = result_create(conn@ptr, \n .     statement), conn = conn, bigint = conn@bigint)",
      "9. initialize(value, ...)",
      "10. initialize(value, ...)",
      "11. result_create(conn@ptr, statement)"
     ]
    }
   ],
   "source": [
    "library(SRAdb)\n",
    "sqlFile <- getSRAdbFile() # this code helps to download the metadata file from sra server.\n",
    "sraCon <- dbConnect(SQLite(),sqlFile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<SQLiteConnection>\n",
       "  Path: F:\\SRAmetadb.sqlite\n",
       "  Extensions: TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Create a connection to the `SQLite` file for querying.\n",
    "sra_dbname <- \"F:/SRAmetadb.sqlite\"\n",
    "sracon <- dbConnect(SQLite(),sra_dbname) \n",
    "sracon <- dbConnect(dbDriver(\"SQLite\"), sra_dbname)\n",
    "sracon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error: no such table: study\n",
     "output_type": "error",
     "traceback": [
      "Error: no such table: study\nTraceback:\n",
      "1. dbListFields(sracon, \"study\")",
      "2. dbListFields(sracon, \"study\")",
      "3. list_fields(conn, name)",
      "4. dbSendQuery(conn, paste(\"SELECT * FROM \", dbQuoteIdentifier(conn, \n .     name), \"LIMIT 0\"))",
      "5. dbSendQuery(conn, paste(\"SELECT * FROM \", dbQuoteIdentifier(conn, \n .     name), \"LIMIT 0\"))",
      "6. .local(conn, statement, ...)",
      "7. new(\"SQLiteResult\", sql = statement, ptr = result_create(conn@ptr, \n .     statement), conn = conn, bigint = conn@bigint)",
      "8. initialize(value, ...)",
      "9. initialize(value, ...)",
      "10. result_create(conn@ptr, statement)"
     ]
    }
   ],
   "source": [
    "# To investigate the content of the database, define the queries that should be run, determine the fields in the data:\n",
    "sraTables <- dbListTables(sracon)\n",
    "dbListFields(sracon,\"study\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "character(0)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sraTables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## frame the first query to get the accessions and titles from the SRA study where,\n",
    "## the study type contains the keyword `embryo`. \n",
    "## The query can be seen as a typical SQL query:\n",
    "myHit <- dbGetQuery(sracon, \n",
    "                    paste(\"select study_accession,study_title from study where\",\"study_description like'%embryo'\", sep=\" \"))\n",
    "myHit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To do a free text search for the terms of interest\n",
    "myHit <- getSRA( search_terms = \"brain\", out_types = c('run','study'), sracon)\n",
    "head(myHit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine key words for appropriate logic, and take a look at some of the retrieved results.\n",
    "myHit <- getSRA( search_terms ='Alzheimers OR \"EPILEPSY\"', out_types = c('sample'), sracon)\n",
    "head(myHit)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Downloading data from the SRA database\n",
    "\n",
    "After querying the `SRA` database, we can use the `SRAdb` package to download the `FASTQ` data based on the queries.\n",
    "Before that, we will need the queries we want to search for or the ID of the data we want to retrieve."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the SRAdb library at first.\n",
    "library(SRAdb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Frame a query. For instance, use the same query as in the previous recipe to search for `ALZHEIMERS` or `EPILEPSY`，\n",
    "## note that the step requires the `sracon` object created in the previous recipe.\n",
    "sra_dbname <- 'F:/SRAmetadb.sqlite'\n",
    "sracon <- dbConnect(SQLite(),sra_dbname) # sracon <- dbConnect(dbDriver(\"SQLite\"), sra_dbname)\n",
    "sracon\n",
    "myHit <- getSRA( search_terms ='ALZHEIMERS OR \"EPILEPSY\"', out_types = c('sample'), sracon)\n",
    "head(myHit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Choose two of the hits and find their accessions (IDs) using the `sraConvert` function. \n",
    "## And take a look at the object created\n",
    "conversion <- sraConvert( c('ERS354366','SRS266589'), sra_con = sracon)\n",
    "conversion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## To get information on one of the experiments, use the `getSRAinfo` function\n",
    "rs <- getSRAinfo( c(\"SRX100465\"), sracon, sraType = \"sra\")\n",
    "rs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Use the `getSRAfile` function to download the data of interest (the run). \n",
    "## Due to some problems, the `download.file()` can't work, so I download it on the internet via the `ftp site`\n",
    "getSRAfile( c(\"SRR351672\", \"SRR351673\"), sracon, fileType = 'fastq')  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This part deviates in the terms of the conversion of the IDs from the results of the query for free text search. This gives us the accession IDs for the hits retrieved from SRA.<br>\n",
    "The `getSRAfile` function fetches the file from the SRA database."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Reading FASTQ files in R\n",
    "\n",
    "A `FASTQ` file, can have several short sequence reads in it. It is the universally accepted format in the NGS community and for most alignment programs such as Bowtie, which uses these files as input. In order to analyze the data, we need to read in this data in the workspace. <br>\n",
    "We can download a file from the SRA database, or use the example file in the `ShortRead` package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# download files from within R\n",
    "install.packages(\"R.utils\")\n",
    "library(R.utils)\n",
    "download.file(url=\"ftp://ftp.ddbj.nig.ac.jp/ddbj_database/dra/fastq/SRA146/SRA146985/SRX495419/SRR1198924.fastq.bz2\", \n",
    "              destfile = \"F:/SRR1198924.fastq.bz2\")\n",
    "# It is a `.bz` file, use \"bunzip()\" to decompress\n",
    "# download.file(url=\"ftp://ftp.ddbj.nig.ac.jp/ddbj_database/dra/fastq/SRA000/SRA000241/SRA000241.run.xml\", \n",
    "#              destfile = \"D:/Try-practice/Chapter-8/3-read fastq files/SRA000241.xml\")    \n",
    "# As before there are `fastq` files, but now they are `xml` files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# unzip the downloaded file.\n",
    "bunzip2(file = \"F:/SRR1198924.fastq.bz2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "download.file(url=\"ftp://ftp.ddbj.nig.ac.jp/ddbj_database/dra/fastq/SRA000/SRA000241/SRA000241.run.xml\", \n",
    "             destfile = \"F:/SRA000241.xml\")  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ShortRead)  # load the ShortRead package"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To read the downloaded FASTQ file as an R object, use the `readFastq` function. Note that `FASTQ` files are usually heavy and might require huge memory space while reading (it might need approximately 8 GB); therefore, check the machines before working on such files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "MyFastq <- readFastq(\"F:/\", pattern=\".fastq\")\n",
    "MyFastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Look at the first four lines of the file by the generic `readLines` function\n",
    "readLines(\"F:/SRR1198924.fastq\", 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(XML) \n",
    "xmlread <- xmlTreeParse(readLines(\"F:/SRA000241.xml\"), asText=TRUE) # readLines()????????\n",
    "xmlread"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xmltop <- xmlRoot(xmlread) # Get the top-level XML node.#形成根目录列表数据\n",
    "xmltop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class(xmltop) # 对象中的类\n",
    "xmlName(xmltop) #第一层（根目录）节点的名称\n",
    "xmlSize(xmltop) #根目录大小，多少\n",
    "xmlName(xmltop[[1]]) # 查看子目录名"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 匹配查询序列信息\n",
    "getNodeSet(xmltop, \"/RUN_SET//PRIMARY_ID\")   # subset by primary ID\n",
    "getNodeSet(xmltop, \"/RUN_SET//SUBMITTER_ID\")\n",
    "getNodeSet(xmltop, \"/RUN_SET//RUN_ATTRIBUTE\")\n",
    "srr_648 <- getNodeSet(xmltop, \"/RUN_SET//RUN[@accession = 'SRR000648']\") # srr_648 <- xmltop[2]\n",
    "srr_648 <- xmlToDataFrame(srr_648)\n",
    "names(srr_648) <- NULL\n",
    "class(srr_648)\n",
    "srr_648 <- as.character(unlist(srr_648[4]))\n",
    "class(srr_648)\n",
    "srr_648 <- strsplit(srr_648, split = \"sequence\")\n",
    "list(srr_648)\n",
    "srr_648 <- srr_648[[1]]\n",
    "srr_648[2]\n",
    "srr_648 <- strsplit(srr_648[2], split=\"key\")\n",
    "class(648)\n",
    "srr_648 <- unlist(srr_648)\n",
    "srr_648[1]\n",
    "srr_648 <- as.character(srr_648[1])\n",
    "srr_648"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "srr_657 <- getNodeSet(xmltop, \"/RUN_SET//RUN[@accession = 'SRR000657']\") # srr_657 <- xmltop[11]\n",
    "srr_657 <- xmlToDataFrame(srr_657)\n",
    "names(srr_657) <- NULL\n",
    "class(srr_657)\n",
    "srr_657 <- as.character(unlist(srr_657[4]))\n",
    "class(srr_657)\n",
    "srr_657 <- strsplit(srr_657, split = \"sequence\")\n",
    "list(srr_657)\n",
    "srr_657 <- srr_657[[1]]\n",
    "srr_657[2]\n",
    "srr_657 <- strsplit(srr_657[2], split=\"key\")\n",
    "class(srr_657)\n",
    "srr_657 <- unlist(srr_657)\n",
    "srr_657[1]\n",
    "srr_657 <- as.character(srr_657[1])\n",
    "srr_657\n",
    "srr_648"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "example <- rbind(srr_648, srr_657)\n",
    "example\n",
    "class(example)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 3.6.2",
   "language": "R",
   "name": "ir362"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
