{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The contents include:\n",
    "\n",
    "* 1. The SNP association analysis\n",
    "* 2. Running association scans for SNPs\n",
    "* 3. The whole genome SNP association analysis\n",
    "\n",
    "## 1. The SNP association analysis\n",
    "\n",
    "GWAS is used to identify genetic variants associated with a disease phenotype. It's performed by statistical analysis of the GWAS data. To perform the SNP association analysis, we take the data built in package `SNPsocc` as an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Bioconductor version 3.9 (BiocManager 1.30.10), R 3.6.2 (2019-12-12)\n",
      "\n",
      "Installing package(s) 'SNPassoc'\n",
      "\n",
      "also installing the dependencies 'polspline', 'rms', 'haplo.stats'\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "package 'polspline' successfully unpacked and MD5 sums checked\n",
      "package 'rms' successfully unpacked and MD5 sums checked\n",
      "package 'haplo.stats' successfully unpacked and MD5 sums checked\n",
      "package 'SNPassoc' successfully unpacked and MD5 sums checked\n",
      "\n",
      "The downloaded binary packages are in\n",
      "\tC:\\Users\\Master\\AppData\\Local\\Temp\\Rtmpktwtnd\\downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Old packages: 'BH', 'bit', 'caTools', 'cli', 'FactoMineR', 'fansi', 'farver',\n",
      "  'GetoptLong', 'ggridges', 'gplots', 'hms', 'knitr', 'leaps', 'multcomp',\n",
      "  'mvtnorm', 'PMA', 'precrec', 'prettyunits', 'pROC', 'RCurl', 'RSQLite',\n",
      "  'Rttf2pt1', 'shinyjs', 'stringi', 'tinytex', 'xfun', 'XML', 'xts', 'zoo'\n",
      "\n",
      "Loading required package: haplo.stats\n",
      "\n",
      "Loading required package: survival\n",
      "\n",
      "Loading required package: mvtnorm\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "Registered S3 method overwritten by 'SNPassoc':\n",
      "  method            from       \n",
      "  summary.haplo.glm haplo.stats\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 40</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>id</th><th scope=col>casco</th><th scope=col>sex</th><th scope=col>blood.pre</th><th scope=col>protein</th><th scope=col>snp10001</th><th scope=col>snp10002</th><th scope=col>snp10003</th><th scope=col>snp10004</th><th scope=col>snp10005</th><th scope=col>...</th><th scope=col>snp100026</th><th scope=col>snp100027</th><th scope=col>snp100028</th><th scope=col>snp100029</th><th scope=col>snp100030</th><th scope=col>snp100031</th><th scope=col>snp100032</th><th scope=col>snp100033</th><th scope=col>snp100034</th><th scope=col>snp100035</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>...</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td>1</td><td>Female</td><td>13.7</td><td>75640.52</td><td>TT</td><td>CC</td><td>GG</td><td>GG</td><td>GG</td><td>...</td><td>GG</td><td>CC</td><td>CC</td><td>GG</td><td>AA</td><td>TT</td><td>AA</td><td>AA</td><td>TT</td><td>TT</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>2</td><td>1</td><td>Female</td><td>12.7</td><td>28688.22</td><td>TT</td><td>AC</td><td>GG</td><td>GG</td><td>AG</td><td>...</td><td>GG</td><td>CG</td><td>CT</td><td>GG</td><td>AA</td><td>TT</td><td>AG</td><td>AG</td><td>TT</td><td>TT</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>3</td><td>1</td><td>Female</td><td>12.9</td><td>17279.59</td><td>TT</td><td>CC</td><td>GG</td><td>GG</td><td>GG</td><td>...</td><td>GG</td><td>CC</td><td>CC</td><td>GG</td><td>AA</td><td>TT</td><td>AA</td><td>AA</td><td>TT</td><td>TT</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>4</td><td>1</td><td>Male  </td><td>14.6</td><td>27253.99</td><td>CT</td><td>CC</td><td>GG</td><td>GG</td><td>GG</td><td>...</td><td>GG</td><td>CC</td><td>CT</td><td>AG</td><td>AA</td><td>TT</td><td>AG</td><td>AG</td><td>CT</td><td>TT</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>5</td><td>1</td><td>Female</td><td>13.4</td><td>38066.57</td><td>TT</td><td>AC</td><td>GG</td><td>GG</td><td>GG</td><td>...</td><td>GG</td><td>CG</td><td>CT</td><td>GG</td><td>AA</td><td>TT</td><td>AG</td><td>AG</td><td>TT</td><td>TT</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>6</td><td>1</td><td>Female</td><td>11.3</td><td> 9872.46</td><td>TT</td><td>CC</td><td>GG</td><td>GG</td><td>GG</td><td>...</td><td>GG</td><td>CC</td><td>CC</td><td>GG</td><td>AA</td><td>TT</td><td>AA</td><td>AA</td><td>TT</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 40\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & id & casco & sex & blood.pre & protein & snp10001 & snp10002 & snp10003 & snp10004 & snp10005 & ... & snp100026 & snp100027 & snp100028 & snp100029 & snp100030 & snp100031 & snp100032 & snp100033 & snp100034 & snp100035\\\\\n",
       "  & <int> & <dbl> & <fct> & <dbl> & <dbl> & <fct> & <fct> & <fct> & <fct> & <fct> & ... & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct>\\\\\n",
       "\\hline\n",
       "\t1 & 1 & 1 & Female & 13.7 & 75640.52 & TT & CC & GG & GG & GG & ... & GG & CC & CC & GG & AA & TT & AA & AA & TT & TT\\\\\n",
       "\t2 & 2 & 1 & Female & 12.7 & 28688.22 & TT & AC & GG & GG & AG & ... & GG & CG & CT & GG & AA & TT & AG & AG & TT & TT\\\\\n",
       "\t3 & 3 & 1 & Female & 12.9 & 17279.59 & TT & CC & GG & GG & GG & ... & GG & CC & CC & GG & AA & TT & AA & AA & TT & TT\\\\\n",
       "\t4 & 4 & 1 & Male   & 14.6 & 27253.99 & CT & CC & GG & GG & GG & ... & GG & CC & CT & AG & AA & TT & AG & AG & CT & TT\\\\\n",
       "\t5 & 5 & 1 & Female & 13.4 & 38066.57 & TT & AC & GG & GG & GG & ... & GG & CG & CT & GG & AA & TT & AG & AG & TT & TT\\\\\n",
       "\t6 & 6 & 1 & Female & 11.3 &  9872.46 & TT & CC & GG & GG & GG & ... & GG & CC & CC & GG & AA & TT & AA & AA & TT & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 40\n",
       "\n",
       "| <!--/--> | id &lt;int&gt; | casco &lt;dbl&gt; | sex &lt;fct&gt; | blood.pre &lt;dbl&gt; | protein &lt;dbl&gt; | snp10001 &lt;fct&gt; | snp10002 &lt;fct&gt; | snp10003 &lt;fct&gt; | snp10004 &lt;fct&gt; | snp10005 &lt;fct&gt; | ... ... | snp100026 &lt;fct&gt; | snp100027 &lt;fct&gt; | snp100028 &lt;fct&gt; | snp100029 &lt;fct&gt; | snp100030 &lt;fct&gt; | snp100031 &lt;fct&gt; | snp100032 &lt;fct&gt; | snp100033 &lt;fct&gt; | snp100034 &lt;fct&gt; | snp100035 &lt;fct&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1 | 1 | Female | 13.7 | 75640.52 | TT | CC | GG | GG | GG | ... | GG | CC | CC | GG | AA | TT | AA | AA | TT | TT |\n",
       "| 2 | 2 | 1 | Female | 12.7 | 28688.22 | TT | AC | GG | GG | AG | ... | GG | CG | CT | GG | AA | TT | AG | AG | TT | TT |\n",
       "| 3 | 3 | 1 | Female | 12.9 | 17279.59 | TT | CC | GG | GG | GG | ... | GG | CC | CC | GG | AA | TT | AA | AA | TT | TT |\n",
       "| 4 | 4 | 1 | Male   | 14.6 | 27253.99 | CT | CC | GG | GG | GG | ... | GG | CC | CT | AG | AA | TT | AG | AG | CT | TT |\n",
       "| 5 | 5 | 1 | Female | 13.4 | 38066.57 | TT | AC | GG | GG | GG | ... | GG | CG | CT | GG | AA | TT | AG | AG | TT | TT |\n",
       "| 6 | 6 | 1 | Female | 11.3 |  9872.46 | TT | CC | GG | GG | GG | ... | GG | CC | CC | GG | AA | TT | AA | AA | TT | NA |\n",
       "\n"
      ],
      "text/plain": [
       "  id casco sex    blood.pre protein  snp10001 snp10002 snp10003 snp10004\n",
       "1 1  1     Female 13.7      75640.52 TT       CC       GG       GG      \n",
       "2 2  1     Female 12.7      28688.22 TT       AC       GG       GG      \n",
       "3 3  1     Female 12.9      17279.59 TT       CC       GG       GG      \n",
       "4 4  1     Male   14.6      27253.99 CT       CC       GG       GG      \n",
       "5 5  1     Female 13.4      38066.57 TT       AC       GG       GG      \n",
       "6 6  1     Female 11.3       9872.46 TT       CC       GG       GG      \n",
       "  snp10005 ... snp100026 snp100027 snp100028 snp100029 snp100030 snp100031\n",
       "1 GG       ... GG        CC        CC        GG        AA        TT       \n",
       "2 AG       ... GG        CG        CT        GG        AA        TT       \n",
       "3 GG       ... GG        CC        CC        GG        AA        TT       \n",
       "4 GG       ... GG        CC        CT        AG        AA        TT       \n",
       "5 GG       ... GG        CG        CT        GG        AA        TT       \n",
       "6 GG       ... GG        CC        CC        GG        AA        TT       \n",
       "  snp100032 snp100033 snp100034 snp100035\n",
       "1 AA        AA        TT        TT       \n",
       "2 AG        AG        TT        TT       \n",
       "3 AA        AA        TT        TT       \n",
       "4 AG        AG        CT        TT       \n",
       "5 AG        AG        TT        TT       \n",
       "6 AA        AA        TT        NA       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>snp</th><th scope=col>chr</th><th scope=col>pos</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>snp10001</td><td>Chr1</td><td>2987398</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>snp10002</td><td>Chr1</td><td>1913558</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>snp10003</td><td>Chr1</td><td>1982067</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>snp10004</td><td>Chr1</td><td> 447403</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>snp10005</td><td>Chr1</td><td>2212031</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>snp10006</td><td>Chr1</td><td>2515720</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & snp & chr & pos\\\\\n",
       "  & <fct> & <fct> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & snp10001 & Chr1 & 2987398\\\\\n",
       "\t2 & snp10002 & Chr1 & 1913558\\\\\n",
       "\t3 & snp10003 & Chr1 & 1982067\\\\\n",
       "\t4 & snp10004 & Chr1 &  447403\\\\\n",
       "\t5 & snp10005 & Chr1 & 2212031\\\\\n",
       "\t6 & snp10006 & Chr1 & 2515720\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| <!--/--> | snp &lt;fct&gt; | chr &lt;fct&gt; | pos &lt;dbl&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | snp10001 | Chr1 | 2987398 |\n",
       "| 2 | snp10002 | Chr1 | 1913558 |\n",
       "| 3 | snp10003 | Chr1 | 1982067 |\n",
       "| 4 | snp10004 | Chr1 |  447403 |\n",
       "| 5 | snp10005 | Chr1 | 2212031 |\n",
       "| 6 | snp10006 | Chr1 | 2515720 |\n",
       "\n"
      ],
      "text/plain": [
       "  snp      chr  pos    \n",
       "1 snp10001 Chr1 2987398\n",
       "2 snp10002 Chr1 1913558\n",
       "3 snp10003 Chr1 1982067\n",
       "4 snp10004 Chr1  447403\n",
       "5 snp10005 Chr1 2212031\n",
       "6 snp10006 Chr1 2515720"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "BiocManager::install(\"SNPassoc\")\n",
    "library(SNPassoc)\n",
    "data(SNPs)               # load the data\n",
    "table1 <- head(SNPs)               # take a look at the objects\n",
    "table1\n",
    "table2 <- head(SNPs.info.pos)\n",
    "table2 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The SNP data is a data.frame object that consists of 157 rows and 40 columns. Each row of the data represents a sample that is either a case or control. The first five columns represent the attributes for each sample and the columns from 6:40 are the SNP information. That is why we used these columns while creating the SNP class object. The first five columns are for sample identifiers, for case or control identification (1 is for case and 0 is for control, which means casco). This is followed by information about the blood pressure and protein level of each gender. The table1 shows the first six entries of the SNP data.<br>\n",
    "\n",
    "Table2 shows information about the SNP name, chromosome name, and genomic position."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A setupSNP: 6 × 40</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>id</th><th scope=col>casco</th><th scope=col>sex</th><th scope=col>blood.pre</th><th scope=col>protein</th><th scope=col>snp10001</th><th scope=col>snp10002</th><th scope=col>snp10003</th><th scope=col>snp10004</th><th scope=col>snp10005</th><th scope=col>...</th><th scope=col>snp100026</th><th scope=col>snp100027</th><th scope=col>snp100028</th><th scope=col>snp100029</th><th scope=col>snp100030</th><th scope=col>snp100031</th><th scope=col>snp100032</th><th scope=col>snp100033</th><th scope=col>snp100034</th><th scope=col>snp100035</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>...</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td>1</td><td>Female</td><td>13.7</td><td>75640.52</td><td>T/T</td><td>C/C</td><td>G/G</td><td>G/G</td><td>G/G</td><td>...</td><td>G/G</td><td>C/C</td><td>C/C</td><td>G/G</td><td>A/A</td><td>T/T</td><td>A/A</td><td>A/A</td><td>T/T</td><td>T/T</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>2</td><td>1</td><td>Female</td><td>12.7</td><td>28688.22</td><td>T/T</td><td>A/C</td><td>G/G</td><td>G/G</td><td>A/G</td><td>...</td><td>G/G</td><td>C/G</td><td>C/T</td><td>G/G</td><td>A/A</td><td>T/T</td><td>A/G</td><td>A/G</td><td>T/T</td><td>T/T</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>3</td><td>1</td><td>Female</td><td>12.9</td><td>17279.59</td><td>T/T</td><td>C/C</td><td>G/G</td><td>G/G</td><td>G/G</td><td>...</td><td>G/G</td><td>C/C</td><td>C/C</td><td>G/G</td><td>A/A</td><td>T/T</td><td>A/A</td><td>A/A</td><td>T/T</td><td>T/T</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>4</td><td>1</td><td>Male  </td><td>14.6</td><td>27253.99</td><td>C/T</td><td>C/C</td><td>G/G</td><td>G/G</td><td>G/G</td><td>...</td><td>G/G</td><td>C/C</td><td>C/T</td><td>A/G</td><td>A/A</td><td>T/T</td><td>A/G</td><td>A/G</td><td>C/T</td><td>T/T</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>5</td><td>1</td><td>Female</td><td>13.4</td><td>38066.57</td><td>T/T</td><td>A/C</td><td>G/G</td><td>G/G</td><td>G/G</td><td>...</td><td>G/G</td><td>C/G</td><td>C/T</td><td>G/G</td><td>A/A</td><td>T/T</td><td>A/G</td><td>A/G</td><td>T/T</td><td>T/T</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>6</td><td>1</td><td>Female</td><td>11.3</td><td> 9872.46</td><td>T/T</td><td>C/C</td><td>G/G</td><td>G/G</td><td>G/G</td><td>...</td><td>G/G</td><td>C/C</td><td>C/C</td><td>G/G</td><td>A/A</td><td>T/T</td><td>A/A</td><td>A/A</td><td>T/T</td><td>NA </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A setupSNP: 6 × 40\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & id & casco & sex & blood.pre & protein & snp10001 & snp10002 & snp10003 & snp10004 & snp10005 & ... & snp100026 & snp100027 & snp100028 & snp100029 & snp100030 & snp100031 & snp100032 & snp100033 & snp100034 & snp100035\\\\\n",
       "  & <int> & <dbl> & <fct> & <dbl> & <dbl> & <fct> & <fct> & <fct> & <fct> & <fct> & ... & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct>\\\\\n",
       "\\hline\n",
       "\t1 & 1 & 1 & Female & 13.7 & 75640.52 & T/T & C/C & G/G & G/G & G/G & ... & G/G & C/C & C/C & G/G & A/A & T/T & A/A & A/A & T/T & T/T\\\\\n",
       "\t2 & 2 & 1 & Female & 12.7 & 28688.22 & T/T & A/C & G/G & G/G & A/G & ... & G/G & C/G & C/T & G/G & A/A & T/T & A/G & A/G & T/T & T/T\\\\\n",
       "\t3 & 3 & 1 & Female & 12.9 & 17279.59 & T/T & C/C & G/G & G/G & G/G & ... & G/G & C/C & C/C & G/G & A/A & T/T & A/A & A/A & T/T & T/T\\\\\n",
       "\t4 & 4 & 1 & Male   & 14.6 & 27253.99 & C/T & C/C & G/G & G/G & G/G & ... & G/G & C/C & C/T & A/G & A/A & T/T & A/G & A/G & C/T & T/T\\\\\n",
       "\t5 & 5 & 1 & Female & 13.4 & 38066.57 & T/T & A/C & G/G & G/G & G/G & ... & G/G & C/G & C/T & G/G & A/A & T/T & A/G & A/G & T/T & T/T\\\\\n",
       "\t6 & 6 & 1 & Female & 11.3 &  9872.46 & T/T & C/C & G/G & G/G & G/G & ... & G/G & C/C & C/C & G/G & A/A & T/T & A/A & A/A & T/T & NA \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A setupSNP: 6 × 40\n",
       "\n",
       "| <!--/--> | id &lt;int&gt; | casco &lt;dbl&gt; | sex &lt;fct&gt; | blood.pre &lt;dbl&gt; | protein &lt;dbl&gt; | snp10001 &lt;fct&gt; | snp10002 &lt;fct&gt; | snp10003 &lt;fct&gt; | snp10004 &lt;fct&gt; | snp10005 &lt;fct&gt; | ... ... | snp100026 &lt;fct&gt; | snp100027 &lt;fct&gt; | snp100028 &lt;fct&gt; | snp100029 &lt;fct&gt; | snp100030 &lt;fct&gt; | snp100031 &lt;fct&gt; | snp100032 &lt;fct&gt; | snp100033 &lt;fct&gt; | snp100034 &lt;fct&gt; | snp100035 &lt;fct&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1 | 1 | Female | 13.7 | 75640.52 | T/T | C/C | G/G | G/G | G/G | ... | G/G | C/C | C/C | G/G | A/A | T/T | A/A | A/A | T/T | T/T |\n",
       "| 2 | 2 | 1 | Female | 12.7 | 28688.22 | T/T | A/C | G/G | G/G | A/G | ... | G/G | C/G | C/T | G/G | A/A | T/T | A/G | A/G | T/T | T/T |\n",
       "| 3 | 3 | 1 | Female | 12.9 | 17279.59 | T/T | C/C | G/G | G/G | G/G | ... | G/G | C/C | C/C | G/G | A/A | T/T | A/A | A/A | T/T | T/T |\n",
       "| 4 | 4 | 1 | Male   | 14.6 | 27253.99 | C/T | C/C | G/G | G/G | G/G | ... | G/G | C/C | C/T | A/G | A/A | T/T | A/G | A/G | C/T | T/T |\n",
       "| 5 | 5 | 1 | Female | 13.4 | 38066.57 | T/T | A/C | G/G | G/G | G/G | ... | G/G | C/G | C/T | G/G | A/A | T/T | A/G | A/G | T/T | T/T |\n",
       "| 6 | 6 | 1 | Female | 11.3 |  9872.46 | T/T | C/C | G/G | G/G | G/G | ... | G/G | C/C | C/C | G/G | A/A | T/T | A/A | A/A | T/T | NA  |\n",
       "\n"
      ],
      "text/plain": [
       "  id casco sex    blood.pre protein  snp10001 snp10002 snp10003 snp10004\n",
       "1 1  1     Female 13.7      75640.52 T/T      C/C      G/G      G/G     \n",
       "2 2  1     Female 12.7      28688.22 T/T      A/C      G/G      G/G     \n",
       "3 3  1     Female 12.9      17279.59 T/T      C/C      G/G      G/G     \n",
       "4 4  1     Male   14.6      27253.99 C/T      C/C      G/G      G/G     \n",
       "5 5  1     Female 13.4      38066.57 T/T      A/C      G/G      G/G     \n",
       "6 6  1     Female 11.3       9872.46 T/T      C/C      G/G      G/G     \n",
       "  snp10005 ... snp100026 snp100027 snp100028 snp100029 snp100030 snp100031\n",
       "1 G/G      ... G/G       C/C       C/C       G/G       A/A       T/T      \n",
       "2 A/G      ... G/G       C/G       C/T       G/G       A/A       T/T      \n",
       "3 G/G      ... G/G       C/C       C/C       G/G       A/A       T/T      \n",
       "4 G/G      ... G/G       C/C       C/T       A/G       A/A       T/T      \n",
       "5 G/G      ... G/G       C/G       C/T       G/G       A/A       T/T      \n",
       "6 G/G      ... G/G       C/C       C/C       G/G       A/A       T/T      \n",
       "  snp100032 snp100033 snp100034 snp100035\n",
       "1 A/A       A/A       T/T       T/T      \n",
       "2 A/G       A/G       T/T       T/T      \n",
       "3 A/A       A/A       T/T       T/T      \n",
       "4 A/G       A/G       C/T       T/T      \n",
       "5 A/G       A/G       T/T       T/T      \n",
       "6 A/A       A/A       T/T       NA       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Create an `SNP` object using the `setupSNP` function with the `SNP` information in columns 6 to 40\n",
    "mySNP <- setupSNP(SNPs, 6:40, sep=\"\")\n",
    "head(mySNP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "SNP: snp10001  adjusted by: sex blood.pre \n",
       "              0    %   1    %   OR lower upper p-value   AIC\n",
       "Codominant                                                  \n",
       "T/T          24 51.1  68 61.8 1.00             0.15410 195.8\n",
       "C/T          21 44.7  32 29.1 0.55  0.26  1.14              \n",
       "C/C           2  4.3  10  9.1 1.74  0.35  8.63              \n",
       "Dominant                                                    \n",
       "T/T          24 51.1  68 61.8 1.00             0.22859 196.1\n",
       "C/T-C/C      23 48.9  42 38.2 0.65  0.32  1.31              \n",
       "Recessive                                                   \n",
       "T/T-C/T      45 95.7 100 90.9 1.00             0.28494 196.4\n",
       "C/C           2  4.3  10  9.1 2.22  0.46 10.70              \n",
       "Overdominant                                                \n",
       "T/T-C/C      26 55.3  78 70.9 1.00             0.07188 194.3\n",
       "C/T          21 44.7  32 29.1 0.52  0.25  1.06              \n",
       "log-Additive                                                \n",
       "0,1,2        47 29.9 110 70.1 0.87  0.51  1.49 0.60861 197.3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## To do an association test for the trait and variables of your interest with `association` function\n",
    "myres <- association(casco~sex+snp10001+blood.pre, data = mySNP,\n",
    "                     model.interaction = c(\"dominant\",\"codominant\"))\n",
    "myres   # Take a look at the results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `association` function perfroms an association analysis between a single SNP and a dependent variable via model fitting and computing statistics. The arguments include the model to be fitted (formula object). This needs the train of interest and the SNP under investigation separated by a `~` operator. It is possible to add other variables to the formula using the `+` operator. Adding more SNPs is possible, but the analysis is done only for the first SNP and adjusted by the others. The `association` function can model the dependence based on the codominant, dominant, recessive, overdominant, and log-additive genetic models of inheritance. This is supplied to the function under the `model.interaction` argument.\n",
    "\n",
    "In this recipe, we tried to find the dependence of case and control on the first SNP—snp10001—the sex, and the blood pressure. We used two genetic models: dominant and codominant. The result looks like a matrix. The columns contain information about the sample size and percentages for each genotype, the odds ratio and its 95 percent confidence interval (which takes the most frequent homozygous genotype as the reference), and the p-value corresponding to the likelihood ratio test obtained from a comparison with the null model. Besides, the matrix also has the `Akaike Information Criterion`(AIC) of each genetic model. Thus, the `myres` table reflects the dependence of phenotypes on SNP\n",
    "(together with other factors), which shows the result of the association analysis test for `snp10001`.\n",
    "There are other packages for performing similar analyses. Some examples are `snp.plotter` and `GenABEL`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Running association scans for SNPs\n",
    "\n",
    "\"The SNP association analysis\" presented the approach to discover the dependence of a phenotype on a single SNP. But what if we have to keep an eye on all of the SNPs and scan the entire set? The following will discuss computing the significance of the dependence of phenotypes on all SNPs available in the data.<br>\n",
    "\n",
    "Also take the `SNPassoc` and the data as an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "SNP: snp10001  adjusted by: sex blood.pre \n",
       "              0    %   1    %   OR lower upper p-value   AIC\n",
       "Codominant                                                  \n",
       "T/T          24 51.1  68 61.8 1.00             0.15410 195.8\n",
       "C/T          21 44.7  32 29.1 0.55  0.26  1.14              \n",
       "C/C           2  4.3  10  9.1 1.74  0.35  8.63              \n",
       "Dominant                                                    \n",
       "T/T          24 51.1  68 61.8 1.00             0.22859 196.1\n",
       "C/T-C/C      23 48.9  42 38.2 0.65  0.32  1.31              \n",
       "Recessive                                                   \n",
       "T/T-C/T      45 95.7 100 90.9 1.00             0.28494 196.4\n",
       "C/C           2  4.3  10  9.1 2.22  0.46 10.70              \n",
       "Overdominant                                                \n",
       "T/T-C/C      26 55.3  78 70.9 1.00             0.07188 194.3\n",
       "C/T          21 44.7  32 29.1 0.52  0.25  1.06              \n",
       "log-Additive                                                \n",
       "0,1,2        47 29.9 110 70.1 0.87  0.51  1.49 0.60861 197.3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(SNPassoc)\n",
    "data(SNPs) \n",
    "mySNP <- setupSNP(SNPs, 6:40, sep=\"\")\n",
    "myres <- association(casco~sex+snp10001+blood.pre, data = mySNP,\n",
    "                     model.interaction = c(\"dominant\",\"codominant\"))\n",
    "myres"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run the `WGassociation` function on the `mySNP` object for all of the genetic models. \n",
    "## The step can also be run avoiding the formula by just providing the name of the phenotype \n",
    "## under study since the function runs across all of the SNPs anyway.<br>\n",
    "myres <- WGassociation(protein, data = mySNP, model = \"all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A WGassociation: 6 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>comments</th><th scope=col>codominant</th><th scope=col>dominant</th><th scope=col>recessive</th><th scope=col>overdominant</th><th scope=col>log-additive</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>snp10001</th><td>NA         </td><td>0.005861132</td><td>0.005164929</td><td>0.01604756</td><td>0.1230616</td><td>0.001403687</td></tr>\n",
       "\t<tr><th scope=row>snp10002</th><td>NA         </td><td>0.785547500</td><td>0.933029300</td><td>0.48703940</td><td>0.8728754</td><td>0.768461800</td></tr>\n",
       "\t<tr><th scope=row>snp10003</th><td>Monomorphic</td><td>         NA</td><td>         NA</td><td>        NA</td><td>       NA</td><td>         NA</td></tr>\n",
       "\t<tr><th scope=row>snp10004</th><td>Monomorphic</td><td>         NA</td><td>         NA</td><td>        NA</td><td>       NA</td><td>         NA</td></tr>\n",
       "\t<tr><th scope=row>snp10005</th><td>NA         </td><td>0.633062200</td><td>0.438808500</td><td>0.50130180</td><td>0.5542676</td><td>0.372673000</td></tr>\n",
       "\t<tr><th scope=row>snp10006</th><td>Monomorphic</td><td>         NA</td><td>         NA</td><td>        NA</td><td>       NA</td><td>         NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A WGassociation: 6 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & comments & codominant & dominant & recessive & overdominant & log-additive\\\\\n",
       "  & <fct> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tsnp10001 & NA          & 0.005861132 & 0.005164929 & 0.01604756 & 0.1230616 & 0.001403687\\\\\n",
       "\tsnp10002 & NA          & 0.785547500 & 0.933029300 & 0.48703940 & 0.8728754 & 0.768461800\\\\\n",
       "\tsnp10003 & Monomorphic &          NA &          NA &         NA &        NA &          NA\\\\\n",
       "\tsnp10004 & Monomorphic &          NA &          NA &         NA &        NA &          NA\\\\\n",
       "\tsnp10005 & NA          & 0.633062200 & 0.438808500 & 0.50130180 & 0.5542676 & 0.372673000\\\\\n",
       "\tsnp10006 & Monomorphic &          NA &          NA &         NA &        NA &          NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A WGassociation: 6 × 6\n",
       "\n",
       "| <!--/--> | comments &lt;fct&gt; | codominant &lt;dbl&gt; | dominant &lt;dbl&gt; | recessive &lt;dbl&gt; | overdominant &lt;dbl&gt; | log-additive &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| snp10001 | NA          | 0.005861132 | 0.005164929 | 0.01604756 | 0.1230616 | 0.001403687 |\n",
       "| snp10002 | NA          | 0.785547500 | 0.933029300 | 0.48703940 | 0.8728754 | 0.768461800 |\n",
       "| snp10003 | Monomorphic |          NA |          NA |         NA |        NA |          NA |\n",
       "| snp10004 | Monomorphic |          NA |          NA |         NA |        NA |          NA |\n",
       "| snp10005 | NA          | 0.633062200 | 0.438808500 | 0.50130180 | 0.5542676 | 0.372673000 |\n",
       "| snp10006 | Monomorphic |          NA |          NA |         NA |        NA |          NA |\n",
       "\n"
      ],
      "text/plain": [
       "         comments    codominant  dominant    recessive  overdominant\n",
       "snp10001 NA          0.005861132 0.005164929 0.01604756 0.1230616   \n",
       "snp10002 NA          0.785547500 0.933029300 0.48703940 0.8728754   \n",
       "snp10003 Monomorphic          NA          NA         NA        NA   \n",
       "snp10004 Monomorphic          NA          NA         NA        NA   \n",
       "snp10005 NA          0.633062200 0.438808500 0.50130180 0.5542676   \n",
       "snp10006 Monomorphic          NA          NA         NA        NA   \n",
       "         log-additive\n",
       "snp10001 0.001403687 \n",
       "snp10002 0.768461800 \n",
       "snp10003          NA \n",
       "snp10004          NA \n",
       "snp10005 0.372673000 \n",
       "snp10006          NA "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "myres <- WGassociation(protein~1, data = mySNP, model = \"all\")\n",
    "head(myres)           # Take a look at the result matrix computed "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `WGassociation` function works in a similar way to the `association` function. The main difference is that it performs many association tests in parallel using the `mclapply` function. The function takes each SNP as a variable that the phenotype depends on and runs the association function on each of them in parallel. These parallel runs return a matrix for each SNP that the corresponding p-values are extracted from, and this ultimately returns as the overall result. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>0.005164929</li>\n",
       "\t<li>0.9330293</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>0.4388085</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>0.30005423</li>\n",
       "\t<li>0.8720384</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>0.1292051</li>\n",
       "\t<li>0.5835992</li>\n",
       "\t<li>0.09877879</li>\n",
       "\t<li>0.01511489</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>0.7912645</li>\n",
       "\t<li>0.8869233</li>\n",
       "\t<li>0.007494522</li>\n",
       "\t<li>0.31380895</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>0.9954381</li>\n",
       "\t<li>0.017737798</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>0.7149631</li>\n",
       "\t<li>0.00504041</li>\n",
       "\t<li>0.021833942</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>0.006561046</li>\n",
       "\t<li>0.004546931</li>\n",
       "\t<li>0.021833942</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0.005164929\n",
       "\\item 0.9330293\n",
       "\\item <NA>\n",
       "\\item <NA>\n",
       "\\item 0.4388085\n",
       "\\item <NA>\n",
       "\\item <NA>\n",
       "\\item 0.30005423\n",
       "\\item 0.8720384\n",
       "\\item <NA>\n",
       "\\item 0.1292051\n",
       "\\item 0.5835992\n",
       "\\item 0.09877879\n",
       "\\item 0.01511489\n",
       "\\item <NA>\n",
       "\\item <NA>\n",
       "\\item 0.7912645\n",
       "\\item 0.8869233\n",
       "\\item 0.007494522\n",
       "\\item 0.31380895\n",
       "\\item <NA>\n",
       "\\item <NA>\n",
       "\\item 0.9954381\n",
       "\\item 0.017737798\n",
       "\\item <NA>\n",
       "\\item <NA>\n",
       "\\item 0.7149631\n",
       "\\item 0.00504041\n",
       "\\item 0.021833942\n",
       "\\item <NA>\n",
       "\\item <NA>\n",
       "\\item 0.006561046\n",
       "\\item 0.004546931\n",
       "\\item 0.021833942\n",
       "\\item <NA>\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0.005164929\n",
       "2. 0.9330293\n",
       "3. &lt;NA&gt;\n",
       "4. &lt;NA&gt;\n",
       "5. 0.4388085\n",
       "6. &lt;NA&gt;\n",
       "7. &lt;NA&gt;\n",
       "8. 0.30005423\n",
       "9. 0.8720384\n",
       "10. &lt;NA&gt;\n",
       "11. 0.1292051\n",
       "12. 0.5835992\n",
       "13. 0.09877879\n",
       "14. 0.01511489\n",
       "15. &lt;NA&gt;\n",
       "16. &lt;NA&gt;\n",
       "17. 0.7912645\n",
       "18. 0.8869233\n",
       "19. 0.007494522\n",
       "20. 0.31380895\n",
       "21. &lt;NA&gt;\n",
       "22. &lt;NA&gt;\n",
       "23. 0.9954381\n",
       "24. 0.017737798\n",
       "25. &lt;NA&gt;\n",
       "26. &lt;NA&gt;\n",
       "27. 0.7149631\n",
       "28. 0.00504041\n",
       "29. 0.021833942\n",
       "30. &lt;NA&gt;\n",
       "31. &lt;NA&gt;\n",
       "32. 0.006561046\n",
       "33. 0.004546931\n",
       "34. 0.021833942\n",
       "35. &lt;NA&gt;\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] 0.005164929 0.933029300          NA          NA 0.438808500          NA\n",
       " [7]          NA 0.300054230 0.872038400          NA 0.129205100 0.583599200\n",
       "[13] 0.098778790 0.015114890          NA          NA 0.791264500 0.886923300\n",
       "[19] 0.007494522 0.313808950          NA          NA 0.995438100 0.017737798\n",
       "[25]          NA          NA 0.714963100 0.005040410 0.021833942          NA\n",
       "[31]          NA 0.006561046 0.004546931 0.021833942          NA"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## To get the p-values for specific genetic models,\n",
    "## use the name of the mode as the function and the complete result matrix as the argument\n",
    "\n",
    "dominant(myres)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>0.016047565</li>\n",
       "\t<li>0.4870394</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>0.5013018</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>0.08652457</li>\n",
       "\t<li>0.4781536</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>0.2728859</li>\n",
       "\t<li>0.4792929</li>\n",
       "\t<li>0.11039996</li>\n",
       "\t<li>0.11952358</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>0.4596597</li>\n",
       "\t<li>0.4781536</li>\n",
       "\t<li>0.147754655</li>\n",
       "\t<li>0.08652457</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>0.4819481</li>\n",
       "\t<li>0.005475794</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>0.9490049</li>\n",
       "\t<li>0.08758219</li>\n",
       "\t<li>0.005475794</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "\t<li>0.09655548</li>\n",
       "\t<li>0.068387908</li>\n",
       "\t<li>0.005475794</li>\n",
       "\t<li>&lt;NA&gt;</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0.016047565\n",
       "\\item 0.4870394\n",
       "\\item <NA>\n",
       "\\item <NA>\n",
       "\\item 0.5013018\n",
       "\\item <NA>\n",
       "\\item <NA>\n",
       "\\item 0.08652457\n",
       "\\item 0.4781536\n",
       "\\item <NA>\n",
       "\\item 0.2728859\n",
       "\\item 0.4792929\n",
       "\\item 0.11039996\n",
       "\\item 0.11952358\n",
       "\\item <NA>\n",
       "\\item <NA>\n",
       "\\item 0.4596597\n",
       "\\item 0.4781536\n",
       "\\item 0.147754655\n",
       "\\item 0.08652457\n",
       "\\item <NA>\n",
       "\\item <NA>\n",
       "\\item 0.4819481\n",
       "\\item 0.005475794\n",
       "\\item <NA>\n",
       "\\item <NA>\n",
       "\\item 0.9490049\n",
       "\\item 0.08758219\n",
       "\\item 0.005475794\n",
       "\\item <NA>\n",
       "\\item <NA>\n",
       "\\item 0.09655548\n",
       "\\item 0.068387908\n",
       "\\item 0.005475794\n",
       "\\item <NA>\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0.016047565\n",
       "2. 0.4870394\n",
       "3. &lt;NA&gt;\n",
       "4. &lt;NA&gt;\n",
       "5. 0.5013018\n",
       "6. &lt;NA&gt;\n",
       "7. &lt;NA&gt;\n",
       "8. 0.08652457\n",
       "9. 0.4781536\n",
       "10. &lt;NA&gt;\n",
       "11. 0.2728859\n",
       "12. 0.4792929\n",
       "13. 0.11039996\n",
       "14. 0.11952358\n",
       "15. &lt;NA&gt;\n",
       "16. &lt;NA&gt;\n",
       "17. 0.4596597\n",
       "18. 0.4781536\n",
       "19. 0.147754655\n",
       "20. 0.08652457\n",
       "21. &lt;NA&gt;\n",
       "22. &lt;NA&gt;\n",
       "23. 0.4819481\n",
       "24. 0.005475794\n",
       "25. &lt;NA&gt;\n",
       "26. &lt;NA&gt;\n",
       "27. 0.9490049\n",
       "28. 0.08758219\n",
       "29. 0.005475794\n",
       "30. &lt;NA&gt;\n",
       "31. &lt;NA&gt;\n",
       "32. 0.09655548\n",
       "33. 0.068387908\n",
       "34. 0.005475794\n",
       "35. &lt;NA&gt;\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] 0.016047565 0.487039400          NA          NA 0.501301800          NA\n",
       " [7]          NA 0.086524570 0.478153600          NA 0.272885900 0.479292900\n",
       "[13] 0.110399960 0.119523580          NA          NA 0.459659700 0.478153600\n",
       "[19] 0.147754655 0.086524570          NA          NA 0.481948100 0.005475794\n",
       "[25]          NA          NA 0.949004900 0.087582190 0.005475794          NA\n",
       "[31]          NA 0.096555480 0.068387908 0.005475794          NA"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "recessive(myres)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "$snp10001\n",
      "\n",
      "SNP: snp10001  adjusted by: \n",
      "               n    me   se    dif  lower   upper  p-value  AIC\n",
      "Codominant                                                     \n",
      "T/T           92 47419 2393      0                0.005861 3602\n",
      "C/T           53 38987 3177  -8432 -16165  -698.2              \n",
      "C/C           12 27413 6061 -20006 -33770 -6241.5              \n",
      "Dominant                                                       \n",
      "T/T           92 47419 2393      0                0.005165 3603\n",
      "C/T-C/C       65 36851 2858 -10568 -17870 -3266.8              \n",
      "Recessive                                                      \n",
      "T/T-C/T      145 44337 1935      0                0.016048 3605\n",
      "C/C           12 27413 6061 -16924 -30549 -3298.9              \n",
      "Overdominant                                                   \n",
      "T/T-C/C      104 45111 2308      0                0.123062 3608\n",
      "C/T           53 38987 3177  -6123 -13864  1617.1              \n",
      "log-Additive                                                   \n",
      "0,1,2                        -9332 -14955 -3708.6 0.001404 3601\n",
      "\n",
      "$snp10002\n",
      "\n",
      "SNP: snp10002  adjusted by: \n",
      "               n    me   se    dif  lower upper p-value  AIC\n",
      "Codominant                                                  \n",
      "C/C           74 42876 2890    0.0               0.7855 3612\n",
      "A/C           78 42740 2576 -135.8  -7648  7377             \n",
      "A/A            5 50262 6879 7385.6 -14006 28777             \n",
      "Dominant                                                    \n",
      "C/C           74 42876 2890    0.0               0.9330 3611\n",
      "A/C-A/A       83 43193 2456  317.3  -7072  7706             \n",
      "Recessive                                                   \n",
      "C/C-A/C      152 42806 1924    0.0               0.4870 3610\n",
      "A/A            5 50262 6879 7455.3 -13518 28429             \n",
      "Overdominant                                                \n",
      "C/C-A/A       79 43343 2742    0.0               0.8729 3611\n",
      "A/C           78 42740 2576 -603.2  -7980  6773             \n",
      "log-Additive                                                \n",
      "0,1,2                        996.5  -5626  7619  0.7685 3611\n",
      "\n",
      "$snp10003\n",
      "\n",
      "SNP: snp10003  adjusted by: \n",
      "Monomorphic\n",
      "\n",
      "$snp10004\n",
      "\n",
      "SNP: snp10004  adjusted by: \n",
      "Monomorphic\n",
      "\n",
      "$snp10005\n",
      "\n",
      "SNP: snp10005  adjusted by: \n",
      "               n    me    se   dif  lower upper p-value  AIC\n",
      "Codominant                                                  \n",
      "G/G           84 41685  2598     0               0.6331 3612\n",
      "A/G           70 44286  2804  2601  -4880 10083             \n",
      "A/A            3 52119 10583 10434 -16729 37598             \n",
      "Dominant                                                    \n",
      "G/G           84 41685  2598     0               0.4388 3610\n",
      "A/G-A/A       73 44608  2718  2923  -4458 10304             \n",
      "Recessive                                                   \n",
      "G/G-A/G      154 42867  1902     0               0.5013 3610\n",
      "A/A            3 52119 10583  9252 -17651 36155             \n",
      "Overdominant                                                \n",
      "G/G-A/A       87 42044  2534     0               0.5543 3611\n",
      "A/G           70 44286  2804  2241  -5171  9654             \n",
      "log-Additive                                                \n",
      "0,1,2                         3127  -3729  9984  0.3727 3610\n",
      "\n",
      "$snp10006\n",
      "\n",
      "SNP: snp10006  adjusted by: \n",
      "Monomorphic\n",
      "\n",
      "$snp10007\n",
      "\n",
      "SNP: snp10007  adjusted by: \n",
      "Monomorphic\n",
      "\n",
      "$snp10008\n",
      "\n",
      "SNP: snp10008  adjusted by: \n",
      "               n    me   se      dif  lower upper p-value  AIC\n",
      "Codominant                                                    \n",
      "C/C          104 44436 2239      0.0              0.20627 3610\n",
      "C/G           44 42421 3699  -2015.3 -10269  6238             \n",
      "G/G            9 30004 8185 -14432.0 -30379  1514             \n",
      "Dominant                                                      \n",
      "C/C          104 44436 2239      0.0              0.30005 3610\n",
      "C/G-G/G       53 40312 3401  -4123.8 -11897  3649             \n",
      "Recessive                                                     \n",
      "C/C-C/G      148 43837 1914      0.0              0.08652 3608\n",
      "G/G            9 30004 8185 -13832.9 -29550  1884             \n",
      "Overdominant                                                  \n",
      "C/C-G/G      113 43286 2182      0.0              0.83655 3611\n",
      "C/G           44 42421 3699   -865.8  -9078  7346             \n",
      "log-Additive                                                  \n",
      "0,1,2                        -4721.9 -10880  1436 0.13493 3609\n",
      "\n",
      "$snp10009\n",
      "\n",
      "SNP: snp10009  adjusted by: \n",
      "               n    me   se     dif  lower upper p-value  AIC\n",
      "Codominant                                                   \n",
      "A/A           72 43225 2942     0.0               0.7474 3589\n",
      "A/G           79 42130 2553 -1095.1  -8637  6446             \n",
      "G/G            5 50262 6879  7036.7 -14370 28443             \n",
      "Dominant                                                     \n",
      "A/A           72 43225 2942     0.0               0.8720 3588\n",
      "A/G-G/G       84 42614 2437  -611.0  -8034  6812             \n",
      "Recessive                                                    \n",
      "A/A-A/G      151 42652 1931     0.0               0.4782 3588\n",
      "G/G            5 50262 6879  7609.7 -13367 28586             \n",
      "Overdominant                                                 \n",
      "A/A-G/G       77 43682 2786     0.0               0.6815 3588\n",
      "A/G           79 42130 2553 -1552.0  -8950  5846             \n",
      "log-Additive                                                 \n",
      "0,1,2                         272.5  -6384  6929  0.9362 3588\n",
      "\n",
      "$snp100010\n",
      "\n",
      "SNP: snp100010  adjusted by: \n",
      "Monomorphic\n",
      "\n",
      "$snp100011\n",
      "\n",
      "SNP: snp100011  adjusted by: \n",
      "               n    me   se    dif  lower upper p-value  AIC\n",
      "Codominant                                                  \n",
      "G/G          154 43441 1898      0               0.3054 3610\n",
      "C/G            2 25303  796 -18138 -50884 14608             \n",
      "C/C            1 17280    0 -26162 -72323 19999             \n",
      "Dominant                                                    \n",
      "G/G          154 43441 1898      0               0.1292 3609\n",
      "C/G-C/C        3 22629 2714 -20813 -47555  5930             \n",
      "Recessive                                                   \n",
      "G/G-C/G      156 43209 1881      0               0.2729 3610\n",
      "C/C            1 17280    0 -25929 -72115 20256             \n",
      "Overdominant                                                \n",
      "G/G-C/C      155 43273 1893      0               0.2841 3610\n",
      "C/G            2 25303  796 -17969 -50739 14800             \n",
      "log-Additive                                                \n",
      "0,1,2                       -14752 -33641  4136  0.1279 3609\n",
      "\n",
      "$snp100012\n",
      "\n",
      "SNP: snp100012  adjusted by: \n",
      "               n    me    se   dif  lower upper p-value  AIC\n",
      "Codominant                                                  \n",
      "G/G           84 41685  2598     0               0.7057 3564\n",
      "C/G           68 43388  2807  1704  -5795  9203             \n",
      "C/C            3 52119 10583 10434 -16576 37444             \n",
      "Dominant                                                    \n",
      "G/G           84 41685  2598     0               0.5836 3563\n",
      "C/G-C/C       71 43757  2721  2073  -5323  9469             \n",
      "Recessive                                                   \n",
      "G/G-C/G      152 42447  1902     0               0.4793 3562\n",
      "C/C            3 52119 10583  9672 -17058 36402             \n",
      "Overdominant                                                \n",
      "G/G-C/C       87 42044  2534     0               0.7234 3563\n",
      "C/G           68 43388  2807  1344  -6086  8774             \n",
      "log-Additive                                                \n",
      "0,1,2                         2422  -4438  9283  0.4899 3563\n",
      "\n",
      "$snp100013\n",
      "\n",
      "SNP: snp100013  adjusted by: \n",
      "               n    me   se    dif  lower   upper p-value  AIC\n",
      "Codominant                                                    \n",
      "A/A          101 44101 2274      0                0.14971 3330\n",
      "A/G           35 39006 3968  -5096 -13978 3786.72             \n",
      "G/G            9 30004 8185 -14097 -29851 1655.53             \n",
      "Dominant                                                      \n",
      "A/A          101 44101 2274      0                0.09878 3329\n",
      "A/G-G/G       44 37164 3572  -6937 -15119 1245.49             \n",
      "Recessive                                                     \n",
      "A/A-A/G      136 42790 1976      0                0.11040 3330\n",
      "G/G            9 30004 8185 -12786 -28387 2814.54             \n",
      "Overdominant                                                  \n",
      "A/A-G/G      110 42948 2213      0                0.38420 3331\n",
      "A/G           35 39006 3968  -3942 -12794 4909.72             \n",
      "log-Additive                                                  \n",
      "0,1,2                        -6209 -12492   74.44 0.05475 3328\n",
      "\n",
      "$snp100014\n",
      "\n",
      "SNP: snp100014  adjusted by: \n",
      "               n    me   se   dif   lower upper p-value  AIC\n",
      "Codominant                                                  \n",
      "C/C           52 36181 3505     0               0.03796 3515\n",
      "A/C           74 44761 2280  8580   359.1 16802             \n",
      "A/A           27 48998 5330 12817  2039.5 23594             \n",
      "Dominant                                                    \n",
      "C/C           52 36181 3505     0               0.01511 3514\n",
      "A/C-A/A      101 45894 2188  9713  1967.0 17459             \n",
      "Recessive                                                   \n",
      "C/C-A/C      126 41220 1998     0               0.11952 3517\n",
      "A/A           27 48998 5330  7777 -1958.7 17513             \n",
      "Overdominant                                                \n",
      "C/C-A/A       79 40561 3000     0               0.27139 3518\n",
      "A/C           74 44761 2280  4200 -3257.1 11657             \n",
      "log-Additive                                                \n",
      "0,1,2                        6759  1521.6 11996 0.01245 3513\n",
      "\n",
      "$snp100015\n",
      "\n",
      "SNP: snp100015  adjusted by: \n",
      "               n    me   se   dif lower upper p-value  AIC\n",
      "Codominant                                                \n",
      "G/G          144 41795 1879     0             0.02626 3606\n",
      "A/G           13 56875 8303 15080  1908 28253             \n",
      "log-Additive                                              \n",
      "0,1,2                       15080  1908 28253         3606\n",
      "\n",
      "$snp100016\n",
      "\n",
      "SNP: snp100016  adjusted by: \n",
      "Monomorphic\n",
      "\n",
      "$snp100017\n",
      "\n",
      "SNP: snp100017  adjusted by: \n",
      "               n    me   se       dif  lower upper p-value  AIC\n",
      "Codominant                                                     \n",
      "T/T           67 43206 3131     0.000               0.7064 3564\n",
      "C/T           83 41713 2403 -1492.727  -9043  6057             \n",
      "C/C            5 50262 6879  7055.818 -14255 28367             \n",
      "Dominant                                                       \n",
      "T/T           67 43206 3131     0.000               0.7913 3563\n",
      "C/T-C/C       88 42199 2303 -1007.014  -8451  6437             \n",
      "Recessive                                                      \n",
      "T/T-C/T      150 42380 1924     0.000               0.4597 3562\n",
      "C/C            5 50262 6879  7881.794 -12958 28721             \n",
      "Overdominant                                                   \n",
      "T/T-C/C       72 43696 2952     0.000               0.5997 3563\n",
      "C/T           83 41713 2403 -1982.715  -9372  5407             \n",
      "log-Additive                                                   \n",
      "0,1,2                          -3.555  -6688  6681  0.9992 3563\n",
      "\n",
      "$snp100018\n",
      "\n",
      "SNP: snp100018  adjusted by: \n",
      "               n    me   se     dif  lower upper p-value  AIC\n",
      "Codominant                                                   \n",
      "T/T           67 43206 3131     0.0               0.7529 3589\n",
      "C/T           84 42210 2426  -995.6  -8578  6586             \n",
      "C/C            5 50262 6879  7055.8 -14403 28515             \n",
      "Dominant                                                     \n",
      "T/T           67 43206 3131     0.0               0.8869 3588\n",
      "C/T-C/C       89 42662 2324  -543.3  -8019  6933             \n",
      "Recessive                                                    \n",
      "T/T-C/T      151 42652 1931     0.0               0.4782 3588\n",
      "C/C            5 50262 6879  7609.7 -13367 28586             \n",
      "Overdominant                                                 \n",
      "T/T-C/C       72 43696 2952     0.0               0.6953 3588\n",
      "C/T           84 42210 2426 -1485.6  -8906  5934             \n",
      "log-Additive                                                 \n",
      "0,1,2                         339.2  -6377  7056  0.9213 3588\n",
      "\n",
      "$snp100019\n",
      "\n",
      "SNP: snp100019  adjusted by: \n",
      "               n    me   se   dif lower upper  p-value  AIC\n",
      "Codominant                                                 \n",
      "G/G           50 35755 3484     0             0.024010 3605\n",
      "C/G           75 45610 2265  9855  1592 18118              \n",
      "C/C           32 48418 4909 12664  2418 22909              \n",
      "Dominant                                                   \n",
      "G/G           50 35755 3484     0             0.007495 3604\n",
      "C/G-C/C      107 46450 2152 10695  2959 18432              \n",
      "Recessive                                                  \n",
      "G/G-C/G      125 41668 1986     0             0.147755 3609\n",
      "C/C           32 48418 4909  6750 -2344 15845              \n",
      "Overdominant                                               \n",
      "G/G-C/C       82 40696 2924     0             0.191682 3609\n",
      "C/G           75 45610 2265  4914 -2430 12257              \n",
      "log-Additive                                               \n",
      "0,1,2                        6711  1651 11771 0.010244 3604\n",
      "\n",
      "$snp100020\n",
      "\n",
      "SNP: snp100020  adjusted by: \n",
      "               n    me   se      dif  lower upper p-value  AIC\n",
      "Codominant                                                    \n",
      "G/G          105 44378 2219      0.0              0.21002 3610\n",
      "A/G           43 42516 3785  -1862.1 -10172  6448             \n",
      "A/A            9 30004 8185 -14373.9 -30316  1568             \n",
      "Dominant                                                      \n",
      "G/G          105 44378 2219      0.0              0.31381 3610\n",
      "A/G-A/A       52 40350 3466  -4027.6 -11839  3784             \n",
      "Recessive                                                     \n",
      "G/G-A/G      148 43837 1914      0.0              0.08652 3608\n",
      "A/A            9 30004 8185 -13832.9 -29550  1884             \n",
      "Overdominant                                                  \n",
      "G/G-A/A      114 43243 2163      0.0              0.86338 3611\n",
      "A/G           43 42516 3785   -727.3  -8998  7543             \n",
      "log-Additive                                                  \n",
      "0,1,2                        -4655.2 -10827  1517 0.14135 3609\n",
      "\n",
      "$snp100021\n",
      "\n",
      "SNP: snp100021  adjusted by: \n",
      "Monomorphic\n",
      "\n",
      "$snp100022\n",
      "\n",
      "SNP: snp100022  adjusted by: \n",
      "Monomorphic\n",
      "\n",
      "$snp100023\n",
      "\n",
      "SNP: snp100023  adjusted by: \n",
      "               n    me   se     dif  lower upper p-value  AIC\n",
      "Codominant                                                   \n",
      "T/T           71 42925 2968    0.00               0.7754 3546\n",
      "A/T           78 42431 2601 -493.69  -8137  7150             \n",
      "A/A            5 50262 6879 7336.83 -14225 28898             \n",
      "Dominant                                                     \n",
      "T/T           71 42925 2968    0.00               0.9954 3544\n",
      "A/T-A/A       83 42903 2480  -21.98  -7543  7499             \n",
      "Recessive                                                    \n",
      "T/T-A/T      149 42666 1957    0.00               0.4819 3544\n",
      "A/A            5 50262 6879 7595.28 -13523 28714             \n",
      "Overdominant                                                 \n",
      "T/T-A/A       76 43407 2809    0.00               0.7989 3544\n",
      "A/T           78 42431 2601 -976.38  -8474  6521             \n",
      "log-Additive                                                 \n",
      "0,1,2                        752.45  -5982  7487  0.8269 3544\n",
      "\n",
      "$snp100024\n",
      "\n",
      "SNP: snp100024  adjusted by: \n",
      "               n    me   se    dif  lower upper  p-value  AIC\n",
      "Codominant                                                   \n",
      "T/T           91 46651 2339      0              0.007236 3580\n",
      "C/T           51 40730 3423  -5921 -13775  1934              \n",
      "C/C           14 26373 5560 -20278 -33169 -7387              \n",
      "Dominant                                                     \n",
      "T/T           91 46651 2339      0              0.017738 3582\n",
      "C/T-C/C       65 37638 3013  -9013 -16384 -1643              \n",
      "Recessive                                                    \n",
      "T/T-C/T      142 44525 1946      0              0.005476 3580\n",
      "C/C           14 26373 5560 -18151 -30778 -5524              \n",
      "Overdominant                                                 \n",
      "T/T-C/C      105 43948 2253      0              0.424421 3587\n",
      "C/T           51 40730 3423  -3217 -11090  4656              \n",
      "log-Additive                                                 \n",
      "0,1,2                        -8554 -14037 -3071 0.002632 3579\n",
      "\n",
      "$snp100025\n",
      "\n",
      "SNP: snp100025  adjusted by: \n",
      "Monomorphic\n",
      "\n",
      "$snp100026\n",
      "\n",
      "SNP: snp100026  adjusted by: \n",
      "Monomorphic\n",
      "\n",
      "$snp100027\n",
      "\n",
      "SNP: snp100027  adjusted by: \n",
      "               n    me   se     dif  lower upper p-value  AIC\n",
      "Codominant                                                   \n",
      "C/C           68 43579 3107     0.0               0.9285 3567\n",
      "C/G           82 42103 2455 -1476.6  -9093  6140             \n",
      "G/G            5 43460 8071  -119.2 -21636 21398             \n",
      "Dominant                                                     \n",
      "C/C           68 43579 3107     0.0               0.7150 3566\n",
      "C/G-G/G       87 42181 2350 -1398.6  -8891  6094             \n",
      "Recessive                                                    \n",
      "C/C-C/G      150 42772 1940     0.0               0.9490 3566\n",
      "G/G            5 43460 8071   688.1 -20363 21739             \n",
      "Overdominant                                                 \n",
      "C/C-G/G       73 43571 2936     0.0               0.6997 3565\n",
      "C/G           82 42103 2455 -1468.4  -8916  5979             \n",
      "log-Additive                                                 \n",
      "0,1,2                       -1056.1  -7780  5668  0.7586 3565\n",
      "\n",
      "$snp100028\n",
      "\n",
      "SNP: snp100028  adjusted by: \n",
      "               n    me   se   dif lower upper  p-value  AIC\n",
      "Codominant                                                 \n",
      "T/T           50 35399 3489     0             0.014803 3582\n",
      "C/T           72 45474 2317 10075  1744 18406              \n",
      "C/C           34 49158 4656 13759  3699 23818              \n",
      "Dominant                                                   \n",
      "T/T           50 35399 3489     0             0.005040 3581\n",
      "C/T-C/C      106 46656 2163 11256  3503 19010              \n",
      "Recessive                                                  \n",
      "T/T-C/T      122 41345 2020     0             0.087582 3586\n",
      "C/C           34 49158 4656  7813 -1094 16720              \n",
      "Overdominant                                               \n",
      "T/T-C/C       84 40968 2883     0             0.235350 3588\n",
      "C/T           72 45474 2317  4506 -2907 11918              \n",
      "log-Additive                                               \n",
      "0,1,2                        7166  2184 12148 0.005452 3581\n",
      "\n",
      "$snp100029\n",
      "\n",
      "SNP: snp100029  adjusted by: \n",
      "               n    me   se    dif  lower upper  p-value  AIC\n",
      "Codominant                                                   \n",
      "G/G           94 46388 2280      0              0.008589 3580\n",
      "A/G           48 40875 3611  -5513 -13488  2462              \n",
      "A/A           14 26373 5560 -20015 -32893 -7137              \n",
      "Dominant                                                     \n",
      "G/G           94 46388 2280      0              0.021834 3583\n",
      "A/G-A/A       62 37601 3142  -8787 -16221 -1353              \n",
      "Recessive                                                    \n",
      "G/G-A/G      142 44525 1946      0              0.005476 3580\n",
      "A/A           14 26373 5560 -18151 -30778 -5524              \n",
      "Overdominant                                                 \n",
      "G/G-A/A      108 43794 2200      0              0.476008 3588\n",
      "A/G           48 40875 3611  -2918 -10923  5087              \n",
      "log-Additive                                                 \n",
      "0,1,2                        -8355 -13847 -2864 0.003331 3579\n",
      "\n",
      "$snp100030\n",
      "\n",
      "SNP: snp100030  adjusted by: \n",
      "Monomorphic\n",
      "\n",
      "$snp100031\n",
      "\n",
      "SNP: snp100031  adjusted by: \n",
      "[1] Genot 65\\\\%\n",
      "\n",
      "$snp100032\n",
      "\n",
      "SNP: snp100032  adjusted by: \n",
      "               n    me   se   dif lower upper  p-value  AIC\n",
      "Codominant                                                 \n",
      "G/G           52 36085 3400     0             0.019427 3581\n",
      "A/G           70 45699 2316  9614  1364 17865              \n",
      "A/A           34 49158 4656 13073  3134 23013              \n",
      "Dominant                                                   \n",
      "G/G           52 36085 3400     0             0.006561 3580\n",
      "A/G-A/A      104 46830 2171 10745  3103 18387              \n",
      "Recessive                                                  \n",
      "G/G-A/G      122 41601 2005     0             0.096555 3584\n",
      "A/A           34 49158 4656  7557 -1302 16415              \n",
      "Overdominant                                               \n",
      "G/G-A/A       86 41253 2829     0             0.239934 3586\n",
      "A/G           70 45699 2316  4446 -2940 11832              \n",
      "log-Additive                                               \n",
      "0,1,2                        6833  1919 11747 0.007170 3580\n",
      "\n",
      "$snp100033\n",
      "\n",
      "SNP: snp100033  adjusted by: \n",
      "               n    me   se   dif   lower upper  p-value  AIC\n",
      "Codominant                                                   \n",
      "G/G           49 34773 3502     0               0.012527 3492\n",
      "A/G           69 44984 2400 10211  1716.3 18706              \n",
      "A/A           34 49158 4656 14386  4235.8 24535              \n",
      "Dominant                                                     \n",
      "G/G           49 34773 3502     0               0.004547 3491\n",
      "A/G-A/A      103 46362 2219 11589  3704.7 19474              \n",
      "Recessive                                                    \n",
      "G/G-A/G      118 40744 2065     0               0.068388 3496\n",
      "A/A           34 49158 4656  8414  -569.8 17399              \n",
      "Overdominant                                                 \n",
      "G/G-A/A       83 40665 2902     0               0.265441 3498\n",
      "A/G           69 44984 2400  4319 -3253.7 11891              \n",
      "log-Additive                                                 \n",
      "0,1,2                        7445  2414.8 12475 0.004282 3491\n",
      "\n",
      "$snp100034\n",
      "\n",
      "SNP: snp100034  adjusted by: \n",
      "               n    me   se    dif  lower upper  p-value  AIC\n",
      "Codominant                                                   \n",
      "T/T           94 46388 2280      0              0.008589 3580\n",
      "C/T           48 40875 3611  -5513 -13488  2462              \n",
      "C/C           14 26373 5560 -20015 -32893 -7137              \n",
      "Dominant                                                     \n",
      "T/T           94 46388 2280      0              0.021834 3583\n",
      "C/T-C/C       62 37601 3142  -8787 -16221 -1353              \n",
      "Recessive                                                    \n",
      "T/T-C/T      142 44525 1946      0              0.005476 3580\n",
      "C/C           14 26373 5560 -18151 -30778 -5524              \n",
      "Overdominant                                                 \n",
      "T/T-C/C      108 43794 2200      0              0.476008 3588\n",
      "C/T           48 40875 3611  -2918 -10923  5087              \n",
      "log-Additive                                                 \n",
      "0,1,2                        -8355 -13847 -2864 0.003331 3579\n",
      "\n",
      "$snp100035\n",
      "\n",
      "SNP: snp100035  adjusted by: \n",
      "Monomorphic\n",
      "\n",
      "attr(,\"label.SNPs\")\n",
      " [1] \"snp10001\"  \"snp10002\"  \"snp10003\"  \"snp10004\"  \"snp10005\"  \"snp10006\" \n",
      " [7] \"snp10007\"  \"snp10008\"  \"snp10009\"  \"snp100010\" \"snp100011\" \"snp100012\"\n",
      "[13] \"snp100013\" \"snp100014\" \"snp100015\" \"snp100016\" \"snp100017\" \"snp100018\"\n",
      "[19] \"snp100019\" \"snp100020\" \"snp100021\" \"snp100022\" \"snp100023\" \"snp100024\"\n",
      "[25] \"snp100025\" \"snp100026\" \"snp100027\" \"snp100028\" \"snp100029\" \"snp100030\"\n",
      "[31] \"snp100031\" \"snp100032\" \"snp100033\" \"snp100034\" \"snp100035\"\n",
      "attr(,\"models\")\n",
      "[1] 1 2 3 4 5\n",
      "attr(,\"quantitative\")\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " SNPs (n) Genot error (%) Monomorphic (%) Significant* (n) (%)\n",
      "       35               0            34.3                0   0\n",
      "\n",
      " *Number of statistically significant associations at level 1e-06\n",
      "Warning: No SNP is statistically significant after \n",
      "         Bonferroni Correction under codominant model \n",
      "Warning: No SNP is statistically significant after \n",
      "         Bonferroni Correction under dominant model \n",
      "Warning: No SNP is statistically significant after \n",
      "         Bonferroni Correction under recessive model \n",
      "Warning: No SNP is statistically significant after \n",
      "         Bonferroni Correction under overdominant model \n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAZlBMVEUAAAAAAP9NTU1NTf9o\naGhoaP98fHx8fP+MjIyMjP+ampqamv+np6enp/+ysrKysv+9vb29vf/Hx8fHx//Q0NDQ0P/Z\n2dnZ2f/h4eHh4f/p6enp6f/uqbjw8PDw8P//AAD/tcX///8N32XCAAAACXBIWXMAABJ0AAAS\ndAHeZh94AAAgAElEQVR4nO2di1rbPLOFZQIESCmlFGjL33zb93+T23IOPts6LMkjeb1PmxhH\nnhnJWtbBSqxKQog3au0ACMkBCokQABQSIQAoJEIAUEiEAKCQCAFAIRECgEIi4lGt19GP5nZY\nfOoDhUTEo+paalZVKSRCJlB1NaWQiDNFUXinuHDwC2VFGiEpddrSb+fN81/NZ5dDrlunVE3y\ntilgjEQwOCF97szSReX/NKPvnWSq9a88S0g1O7ufnQ+5bp6VV/ZtlNjKTyGJxry9iWcpPgMh\nleVQGU3CsuztVD0zFNLmoJA010ZlVkhKqRkhXT9tunrQvh2FtDrvj0Vx+FNvfu6LYv952v28\nKw7Hc/Vv9lc7jodi91r+2Re79/IsEP1Spb971Un+HIpz6tbuoiZ21kAYCanXtSu7QhpNjo6R\nrMnzqY7/aTZf9O693tqdKn9r/3ln8VjveW+EVKcvKsl8nhIXn53dyQvpOuqZEVK7RWpppT2y\nYtcuV96ryn6spLKvmhK9edzXonqtd+/ryt/eX+046GOq17eiuGsJSRvRO+6Kt1J/tu/uTldG\nnVkD1RJRV0haKO0Wqem4XXSorjYaU+AYyWo81vo4PlYvh7oZ+dQi0RI6bRfd/dWOr9Zr86Jb\ntJZYBrtTFpID0es1hbQyrQq+O21Wnbfr7vptuL/12lbJ+e34etgPdlNImTkkXfqtyIg0htuz\nQjoNnyikzB2SLvAWqRoS7d+OWxdSdCiklXmsB0CDMdLj9BipnBdS/7MMhKQmbvosTBdY1m0/\nKVBIK3OZtburpXKdnXu7bBfd/ctC2mnZPeclpNbrYDfQw7oWiB/nm0Sfzeaz3j1yH0nvXxbS\nS3E+8quzW9+lTXXZKoVEDHjfX1c26M39+2l3pZ7D11kKzf5lIZWvei3DUcumvfur6iw+RsqR\nMf9pRt87yRohdVd/q9aO8nSfqPX5tW6fu4WjSVtW+vbtoJCIeK5jpPP/Zo1Ce4ca+bxsHTSe\ndPRN2QuDQiLiubZIg2rf3TEii7Ic19NImqE5+xgJEcyMkJol37ZCGlssTiGRnJlvkUrnFmnq\nQwqJZIl/184kKYVEMqe5Idudc+jvGPm82Tme9LJ33JxNjM65I4RcoZAIAYAREuVINg6FRAgA\nCokQABQSIQAoJEIAUEiEAKCQCAFAIRECgEIiBACFRAgACokQABQSIQAwEpj+EQu+8z2f9xnY\nIhECgEIiBACFRAgACokQABQSIQAoJEIAUEiEAKCQCAFAIRECgEIiBACFRAgAiATsH8tESF4g\nJODygDNCsgKggPr3x6kksmkoJEIAsGtH8GywPnCygcBRG5zH5fQ3QaOuLxF8Rat6C54oJIIm\nopDitX1LnizCuD5+0MsKyZ9o1TueZBc9mUdxNUUhkQVidbjSFdLExWaLo0sigBS7duryRiGR\ntVGXNi9026dMPVkLqWNQXbCyRIgHTd3u7PS0OeZlzNGkBQtn04dQSCQKM3XbptYPjm1XX0sF\nNSYAxLxzQKSjOm84s4vV21VLTfV1kVDLhnd6Cok0wITUVGrjCu4kBET1BQqJSiIaNdhwN3Su\nWFaW3PplgGAB6SkkcgUlJI92wrr98p/8gy4RopJIuxJ4zqR5mVjSktOMwpw9pBUKieCE5N3h\nUhMdQ7SGTkahVqikzaMm/3CwhahPw6ltgNFRN0ArFNLmQQoJQauDGEpDjR+cFQElR6aIsZRU\nzf65ArHuzFBImyHKgkhxQoq1DpRCWp1I3zmIcmkeLlkL7NCAOOULFpKEgkuMSFfM5ISUWE2i\nkFYm3uqqCIod2ne/D+QVSHzQQkquANYm4jLF8H2cEQeuPlOrRxTSSlxvaMTq2kVgLB+OeUuu\nSOBCSq8IIjP4xlg+vwo4mhHH1XJegawAhRSUjkjcvjGWDnMLmiGmJIMXUoKFEIym2yZIQsHi\nmDDs4k9KWZlDIQUk4kTCIuGHsVNmHdwJKTMbAggpxWIIg4WQok5MR162ae8uwRpEIYXEYkYu\ncKGFP0PTNq29pViBQggpyYIIg/nAKGyZ9a0H8EYh4a2kWBCrE7LQgCsOzF04O0uy+mCC/p+m\neVe9v/lu8q6ExIF/t6sPgsthhiAtUprXFCgOk93hCg14o9TKhZuvNCsPhRQCt3tGoUoNeH/H\n1oeDq0TrThghpVoaIIQt1ATe4LF14eAq0apDIcFxX8IQ+f4O0B9OSKnWnEBCSrY8vPFaCRT3\n/g7O37IdIbcBwkEhQfFdUBf1/k7pH6+JD8MUdunEEawgky0RH7wzjS814OjFy4ahn2SrDYWE\nA3F5RxcbsNPlZ0LGksNwWIwC24/nM7CSbpm4Ea2XhDYXpxnFJZIJZl6SQoo4bodb83WJa2wS\nrjN+t8pmGqmUC8Ue4Lf2kOUWZWCCc5JylQnYH0m5WOzAfvk14n1SiEuck5RrDIXkT/xhDdyS\nh0/cHaKkK4xt8OPpbfbmQ6hfYVhjuOXuE6fWpOsLheSMir44LqQZV5+4VXRpV5eQQkq8aBZQ\n15cwpmNbcXSKa/bSri1Bz1naRbNAQCFhzGKukciD5lcrubiXQ9iLX+KFM0+4rl205Qaxvcyl\nTryuJCYkMb+yqAkYjL9liTqyvaefEoG74/iZ4dQL3JA463ZGDsM1MVbJkz+tSQkp5LBEGp7Z\ndJ86CDppHuuKG5/QE0T4e/4iijx8EFGW7YwehpvQNj9AxEn1IikhSenaRRmpRVm2M3ZY0PUQ\n2c7vpiUkGZMNkWKIsWxn5MCQDVLGt+6D3/vLoIy6RNNyjFU7w0PD6mjioAwqSTJCUjN/Wdta\n42If0ZXkJnPsoAx0FGE1CuYufd+KpxZcj47btYywasfDBi46CsnECuIm/ehlzFkM1xe3Q+Ox\nZpDLxQsTUg46iiAkbxeTpzTSHO9yIMHAjUJc7Mwb8rqOIQzJQryQ5k6nY81269qtMmGIm2B2\n8h6kpzHopbsaEkUEIYWcRHJUkupvGBzj5MkXe6/wG3dwP2r2z1QRLSSDem7XTgxTGx6/2v0r\n2NoBV//gfvXwUArJ3IpnuwFJNply/DeQemlMXeCxdI2PdKLUfCyiDElCrJDQva45vSwNq9c8\n2Xa+g0Rq+pOg5gZRhgQRRUgOHRTPHpttilmZ2YSCx64kIsXgOYGEMiQIkUKybwLmpw2N7E0k\nErG6z5RgsfZLwdORGtlKnDhCMrSi3K1N13ebMVSmNwsRQC8oarCRPIKE5HZ7pzl87GCzxmjB\nCKlBlo3qvaePHCEpb0Ouc9ujx4iSlJBgcFF49D2EEktIhku3AOG43G0djUbOWTYJJp7WhKha\nGJGENFkVWg+zwNRdmBU5SjIJJp7wRV1i5BBx1q496Tn6PBjYE+/87UgV0tTT3uLFK6pkBBFd\nSFOPUwIB6yBKqi1jwXQllaCQMushrt21g4PyJOo8G40w0ypgUVcqABaZ8Xo6H8fCgUmsgLPr\nIZrnZe4mWk4lQmJAIU1MGhCSOwGFZGHFxBHGTGrBbDTbJj3EhLJNIW3Ik6hgMst2rMmGDRat\nPE+igsks25Gmv7dYtPI8iQoms2xTSBvyJCqYzLJNIW3Ik6hgMss2hbQhT6KCySzbFNKGPIkK\nJrNsY4REyMahkAgBQCERAoBCIgQAhUQIAAqJEAAUEiEAKCRCAFBIhACgkAgBQCERAoBCIgQA\nhSSPoii8U1w4+IVCTKGQ5IET0ufOLB3xhkKSh3l7E88SWYBCkgeFlCAUkiied8XheK7+n/ui\n2H/qrWrH8VDsXss/+2L3Xp4Fol+q9HevOsmfQ3FO3dpd1KyUlY1BIUliryv+7lT5n08yeClr\nQeidxWO9570RUp2+qCTzeUpcfHZ2U0jxoJAE8Vpp4FirqWpi9Pax2v5Ta+NQvtevb0Vx1xLS\nUeut2nFXvJX6s313N2UUDQpJEPu6Sfmsq//hsn2o9fDVem1eKpG1xTLYTSFFg0ISxLne12+7\ny/busr/12lbJ+e34etgPdlNI0aCQBDGigdb2rJAei8uAiEJaBQpJEO4tUjUk2r8dKaT1oJAE\n8Tg9RirnhdT/jEKKDYUkiLfLTF1RS6g9a1fOC2mnZfdMIa0HhSSJkftIz6WJkF6K85Ffnd36\nLi2XrUaBQhJFpZ7D11kK73plg17HYDLZ8KrXMhy1bNq7v6rO4mPcHGwVCokQABQSIQAoJEIA\nUEiEAKCQCAFAIRECgEIiBACFRAgAjJAoR7JxKCRCAFBIhACgkAgBQCERAoBCIgQAhUQIAAqJ\nEAAUEiEAKCRCAFBIhACgkAgBQCERAgAjgf80fOd75u8zsEUiBACFRAgACokQABQSIQAoJEIA\nUEiEAKCQCCmVdw2mkIhY/Ku3qSP/KgwJNVqGyZYAVG9jR96uEJEqKongQVTveJ4AgdYXDiqJ\ngAkvJKXOLYCMrh2FRIIQrmunTrR3eJv0NVCya0cCgalWHSt9BcHgZAPJm6ZdW5aQRz3m9Pd2\n2OT1zm6k5V5EFNJmiDabjMYratspC1dnFNJWiDebDMerKbW9fjj6stHquYs5ckia52dbJCwk\nv7htZegmW/ODridiXEipnqLtkNpJ6gTrVLtjdtMshTRxNlI7R9skscmGXrTW0Xtk1+FIWyGN\n33udaKgIcWZYn+yGOn5TFNZHWwup7N4Qbu4RU0gEydj1OsI09ox7VHrVe+9/RCURHOO1ybCO\nQdbr2BkBTn9TSATHRG0yqd6osWDQWfO5CwWVJBnVeZPOdJhLGQBOqdhYopCCImiibKZnLpCZ\nKGfLFFvgFtYopJAIui2gBhuSmQ9yun7DM2c+vYF0l8Q5ioikoaMa2QrnzNfHYv8tjFuHSGzT\nGVkRUmfEIEhIanQznDM/J8tHj2gmUDfa0CxUSEIqjRzkdO3UxHYwX373Q+3TBByNGlmmkIIi\nZbJBzfwVxpmPj6ArdAKBFdL6+SGjRBWSd0Nss0ogWo1b8EQhbYH+aQmupNBDpFbKSFVuyROF\nFApJRTGIRVJwAyxv3kTJy6InsJBkn6GYSCqIkRmuFaIwhkKC2UseKZMMJ0aCERVfF8uFOXl2\n7USfoXjIKoSxaGRF2MEqtFwnG0SfoWgIK4PRcMLfdnH0IKzwTIELKdWCACKsBCbCCRWlbzdf\nWOmZQiHBkVYAaQlJWumZQiGBkTXNUM6cj0BL0zw9SCs+U/BCSrYoIMjL/HREQWL1FJK88jOE\nQoIiL+8zEYUI1nctkrwCNCSAkNItDG8E5jz8TcQ5k7YeBBagIRQSEIEZj3ATcc4iheRlJd3i\n8EHcNIMmxl3EOYOWHiQWoRkUEgqZmV6IChy076I+mWVoBCb0/2mad9X7m+98z+J9hiAtUspX\nFkeE5ngxLGjcvmuRhBaiERQSBKEZNggLGTmFBLeScpHYI3KaQRNXSL5rkaSWohEUkj9iM2sU\nGCz6KUMUkoeVpMvEErl5lSEkUw9yy9GEQEJKvFTyIG4F9l6LlHaVoZDcETs0OmMaXqg6YOdA\neGEuEawQEy8XA5TwTMYdm3ivRRJdlsuYh6/az7lctpJ4uSyjri9CsTi1ob1RSIZJxz5KvGAW\nkS4km9D8s+G9FElwSRrhV9wzjVTyJbOI8K6dKCFFXmKxAuEGmoJLBjRLIHqyIe5qUW+dSC5K\nIwLO2IgtG+FNCQZhQoq8DD0+thkYT5+UkKQPbiBgzivw6Pkk6Z+OgEKSWjoIIQnNWoN1gD45\n8r9PJL48F9mgkABdO6k5u2IfYHAh+S99EE3Qu9pSy8f38T1S89XgEKF7pvyXIskv0EU2KSQ/\nUsiVS4zO+fJeipRCiS6xaSE5xSc9U5qoGfNfipRCkS4RVEgCS6gbkUMnTV6WRnAL0jFr3kuR\nkijSJTYmpEFAcaeJIxFcEa5HpTUpZUVYIUkro5FwrBqlBKYZNME7aa4HUUiuVkSV0YQMvLsm\n0ogapv8KikRKdYHAQpJUStNq95+/FQWFtAKJCcmjbzV3pP+teUEI1tFY+lSKdYG0hOSxJsF7\nrVcyJ9wrUNsrlf9SpGTKdZ7QQoKWk3I3uHTUUgVKZJqh9NWR5fH+tw+SKdcFNiIk3xYnodPt\nE6ptAQMWUCRUsrMEFxLEw+U7uK5dO7MxUBJTJkv4NkiRhZRQyc4jXkjdb7I79bCMD0rmvvIM\nfrHaXakAveyUinaW8EKyWIvVb/Unfg1iJCnE/5TZlE62b6x296cdfXhbEIggIbWvhjMauqYw\nN2vByPRsUicbFayJHcAEalJlO4scITX98yUNXQ4Icvdn0CxaHr8uuGgNStfZmRpspE8EIRn6\ncJiSC3OyU+7DQ+dIl+4IuFv2NyEOOUJympILeLJTPMngmOfLwMNZimW7QAwhLX6suu92rsPM\nGLpOtK8LPOQ8bgnEYHUhGQ6IZi3YfmBg8/qSEiECni7dAM4SJoqQpm/QYDpQ+FlrCqllNfl7\nAjFYUUggFZ1tmfk0N8iq0mD4ABJ/owkTR0hjhwQeGEe9Mxma1YMZBOCvo8yuVOsICdkYtYzi\nbcpAQqUDl26afecZVhBSEBW1zYe0vgZGlS58tqEeKCRHK9ezEPqES7h6gzGpdFGyDVZSVucp\nkpCiFVt2VzpNU3pTzx+Nme0tPF3KHovceDydL955zlJIo5OSHUlFzHZuTQkI8zJRgw1zKzzP\noWGTvzJ+QlKEbIaAQjK3YtQjNkgTLUmOniDBKNSi+7SyvZQklpA2WLTyPGGCMbkmZphtmJB8\nJhsMHeVVtPI8iQoms2xHmv7eYtHK8yQqmMyyTSFtyJOoYDLLNoW0IU+igsks2xTShjyJCiaz\nbFNIG/IkKpjMso0REiEbh0IiBACFRAgACokQABQSIQAoJEIAUEiEAKCQCAFAIRECgEIiBACF\nRAgACikbiqJYO4QNQyFlA4W0JhQSIQAoJEIAUEhyqPpmn3fFvtp63hW752O98/2xKA5/6s1m\n72e187y32dRdu8+ieNYpK0tdMyQwFJIcKiXsiuKxLKtXva0l8FxvFloozd63004tltZmPUaq\npFM2r1czJDQUkhyqWl81R8fyRevitSheqvao2nWs1FTtb+2tFPKlG5+7zmYtpGf991fdLrUO\nIMGhkORwaljKcl9Pv9Wqeqwbo+Pjn87eKuFrc0xrs9SaetPN1GfnABIcCkkOl/nr4kJnSru1\nVw+Liv273tvaPKW9q/qGj3XPrnUACQ6FJAdjIZWvd0VxbmqazdNHVYeuPM04UEgxoZDk0BJS\nf1fZv9/69XKal2htnhJU46NXPU7iDdq4UEhyuNT8/UkImsd6tHMZI321U381Qvlqt166hbrr\nmSHBoZDkcFFG1Ts76FmDx2bW7q6z907L60+tl9bm+fCqPTpN1LUOIMGhkORwEdLx7nQDSLcn\n5/tIn529f4rr3tbm+fBjcW6J2mZIaCgkOTR9tZdKA4eTAN7315UNzd6vw+6yt9lseoZ3QzMk\nMBQSIQAoJEIAUEiEAOgL6e/Pp3v9fMv7p59/VwmIkBTpCunnbftx6Lc/VwqKkNRoC+nHTdUS\nff+oW6LfH98rUd38WCksQtKiEdLHTaWbf+3P/n6vdn1ED4mQ9GiENNr8VI1UxGAISZVGJ0//\nxj7/9xQtFELSJdozZAnJGQqJEABDCXx8u1XV3offXlYI2RQDCTzUt5CqD9QvDyuEbIu+BO7V\n/UephfRLKfM2iUIiG6cngR/qXu9U9ab5hB2FRDZOTwL3dTNUC+mfunG1QsjW6EmgllDn1cUK\nIVuDQiIEwKBrp9fW1RL6qIdLTlYI2RrTkw33ynzpN4VENs5w+vu2nv7+qDbcrRCyMQYSuL98\nre/e4huy/2n4zvfM32cYtiW/nvQ3+h4s1jWwRSKbh4tWCQFAIRECgEIiBMDghmwbVyuEbA0K\niRAAUxL4++2G30cixJRpCXxX5r/ERSGRjTMtgX9ca0eIKTMS4BiJOGJRdXLxNP3pbwqJuKGi\n1Qg5niY//LjlV82JE+r6siFPM9PfN+arVikk0iCoesfzNCmkm28Wq78pJNJCTocrnicuESJ4\nxEwBxPNEIRECgEIiBACFRAiA1oPGhvSSnveMqIZCIjljUL/NhXQVEYVERKBGtgI78kvTTjk+\nC0ghkQbVeg3uZrgd1JFfonZKRSGRWVTnLbSb0b8COvJL1U7Z7vHZfwWQZI/qvQd2E9qZoeWZ\nr1F8PIwnZYtEZlCDjbBuQntzFNLv+6nJBoAzsgHiTAEMbQfyZmq2l+5vS0a3Y9+QHbdLIZEz\nanQzoJew7oyN9hJ+U/f/yptq58f9+DNkRw1zjEQuRJlLG6+FsRyZpLzVP9TwUD+2716NLf8e\nvxRQSeSEmvwjmJeA7iwsDlYvlPrZLvqBLh/qm7GJ0TlxskGiCGnKLtqfjb0xIX2oer7O9Bmy\nFBK50K9OMZwE82dlrZdYD48uj2E27q+xa0fOxBDSnFGkQztbvdRP9Y/Z3ah/Wk7mMxbUEdEM\n6kGAijFrEujP0lQv+S91+1vP3T3oyQb++AmxY3Uh4RzaGuqnf9Cty7+b+kYSf/yEWBFjzcuS\nQZRDXyGVP/SjY/89KfXAHz8hdkQQ0rI9jEdrK70D/kVySzIkwnIDA3MQj/ZG+tMs3367+qWY\ntk749QZGxgAeHUwMpr+V3WOYW1aopG0TfrmB9zcasH5mj/l1q9TtTycrFNK2mTj/K8xIe7p0\nOnx40K8HpW6+2w2WlHsAJBPCr9uJ9LUet4PHjvr9TSm7wZLyCYFkweTJX2FC2sel47Hjh/37\nZjdYopA2T/AFcD4r3yIcOXncN5dvyFJJm2Xm1K8xIe3sEyukv7pFurVukaikzRJ8Jan3SoOg\nh02MkR6q1ujBYYxEIW2W0ELyvkMa8KCJI3/d64cjucza+UVCEib0SlL/O6TBjpk49Ke+j/TD\nwwqVtEUCCylS6wKc7KuGRvcOKxv+p+E738O8K7fjFDyeGfpr7Z6c19pNWCT5s3jK11hq4O/W\nxxdi9TeVtDUCf7chwlS2/3e8A3x7g0LaGIG/2wCcAZhJ51trQ3wNikraFrG+JOSAmV/EmusQ\nQqKSNkWsLwmFY4NC4u8VBcahgMMJKfjZviyDE9q1C6ckQIbJHA4FbJje7ZZqyEtyezGpyMkG\nnNlRq1RSOFwKOJiQUGd7IJIgj8YbWSL0pBesDr+ONPfYJAopDxwK2Di19XkDne1Ouxbu8ZJ9\nq3+vDxob/DzkTADDj4JFK0hHGQ7Y7As4nJAwZ7slx6CPaO2bvlW3v/TvFf+6VQ9LacuZZ8gG\nCllS3TU5z5LineN6Cm0DtrjnaWe4PgTRsXP07eSn4Ye6v2zejj9pzMTKxK68MDlDslrQCbx6\nOxZHrlQSkc5Bz8e9uj7v8sPzt7/l1yA/un2G0ZZZ/pjOd8hgc/BqN2WjOO6fejW+PZV+Zi86\nfDWxvRpjl7qupIQJqXdC50RkGrRV5uSURADCCQlcbv1KADXuxnIQorp27WAWWyKzAg40wWdt\nWQCDrt11YPTRDJem0zvsdWJ4XtMoaRGCP9HpipqmB6RxSi6n2EzpRfyhbi4PobhX5j+4GlpI\n0sZg6Z1nh37mclrbYgg4Vb46/ZB/qZsff0/T34bPYh6zMr/bmgmd5j16hWPdz1zMZyghpVi+\ngzHSAAcri/ut8B8SY0nxNNfYXwAWxlH2EQBTCUO+kOZsrFDkiTZHjsxmNpCQ0ixgTNQB2wzE\n5BKQrZ3lmQJ2MWpwTJolHFpI3vZNZpjjkWZzFOieK4XURraQTKpAxIJP8hx7ix86kyTrwgik\niftp9BeE/hmtE5rJvU/BGI5N7cfQTkEl2RxB1n0C77ejJwLF0ASubkZ+YPXHjff40KN/Dk94\nTe4QVIqnGKX9ETOultETgVJoIv+4qaTUaZX+fq92fQwOmbVi95mzUR8f1ncm7V3IANiEDi0F\nEVKChXyhHfoP/YPF3z/qpQ2/P77fqtFGasmK3YfTR9mJw1Z1to1YeqcYG/LownYnQ46fSacb\nu/4J/QbzhzLj2+uQLYZ91y69Mxz6RwlCdNjTK+UW/eD//nyqv2x+//Tz7+gBRlZsPo14yPXX\nl9R1h5Hl5M5wkIjVxLaHGbMPkiD09LeTkxCzahOLNAzWbiR3gkMJH2QXOp0uhihCimVt6sAF\ntcx+nFxzFPMXPoBGUivmPgKFBFal2XLBqWQyzq/Dc7EFM9oziB4FmFhCindh7/x+ppVb8G9v\n4jCZHVmlgB0NGO1KjEhCMpoog9SFxpOTOeCvQcMwma+PF6+/J9xNKUHEEZLRrRtIXXC72xrK\nDIZWMOv/XhHC0+BwKSXtQXQhTX7RCVMXchTS6CVmrd8rCiEkMQXtwYpdOxWkLoD6OJK6dia9\n1JS6dn0DggraHUGTDSgJgLKU2OlNaLKhd54TK+gJJE1/p1Z3iTNqdDNlxrPx+/uD7nHdPPww\ne8x5JoVBYqFGttJmNB8freGL99coCJkmm5ozmpEH9XR6ztjvp9GfWzWzQsh2GF8XY3uDn0Ii\nG4dCIuuQ2dTSaG7u1Td27UhQZN2o88dismHmp1ezKhISAVlLRwCM5+Xfj4eb+svmrYebq5lD\ncioREoONCGk2ZftmGiGbIaCQLKyYOMKYSS2YjWYb9Cu6UrJNIW3Ik6hgMsu2RTd1poWTkhtg\nkhw9iQoms2xP3Ecy7RmiwkCaSS0YZnv9JIGE9I1CytKTqGAyy/b4p9/M1qriwkCaSS0YZnv9\nJMHGSE83Bs6BYSDNpBYMs71+knCTDTcWzzSXkxtgkhw9iQoms2wbDYEIIfNQSIQAoJAIAUAh\nEQKgJ6TuKr2H72a/fULI1pkVklI3v9YJi5C06HftvlXaqZqhf79u1bfy33d1wzaJkGV6QvpQ\nt5fNe7264Ul9jx0RIQnSE9J9szboQ/9cw2+z32wgZOMMxki97cx+64WQMFBIhABY6NpRSFIo\nisI7xYWDXyhkhJ5OfjRjonv1Q6vpIXZEZBSckD53ZumIDf0G517dXqa/K0n9q9VEBGDe3ixc\nLrwAABWuSURBVMSzRBoGPbeH67IG/amy+14SCQaFJJvhEOjj261St6fHUaiHv9Ej2gqf+6LY\nf+qNoniu3p6LQv/1vCt2z8eyru+fd8X+tOtwPFf/61H68+Oh2L2Wf/bF7r08C0S/VOnvXnWS\nP4finLq1u6iJn9/M4VzCSjyfKvRLtVlJp/Wq2R3rCl/98ViW++K0WXSPOu8sHus9742Q6vTF\na63Qms/ObgopCBTSOvypavbxWFXuP7U6vsqvul160dX+tVaKrvtledR/7Y+1mrpHVTsO5Xv9\n+lYUdy0hHbXFasdd8Vbqz/bd3ZRRCIZC+qV/9vvmgatVg3KoG4pPLQP9+qYr/Gfd+pS6qp/q\nft2F219SFt2jCi2/5rV5+VN2xDLYTSGFoC+kv/eXyYZ7Do8CsjvV5qLuz91VPbjHequ40NT3\n4pKy6B5VnDV3fW0fdX47vh72g90UUghmpr9vRw8gEDoVvurQlacZhwUhDbdnhfTYN0YhBWPh\nhiwJRKdFqsZHr3UXbdAlK7sasGuRqiHR/u1IIUVhaYkQCUN7tKP7dvU0QD0g+jqnuNT3x+kx\nUjkvpP5nFFJAlhatkjB8tubfSj0zV09p607eQX/22NT3t0vKonvUspB2WnbPFFIUKKSVON8R\neq7/0P2vuiU63tV7d1+t+j5yH+k8nCrnhfRSnI/86uzWd2m5bBXNXNfuKXYwm+Jdr1F4P/+x\nP/XsKl4qKR26w6Vnvef8Z3PUspDKV72W4ahl09799Vi3eATK3GQDbyURYki/+3bbTH/zCxSE\nmDJzQ3aVeAhJkpElQk96idAT+3WEmMOZOUIAUEiEAKCQCAHQCKn/s9+Gz2EmhKCERMmRjYOR\nAIVENg6FRAgACokQABQSIQAoJEIAUEiEAKCQCAEwLgFbYVBIZONQSIQAoJAIAUAhEQKAQiIE\nAEZI/2n4zvfM32dgi0QIAAqJEAAUEiEAKCRCAFBIhACgkAgBQCERAoCrvwkBQCERAoBCIgQA\nhUQIAAqJEAAUEiEAKCRCAFBIhACgkAgBQCERAoBCIgSAhQQuT0waOYRCIhvH8mFiavwQCols\nHNun8ikKiZAh1o+3VBQSIQPsnxPbfrQsH9tMSI3NZMP0IRQS2Tic/iYEAOZL5RQS2TgUEiEA\nKCRCAHCMRAgACokQABQSIQAoJEIAUEiEAKCQCAFAIRECgEIiBACFRAgACokQABQSIQAoJEIA\nUEiEAKCQCAFAIRECgEIiBACFRAgACokQABQSIQAwEvifhu98z/x9BrZIhACgkAgBQCERAoBC\nIgQAhUQIAAqJEAAUEiEAKCRCAFg8sW/m8XwUEtk4Dk/s87NCSI7YSGCYls+QJaQGIQFVUklk\n4wAUoGop+dshJF0AT+yjkAhBPPqSXTuyeSDPkKWOyNbhDVlCAFBI24Edh4AghcTzJBrFMxQQ\naIvE8yQYdX0hAcB27Xie5JKgkFLqi4LHSAnlfHMk17UDBRxHjujJhrRO1bZI6QJfwppQIzn6\nFw181i6FsxUvxhRKQyoYIRlZAbR9Aaa/xdedeH0cMb0pGVFYAim9lpAmv1GHkGyI+0jCz1q8\nUbeU8b1K6vw0QNrzMTmqrqakCkn4mdqakIbVMbUOp2O8SrXfZlOK7NrBrIYieIfr2nuQ0LUb\nXx0ZOwo7hp0vWy1Zftv0ktj9EhNoiZDsEwW6Io+Y6XfBV7/2TwWwemAzjMdmIQ2fr2y7Hhtq\nrZ3M89QvJK/q1G1uRH7lfi4gabFemI3ZpJvmfRacTmQoIUk8T6O/f+TXljfTQa5mArIQVMCY\nPUp1caZ6QWigTNn3JTFug1mGMVkyrkUvYyZhBqMxdijPXtMDi6lG06GvZpajLIzPcKYxzJaJ\ne69YUhZ7GGUqTPyulxib8zCYHQjSvrYlu9S+YxwGtO3P8qXF7iz4T/EExzS0MNXv+mJ1lEso\nES5mhhOwIYVkWzZhKqZxd8HQmmD5XLG5tIdxb3utcSvVeN3rRU9BhWRnPsTVxabaLyddTURu\nLWaI1GbGWjaNrjzyx6lrC8nCfoBCsb+LN/vhak2R1SXGPkzcRNdkzyTEGLU+NNoAYtWunZUD\nlJC8hp8Tx6zbobMqGaeRhn/mFgtosgj9nItZyB9cSFZKunQMPKLyvUYNZxJWHxUZCcmzPvoc\n7DbAXLtUwYQXkkEVGI5MnZcJwNq1RtX+xnxpD96nvwmwzpXd9kRJWoaIJIKQ5o6bPQ0jny5P\nZPuEAzeDYnJVH/CLAOY9h871zsebnAL2J5aQxtbzmt0z7NwUm3LVpMJc6VI7z7B4DaYuW82j\nh5/rSyZEElKnetv32yauu6MdHUxfLLWeByjeZTPwvnMmxBFSq/S9lrh3zIQdvUgYG9kAiXdU\nJfhvk57NAozIwSI3E8NcEyu8jKVB/0oVYlojU8zLRA02LKygSj+zy5g4TM4Tz8EYfkJShGyG\ngEKysGI0QycoSY6eRAWTWbYppA15EhVMZtmOM9lg6CivopXnSVQwmWU7zvS3oaO8ilaeJ1HB\nZJZtCmlDnkQFk1m2KaQNeRIVTGbZppA25ElUMJllm0LakCdRwWSWbYyQCNk4FBIhACgkQgBQ\nSIQAoJAIAUAhEQKAQiIEAIVECAAKiRAAFBIhACiktSmKwvOoevu84+BukvhAIa0NUkifu8Ld\nJPGBQlobkJA8bBEAFNLaUEhZQCGtzbnyf+6LYv952vW8Kw7Hjij+HIrrx+1Pm239UtSctj+L\n4lknKIrPU7rd8zFWlrYIhbQ2J0k8n0Twovfs9dZdW0ifp09rTdSf7k6ftrb7Qior6ZTNa52Q\nSgoHhbQ2da3/U7U3x2Oliz9l+VZtl3q7uKa5K97O+8tXnbI8fdrebk/d1W+VNL/Kr7pdetES\nfD3LlASBQlqbutYf6tamangOZfl42S7GUu5bn7a3B0Kqdr9p8X3W6eq9+2iZ2h4U0tqc+2HF\naXvXVcOls1YeXw/7gVYG2+0dVTP2WIlyV16scCYiJBTS2izJo35/vCrBXEgv9ajpuaSQYkAh\nrY1Ji1SNd/Zvx4FW5oVUjY9e9TiJs+IxoJDWxmSMVJwHOUX308e5MZLu2xXFnd7Yn/REAkIh\nrU1xnhmYm7XbacU813veLil72yNCqtqj00TdixZo5eIxasa2BYW0NkX7PpIe0ZzuDnWE9FJ/\nuKsblpn7SLph023a+VDdF6xbouPd6Xg2S+GgkNbmXOvf9cqG99OuZ62ozsDmdVfcvR5rlehP\nD1/nT5vt+uXrsW52LofuTz27ipc7nTBCbjYLhSSV4qoCkgAUkjiKeqT0eu7nkTSgkMTxcr7p\nw6VxKUEhyeNN336942LtpKCQCAFAIRECgEIiBACFRAgACokQABQSIQAwQqIcycaBSEBRSGTj\nICSgqCSydQAKULWU/O0Qki4UEiEA2LUjBAAnGwgBwOlvQgBQSIQAoJAIAUAhEQKAQiJ4Njj7\nBBPS9oqOTLHF6oDJ8H8avvO9fldC4sC/z4Dr2m3uGkQmUNeXDUEhETgRu3bxRmMLnigkgida\n9Y4n2SVPwFk7KolEJl69W/REIZF0oZAIQYDq2k30RVWNkSfkDVkqiUTiUr1Bo7GuStSVrssl\nE5hAYLYImWNQwy+7PUw2L+PGTW14ozpvZNuEu6zOVnNXCWBGJhQSQQMTkur3t5YPcBMTYKhF\nIREwqvfuY+g6EjI35pTWf6iFXf1NJRFYZXDvcJloyX00NGEPaoVCImpky92Q87BnRiZoDZ2M\nQq1QSAQmJO+Ry+UOkO1Iy80Z1gqVtHlwQsLcJHIbabm4QVqhkLaOmthei1iLbigkAoVCWk56\nbhrHbiu72CM5oib/WAnUcrxlNzYp1fghFBI5oWb+Wok4X46yFNKEvimkFIhRo4BCSqsq2Qqp\nHH3whBrdJKKI0cfpO/BxmFZNshbScFq+OzWfVvY3RJRRN1BIiVUkm8mG6UOkTdWIQdAvJa4i\nJA+HckrOCPD0N8xiHkSaMTIjQjBD+84e5ZSbGbbxjqenkEaJdQ/DkPDNI4XkmV7czQMZCBNS\neBZ6/X6WZEMhhURU1y48Y3mlkJytJFcGARE02RCB0cw6lUB6xRZASAmWwqYIdnoMeit+piRD\nIYVCWimEX30yYdfBnbSyM4BCCoO8Pl34m+YUEtpKguUARl4JhB/GThm1dyav9JahkAIgrzla\nU0j23gQW3yIUEh6J2Q9/hqZNUkjOVlIsCRQSm6MYZ2jGoqUzkQW4BIUERmbWgWt3zF24+pJZ\nhAuEEVKaZQFAZnOEXLtj4cHRldAiXIBCQiI127glB3YunFxJLcR5KCQgYnMNXHNg58LBk9hC\nnCeQkFItDh+kduuQd0ptXTi4EluK81BIKATnGHer1NqFtSfBxTgLhYRBbnMEvcPjaohCcraS\naoG4ITq3uIlpdzumjkQX5BwUEgDJzRFyPs3DDIXkaiXZErFHeFbnw4MEb2DEzI/wopwhmJAS\nLhNTZDdEF3DDFy8bFJKrlXTLxBCVRh5hvS4/CxSSCf/T8D3Nd+VtR4nIR4T3GcK1SAlfXYxQ\n1xfRwHpdvgZwzZZIKCRnkujaweq4/+HLyeQX5zQBhZR0uZiQwmRDjKEJhVRSSJkT5fYNSkhJ\n1xcKKWuMT4HHuYL5SLq+hBRS2iWTA1EWXaOElHZtoZCcSCRnMYQE85FImU5AIbmQSMaswnTN\nE0pIiZTpFBbr20eec7lkJfHCmSKVbNnF6ZgrCqkGUwzbElIqubKN0ylfsFYvlVKdwCb8YdqZ\nRsraekT8bgAJzdSQKIGihJRMqU4QdIyELx7IPVC/JQnJnPE4gaK6j8kU6wRpCQmyKsdvkVw6\nJ9ytp2Z5FCp5OsU6AaYfHal8MOtEvaykc8JdRzxBZygopPn0kVpsjJB82rWEzrdLqNYFbO0k\n2xslSQkJ07XzGGkldL7dp+AoJBcCj5HgSlq1xFM63W6x2l6p7L2MHpFSwU6QmJDWJIVvTVxx\nbnNDr4agkBytgMooSC2OsoRmHSTOfU8ek1bRjhNcSOsOakZMOdpN62TL1RGF5GoFMMs2MOFu\ns3+kqZQSO9e+4ZodDxJSYmU7jnghja8/Ak67GZlK7FyD7lqHcQO8LAoivJC8XGBX8Y1bM2iU\nUjvVgHjDlQqF5GjFox82N4kBmnk1sZXamY5wUn3cqNk/E0WukKbXlDvade5/JnemUTOlgdxQ\nSI5WnCZJDQ6ys7sQY6zFGRGABTx/EpzdqJm/kkWmkAz7bdDFLFM+0zvRwIjDzCKpyT/SRZKQ\nLlXZfPiDvQ0EnR9cD2jEQRpqNbGdMlHGpcbti3U0xhp1TZbgecaGPD1DgzGaYAGPIkdIyjil\nvWnzJi79gTA65BBdXjWylThxZkqNOlamCXuHge93JLU0dQx8/ON33zAWUy/tK3KE5NS1gzn3\nSC+LENHjx46q954+goTk3BjgJ9/LdFumIGGPGKWQuki4Ce5dZwNM0ro2j2sTKOjBKcKcsiSL\neJy1hTTzo3gQ/z6NXIqnOVjM4FkYCglpBSKi2QBc7acqpIAh53gXFch6QsKpaDICDw+Jdu1C\nxtwuzRTLJizrCAnToeuZHO4C25NP4JhTLJJYrCAkvIjGg9jgaQ+d5Q0WqSmxhHQVT8ArvZr5\ni0gjySZ/mnizdhHKTU1sEyQYBSQ6CJ0k4g3ZrMptu2AUkF2NyE9ImXUZTImVbdCppJDcrMRr\nyHPrMtQsqyRatlEKyO08WWRm5ul8FpMNgcnuSqdZrnQRs41SQGY9B/PcqMGGi5XQ5Cik0Tx1\nn43NvvPa+AlJEbIZAgrJwoqJI4gZ1O8PYcxAPCmTQVJ+2U4qmPyElFww8S4OqWU7pWBiTTZs\nsGjleRIVTGbZjjT9vcWiledJVDCZZZtC2pAnUcFklm0KaUOeRAWTWbYppA15EhVMZtmmkDbk\nSVQwmWUbIyRCNg6FRAgACokQABQSIQCiCOmyHmLGmWqnmzZjkGRxeaEBJgEvB3NOuPjhvBmj\noTKmZAxWZwKzvWQAkieDM4nJUwwhXXOyUF0WcmyRxDNXBgGbBGOwbNjMzFysplbK5ZJRgw0n\nTybZjpUnVNVbzhPk4m3kQ0US0pICDKv3QsAGwZTL10LDPJk8ltq/ZEBCMsx2nDyBqp5Bnvxb\nJOMWUYkQksUdg7mATUr/VP4AIS1VO1lCMsx2lDyBql65nCd/IS2buCRYauzjCMlc+UaPOl+6\nsC524Y3yNG8orpAWkpw+Nnogevg8GVW9wcZEuvnx2vzBGFTvfSLVoujNhp+LjpYxCFgtpli2\ncfp8IdsmeQGVjMlkg2lQJkKaNwDJk0nVg+QpipAIyR1/IZnNmRrNU/qbMfFEpjCYK8YkiefJ\nZGrbrO0D3KWYBdSxxpgx7e6SMUzHHN5J4nkyGGmZjsbMhrLupCYkw9sci1P6kCTxPBkFc3ld\nKmDfJPE8UUgRPM1gkAaTJJ4n8yRLs/7+SeJ5opDCeZJVvyUp6ZLCYNbfL0k8TytUPWc42ZAN\nqvceLEk8T2oxhVGa5SSscoQA2F6LJCoYZnv9JBgzmxsjiQqG2V4/iZQxkszcJBIMs71+Egop\ng2CY7fWTUEgZBMNsr59EipBEjfiSC4bZXj+JkMkGQghbpM14EhVMhtnmGGkbnkQFk3W2XZGZ\nm0SCYbbXT0IhZRAMs71+Egopg2CY7fWTSBGSqBFfcsEw2+snETLZQAjZopBEXTAzvDTL8sQW\nKRSiuvBZDxYkeEpojKTUoloNkkT8FZDBhlMamScxkWCyzrYzBhZMnGDMLCcRdZ6zrlESPKUk\npMSUJOo8Z12jJHhKSkiJIWosnOGoW5anaMFsT0iEBAAipEXNGyUBmTHxRAgYRJVTy5YMkoDM\nmHgiBM3mhGQwQx5vLj7irL+kYLLM9rQDc9SyIYMkIDNLSQTNIMb0JCqYDLMNaZFEXRkWk4g6\n0RlWKVmeYgXDwQQhACgkQgDgpr8XZhKWk4DMmCaZjyXiXHw0T6KCyS3buFk73ySxu80SphCj\nehIVTHbZppCc08g5iQkGk122tygkk0l0kzSYJPE8iQomt2xDhLQ8LDFIAjJjnkTEXHxET6KC\nyS7bCCERsnkoJEIAQISkdMvnnQRkxiTJQktvmAaTJPKsv5Rgcss2aLJBLc95LCUBmTFLskjc\nmQ8pSXL0FCkYCsk5jZyTCEuSoycKKWSSRZI6ibAkOXpKSEj1mGSxE2oQC8TMcpLF2VCzNJgk\n0Wf9ZQSTXbYhQiJk61BIhADYpJBUKWguPuKsv6Rgcsv2FoUkauYj6hyLnGCyyzaF5JxGzklM\nMJjssk0hOaeRcxITDCa7bG9RSKLm4iN6EhVMbtnepJAIQUMhEQKAQiIEAIVECAAKiRAAFBIh\nACgkQgBQSIQAoJAIAUAhEQKAQiIEAIVECAAKiRAAFBIhACgkQgBQSIQAoJAIAfD/cRgkCz3y\n3nsAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Use the `WGstats` function to get the detailed result for every SNP (similar to the previous recipe)\n",
    "WGstats(myres)\n",
    "summary(myres)    # View the overall summary\n",
    "\n",
    "\n",
    "plot(myres)  # Plot the p-values for all of the models and SNPs \n",
    "# the horizontal dashed lines (red and pink) represent two different thresholds, namely, \n",
    "# the Bonferroni correction and nominal p-value (0.05), respectively:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The plot shows the p-values for every SNP as an absolute log scale along the y axis and SNPs along the x axis. The facets represent the genetic model. The p-values presented here are equipped with the Bonferroni correction. While plotting a warning, it tells you how many SNPs are significant for every model of inheritance after the Bonferroni correction. Here, in our results, we have no significant association for the overdominant model. The plot shows the plot for the association of all SNPs in the dataset. The plot shows the log p-values to the base 10 for different models along the y axis, and the SNPs are aligned along the x axis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A WGassociation: 35 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>comments</th><th scope=col>log-additive</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>snp10001</th><td>NA         </td><td>0.001403687</td></tr>\n",
       "\t<tr><th scope=row>snp10002</th><td>NA         </td><td>0.768461800</td></tr>\n",
       "\t<tr><th scope=row>snp10003</th><td>Monomorphic</td><td>         NA</td></tr>\n",
       "\t<tr><th scope=row>snp10004</th><td>Monomorphic</td><td>         NA</td></tr>\n",
       "\t<tr><th scope=row>snp10005</th><td>NA         </td><td>0.372673000</td></tr>\n",
       "\t<tr><th scope=row>snp10006</th><td>Monomorphic</td><td>         NA</td></tr>\n",
       "\t<tr><th scope=row>snp10007</th><td>Monomorphic</td><td>         NA</td></tr>\n",
       "\t<tr><th scope=row>snp10008</th><td>NA         </td><td>0.134925200</td></tr>\n",
       "\t<tr><th scope=row>snp10009</th><td>NA         </td><td>0.936159000</td></tr>\n",
       "\t<tr><th scope=row>snp100010</th><td>Monomorphic</td><td>         NA</td></tr>\n",
       "\t<tr><th scope=row>snp100011</th><td>NA         </td><td>0.127867500</td></tr>\n",
       "\t<tr><th scope=row>snp100012</th><td>NA         </td><td>0.489937800</td></tr>\n",
       "\t<tr><th scope=row>snp100013</th><td>NA         </td><td>0.054748610</td></tr>\n",
       "\t<tr><th scope=row>snp100014</th><td>NA         </td><td>0.012451810</td></tr>\n",
       "\t<tr><th scope=row>snp100015</th><td>NA         </td><td>0.026258050</td></tr>\n",
       "\t<tr><th scope=row>snp100016</th><td>Monomorphic</td><td>         NA</td></tr>\n",
       "\t<tr><th scope=row>snp100017</th><td>NA         </td><td>0.999169600</td></tr>\n",
       "\t<tr><th scope=row>snp100018</th><td>NA         </td><td>0.921291100</td></tr>\n",
       "\t<tr><th scope=row>snp100019</th><td>NA         </td><td>0.010243840</td></tr>\n",
       "\t<tr><th scope=row>snp100020</th><td>NA         </td><td>0.141347000</td></tr>\n",
       "\t<tr><th scope=row>snp100021</th><td>Monomorphic</td><td>         NA</td></tr>\n",
       "\t<tr><th scope=row>snp100022</th><td>Monomorphic</td><td>         NA</td></tr>\n",
       "\t<tr><th scope=row>snp100023</th><td>NA         </td><td>0.826949900</td></tr>\n",
       "\t<tr><th scope=row>snp100024</th><td>NA         </td><td>0.002632158</td></tr>\n",
       "\t<tr><th scope=row>snp100025</th><td>Monomorphic</td><td>         NA</td></tr>\n",
       "\t<tr><th scope=row>snp100026</th><td>Monomorphic</td><td>         NA</td></tr>\n",
       "\t<tr><th scope=row>snp100027</th><td>NA         </td><td>0.758636700</td></tr>\n",
       "\t<tr><th scope=row>snp100028</th><td>NA         </td><td>0.005451930</td></tr>\n",
       "\t<tr><th scope=row>snp100029</th><td>NA         </td><td>0.003330572</td></tr>\n",
       "\t<tr><th scope=row>snp100030</th><td>Monomorphic</td><td>         NA</td></tr>\n",
       "\t<tr><th scope=row>snp100031</th><td>Genot 65\\%</td><td>         NA</td></tr>\n",
       "\t<tr><th scope=row>snp100032</th><td>NA         </td><td>0.007170055</td></tr>\n",
       "\t<tr><th scope=row>snp100033</th><td>NA         </td><td>0.004281721</td></tr>\n",
       "\t<tr><th scope=row>snp100034</th><td>NA         </td><td>0.003330572</td></tr>\n",
       "\t<tr><th scope=row>snp100035</th><td>Monomorphic</td><td>         NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A WGassociation: 35 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & comments & log-additive\\\\\n",
       "  & <fct> & <dbl>\\\\\n",
       "\\hline\n",
       "\tsnp10001 & NA          & 0.001403687\\\\\n",
       "\tsnp10002 & NA          & 0.768461800\\\\\n",
       "\tsnp10003 & Monomorphic &          NA\\\\\n",
       "\tsnp10004 & Monomorphic &          NA\\\\\n",
       "\tsnp10005 & NA          & 0.372673000\\\\\n",
       "\tsnp10006 & Monomorphic &          NA\\\\\n",
       "\tsnp10007 & Monomorphic &          NA\\\\\n",
       "\tsnp10008 & NA          & 0.134925200\\\\\n",
       "\tsnp10009 & NA          & 0.936159000\\\\\n",
       "\tsnp100010 & Monomorphic &          NA\\\\\n",
       "\tsnp100011 & NA          & 0.127867500\\\\\n",
       "\tsnp100012 & NA          & 0.489937800\\\\\n",
       "\tsnp100013 & NA          & 0.054748610\\\\\n",
       "\tsnp100014 & NA          & 0.012451810\\\\\n",
       "\tsnp100015 & NA          & 0.026258050\\\\\n",
       "\tsnp100016 & Monomorphic &          NA\\\\\n",
       "\tsnp100017 & NA          & 0.999169600\\\\\n",
       "\tsnp100018 & NA          & 0.921291100\\\\\n",
       "\tsnp100019 & NA          & 0.010243840\\\\\n",
       "\tsnp100020 & NA          & 0.141347000\\\\\n",
       "\tsnp100021 & Monomorphic &          NA\\\\\n",
       "\tsnp100022 & Monomorphic &          NA\\\\\n",
       "\tsnp100023 & NA          & 0.826949900\\\\\n",
       "\tsnp100024 & NA          & 0.002632158\\\\\n",
       "\tsnp100025 & Monomorphic &          NA\\\\\n",
       "\tsnp100026 & Monomorphic &          NA\\\\\n",
       "\tsnp100027 & NA          & 0.758636700\\\\\n",
       "\tsnp100028 & NA          & 0.005451930\\\\\n",
       "\tsnp100029 & NA          & 0.003330572\\\\\n",
       "\tsnp100030 & Monomorphic &          NA\\\\\n",
       "\tsnp100031 & Genot 65\\textbackslash{}\\% &          NA\\\\\n",
       "\tsnp100032 & NA          & 0.007170055\\\\\n",
       "\tsnp100033 & NA          & 0.004281721\\\\\n",
       "\tsnp100034 & NA          & 0.003330572\\\\\n",
       "\tsnp100035 & Monomorphic &          NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A WGassociation: 35 × 2\n",
       "\n",
       "| <!--/--> | comments &lt;fct&gt; | log-additive &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| snp10001 | NA          | 0.001403687 |\n",
       "| snp10002 | NA          | 0.768461800 |\n",
       "| snp10003 | Monomorphic |          NA |\n",
       "| snp10004 | Monomorphic |          NA |\n",
       "| snp10005 | NA          | 0.372673000 |\n",
       "| snp10006 | Monomorphic |          NA |\n",
       "| snp10007 | Monomorphic |          NA |\n",
       "| snp10008 | NA          | 0.134925200 |\n",
       "| snp10009 | NA          | 0.936159000 |\n",
       "| snp100010 | Monomorphic |          NA |\n",
       "| snp100011 | NA          | 0.127867500 |\n",
       "| snp100012 | NA          | 0.489937800 |\n",
       "| snp100013 | NA          | 0.054748610 |\n",
       "| snp100014 | NA          | 0.012451810 |\n",
       "| snp100015 | NA          | 0.026258050 |\n",
       "| snp100016 | Monomorphic |          NA |\n",
       "| snp100017 | NA          | 0.999169600 |\n",
       "| snp100018 | NA          | 0.921291100 |\n",
       "| snp100019 | NA          | 0.010243840 |\n",
       "| snp100020 | NA          | 0.141347000 |\n",
       "| snp100021 | Monomorphic |          NA |\n",
       "| snp100022 | Monomorphic |          NA |\n",
       "| snp100023 | NA          | 0.826949900 |\n",
       "| snp100024 | NA          | 0.002632158 |\n",
       "| snp100025 | Monomorphic |          NA |\n",
       "| snp100026 | Monomorphic |          NA |\n",
       "| snp100027 | NA          | 0.758636700 |\n",
       "| snp100028 | NA          | 0.005451930 |\n",
       "| snp100029 | NA          | 0.003330572 |\n",
       "| snp100030 | Monomorphic |          NA |\n",
       "| snp100031 | Genot 65\\% |          NA |\n",
       "| snp100032 | NA          | 0.007170055 |\n",
       "| snp100033 | NA          | 0.004281721 |\n",
       "| snp100034 | NA          | 0.003330572 |\n",
       "| snp100035 | Monomorphic |          NA |\n",
       "\n"
      ],
      "text/plain": [
       "          comments    log-additive\n",
       "snp10001  NA          0.001403687 \n",
       "snp10002  NA          0.768461800 \n",
       "snp10003  Monomorphic          NA \n",
       "snp10004  Monomorphic          NA \n",
       "snp10005  NA          0.372673000 \n",
       "snp10006  Monomorphic          NA \n",
       "snp10007  Monomorphic          NA \n",
       "snp10008  NA          0.134925200 \n",
       "snp10009  NA          0.936159000 \n",
       "snp100010 Monomorphic          NA \n",
       "snp100011 NA          0.127867500 \n",
       "snp100012 NA          0.489937800 \n",
       "snp100013 NA          0.054748610 \n",
       "snp100014 NA          0.012451810 \n",
       "snp100015 NA          0.026258050 \n",
       "snp100016 Monomorphic          NA \n",
       "snp100017 NA          0.999169600 \n",
       "snp100018 NA          0.921291100 \n",
       "snp100019 NA          0.010243840 \n",
       "snp100020 NA          0.141347000 \n",
       "snp100021 Monomorphic          NA \n",
       "snp100022 Monomorphic          NA \n",
       "snp100023 NA          0.826949900 \n",
       "snp100024 NA          0.002632158 \n",
       "snp100025 Monomorphic          NA \n",
       "snp100026 Monomorphic          NA \n",
       "snp100027 NA          0.758636700 \n",
       "snp100028 NA          0.005451930 \n",
       "snp100029 NA          0.003330572 \n",
       "snp100030 Monomorphic          NA \n",
       "snp100031 Genot 65\\\\%          NA \n",
       "snp100032 NA          0.007170055 \n",
       "snp100033 NA          0.004281721 \n",
       "snp100034 NA          0.003330572 \n",
       "snp100035 Monomorphic          NA "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## The `WGassociation` function performs a whole genome association analysis \n",
    "## by computing just the p-values for each SNP corresponding to the likelihood ratio test\n",
    "resHapMap <- WGassociation(protein, data=mySNP, model=\"log\")\n",
    "resHapMap"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.  The whole genome SNP association analysis\n",
    "\n",
    "We have worked with SNP data where we had a few SNPs and made attempts to understand the phenotype dependence on these SNPs. And then we attempt to study the association analysis of a trait across the whole genome.\n",
    "\n",
    "Use a much larger one called the `HapMap` data set, available within the `SNPassoc` package of R, to perform a whole genome association analysis for SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'data.frame':\t120 obs. of  9307 variables:\n",
      " $ id        : Factor w/ 120 levels \"NA06985\",\"NA06993\",..: 1 2 3 4 5 6 7 8 9 10 ...\n",
      " $ group     : Factor w/ 2 levels \"CEU\",\"YRI\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs10399749: Factor w/ 1 level \"CC\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs11260616: Factor w/ 3 levels \"AA\",\"AT\",\"TT\": 1 2 1 2 1 1 1 1 2 2 ...\n",
      " $ rs4648633 : Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 3 2 3 3 2 3 3 2 2 NA ...\n",
      " $ rs6659552 : Factor w/ 3 levels \"CC\",\"CG\",\"GG\": 3 2 2 3 3 2 2 2 2 3 ...\n",
      " $ rs7550396 : Factor w/ 2 levels \"AG\",\"GG\": 2 2 2 2 2 2 2 2 2 2 ...\n",
      " $ rs12239794: Factor w/ 1 level \"GG\": 1 1 1 NA 1 1 1 1 1 1 ...\n",
      " $ rs6688969 : Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 1 2 2 1 2 1 2 2 1 1 ...\n",
      " $ rs10753357: Factor w/ 3 levels \"AA\",\"AC\",\"CC\": 2 1 1 2 3 3 3 3 3 2 ...\n",
      " $ rs1495243 : Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 3 2 1 3 3 2 2 2 1 2 ...\n",
      " $ rs12047125: Factor w/ 1 level \"GG\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs780983  : Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 1 2 1 1 2 2 2 1 1 1 ...\n",
      " $ rs6681520 : Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 2 2 2 1 2 2 3 2 2 3 ...\n",
      " $ rs12410393: Factor w/ 1 level \"AA\": 1 1 1 1 1 NA 1 1 1 1 ...\n",
      " $ rs12136845: Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 1 1 2 3 1 2 3 2 1 1 ...\n",
      " $ rs7419115 : Factor w/ 2 levels \"AG\",\"GG\": 2 2 2 2 2 2 2 2 2 2 ...\n",
      " $ rs6665656 : Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs1695623 : Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 1 1 1 2 1 1 2 1 1 1 ...\n",
      " $ rs12745075: Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 1 1 3 2 1 3 2 1 2 1 ...\n",
      " $ rs11121187: Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 2 1 2 1 1 1 2 2 2 1 ...\n",
      " $ rs17365584: Factor w/ 2 levels \"CC\",\"CT\": 1 1 2 1 1 1 1 1 1 1 ...\n",
      " $ rs6577401 : Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 3 1 2 1 1 3 1 1 1 2 ...\n",
      " $ rs2067995 : Factor w/ 3 levels \"CC\",\"CG\",\"GG\": 3 3 3 3 3 3 3 3 3 3 ...\n",
      " $ rs12161399: Factor w/ 1 level \"CC\": 1 1 1 1 1 1 1 NA 1 1 ...\n",
      " $ rs12093433: Factor w/ 2 levels \"AA\",\"AG\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs17027135: Factor w/ 2 levels \"AG\",\"GG\": 2 2 2 2 2 2 2 2 2 2 ...\n",
      " $ rs7532257 : Factor w/ 2 levels \"CC\",\"CT\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs6577514 : Factor w/ 3 levels \"GG\",\"GT\",\"TT\": 1 1 1 1 2 2 2 1 1 2 ...\n",
      " $ rs10746482: Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 1 1 2 3 1 3 1 3 3 2 ...\n",
      " $ rs2311019 : Factor w/ 1 level \"TT\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs1220388 : Factor w/ 1 level \"GG\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs544289  : Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 1 1 NA 1 2 1 1 1 1 1 ...\n",
      " $ rs284277  : Factor w/ 3 levels \"GG\",\"GT\",\"TT\": 1 3 1 2 1 3 3 3 2 2 ...\n",
      " $ rs12119523: Factor w/ 2 levels \"CT\",\"TT\": 2 2 2 2 2 2 2 2 2 2 ...\n",
      " $ rs3124625 : Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 2 2 2 3 3 1 2 2 3 2 ...\n",
      " $ rs4846056 : Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 2 2 3 1 1 2 2 1 1 1 ...\n",
      " $ rs4846094 : Factor w/ 2 levels \"AG\",\"GG\": 2 1 1 1 2 2 2 2 2 2 ...\n",
      " $ rs12743366: Factor w/ 2 levels \"CC\",\"CT\": 2 1 2 1 1 2 1 1 1 1 ...\n",
      " $ rs3795746 : Factor w/ 3 levels \"AA\",\"AT\",\"TT\": 1 1 1 2 1 1 1 2 1 1 ...\n",
      " $ rs12066711: Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs7517964 : Factor w/ 2 levels \"CC\",\"CT\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs12745913: Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 1 3 3 2 2 1 1 3 1 3 ...\n",
      " $ rs16853611: Factor w/ 2 levels \"GT\",\"TT\": 2 2 2 2 2 2 2 2 2 2 ...\n",
      " $ rs1416623 : Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 1 1 1 1 1 1 1 2 1 1 ...\n",
      " $ rs4233528 : Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 3 3 3 3 3 3 3 3 3 NA ...\n",
      " $ rs2807564 : Factor w/ 1 level \"AA\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs6681992 : Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 3 3 3 3 2 3 3 2 3 2 ...\n",
      " $ rs4661627 : Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 3 1 2 2 3 2 3 2 2 3 ...\n",
      " $ rs12093675: Factor w/ 2 levels \"AG\",\"GG\": 2 2 2 2 2 2 2 2 2 2 ...\n",
      " $ rs885795  : Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 3 1 3 2 3 3 2 3 3 1 ...\n",
      " $ rs603343  : Factor w/ 1 level \"GG\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs11203327: Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 2 3 3 3 2 1 NA 2 3 NA ...\n",
      " $ rs12754590: Factor w/ 2 levels \"AA\",\"AC\": 2 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs2297915 : Factor w/ 1 level \"CC\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs12405926: Factor w/ 2 levels \"AA\",\"AG\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs11203259: Factor w/ 3 levels \"CC\",\"CT\",\"TT\": NA NA 1 NA 1 1 1 1 1 NA ...\n",
      " $ rs4920306 : Factor w/ 1 level \"CC\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs6684536 : Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 2 1 3 2 2 1 1 1 2 NA ...\n",
      " $ rs2745311 : Factor w/ 3 levels \"GG\",\"GT\",\"TT\": 3 2 3 3 1 2 3 NA 2 1 ...\n",
      " $ rs16862254: Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 2 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs7512412 : Factor w/ 1 level \"GG\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs16862756: Factor w/ 2 levels \"AG\",\"GG\": 2 2 2 2 2 2 2 2 2 2 ...\n",
      " $ rs1266437 : Factor w/ 3 levels \"AA\",\"AC\",\"CC\": 3 3 3 3 2 NA 2 3 2 3 ...\n",
      " $ rs12128039: Factor w/ 1 level \"GG\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs12117738: Factor w/ 2 levels \"CC\",\"CG\": 1 1 1 1 1 2 2 1 1 2 ...\n",
      " $ rs16824016: Factor w/ 1 level \"CC\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs3753047 : Factor w/ 2 levels \"AG\",\"GG\": 2 2 2 2 2 1 2 2 1 1 ...\n",
      " $ rs11584316: Factor w/ 3 levels \"GG\",\"GT\",\"TT\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs3026792 : Factor w/ 2 levels \"CC\",\"CT\": 1 1 2 1 1 1 2 2 1 1 ...\n",
      " $ rs16825746: Factor w/ 2 levels \"AG\",\"GG\": 2 2 2 2 2 2 2 2 2 2 ...\n",
      " $ rs7531732 : Factor w/ 2 levels \"AG\",\"GG\": 2 2 2 2 2 2 2 2 2 2 ...\n",
      " $ rs17356087: Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 2 3 3 3 3 3 2 3 2 3 ...\n",
      " $ rs10799753: Factor w/ 2 levels \"AG\",\"GG\": 2 2 2 2 2 2 2 2 2 2 ...\n",
      " $ rs6426763 : Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 1 2 3 1 2 1 1 1 1 1 ...\n",
      " $ rs681648  : Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 1 2 1 1 1 1 1 2 1 1 ...\n",
      " $ rs4649089 : Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 3 2 2 2 2 1 2 2 1 1 ...\n",
      " $ rs6689554 : Factor w/ 2 levels \"CC\",\"CT\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs6693254 : Factor w/ 3 levels \"AA\",\"AC\",\"CC\": NA 3 2 2 3 2 3 2 2 3 ...\n",
      " $ rs17363576: Factor w/ 2 levels \"AA\",\"AG\": 1 1 NA 1 1 1 1 1 1 1 ...\n",
      " $ rs1702304 : Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 3 3 3 3 1 2 2 3 3 2 ...\n",
      " $ rs12737701: Factor w/ 1 level \"TT\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs298459  : Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 1 2 1 2 1 2 3 1 2 1 ...\n",
      " $ rs6598947 : Factor w/ 3 levels \"CC\",\"CG\",\"GG\": 3 3 3 3 3 3 3 3 3 3 ...\n",
      " $ rs282176  : Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 1 2 1 2 2 1 1 1 1 2 ...\n",
      " $ rs12043612: Factor w/ 2 levels \"CC\",\"CT\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs2272796 : Factor w/ 2 levels \"AA\",\"AG\": 1 1 1 2 1 1 2 1 1 1 ...\n",
      " $ rs514745  : Factor w/ 2 levels \"CC\",\"CT\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs2318372 : Factor w/ 1 level \"TT\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs12045777: Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 2 1 2 2 1 2 3 2 NA 3 ...\n",
      " $ rs10158631: Factor w/ 1 level \"GG\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs10915121: Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 2 2 1 1 2 2 1 2 2 3 ...\n",
      " $ rs12062432: Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 1 1 1 1 2 1 1 1 1 1 ...\n",
      " $ rs397488  : Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 1 1 1 2 1 1 1 2 2 3 ...\n",
      " $ rs12076353: Factor w/ 3 levels \"AA\",\"AT\",\"TT\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ rs670781  : Factor w/ 3 levels \"GG\",\"GT\",\"TT\": 1 2 1 2 2 2 1 1 1 1 ...\n",
      " $ rs12727605: Factor w/ 3 levels \"CC\",\"CT\",\"TT\": 1 2 2 2 2 3 1 3 2 1 ...\n",
      " $ rs12563847: Factor w/ 3 levels \"AA\",\"AG\",\"GG\": 3 3 3 3 3 NA 3 3 NA 3 ...\n",
      " $ rs11584510: Factor w/ 2 levels \"GG\",\"GT\": 2 1 1 1 1 1 1 1 1 1 ...\n",
      "  [list output truncated]\n"
     ]
    }
   ],
   "source": [
    "## Start with loading the `HapMap` data\n",
    "library(SNPassoc)\n",
    "data(HapMap)\n",
    "str(HapMap)     # Check what the data looks like by looking at the first few entries with the `str` function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>id</th><th scope=col>group</th><th scope=col>rs10399749</th><th scope=col>rs11260616</th><th scope=col>rs4648633</th><th scope=col>rs6659552</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>NA06985</td><td>CEU</td><td>CC</td><td>AA</td><td>TT</td><td>GG</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>NA06993</td><td>CEU</td><td>CC</td><td>AT</td><td>CT</td><td>CG</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>NA06994</td><td>CEU</td><td>CC</td><td>AA</td><td>TT</td><td>CG</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>NA07000</td><td>CEU</td><td>CC</td><td>AT</td><td>TT</td><td>GG</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>NA07022</td><td>CEU</td><td>CC</td><td>AA</td><td>CT</td><td>GG</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>NA07034</td><td>CEU</td><td>CC</td><td>AA</td><td>TT</td><td>CG</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & id & group & rs10399749 & rs11260616 & rs4648633 & rs6659552\\\\\n",
       "  & <fct> & <fct> & <fct> & <fct> & <fct> & <fct>\\\\\n",
       "\\hline\n",
       "\t1 & NA06985 & CEU & CC & AA & TT & GG\\\\\n",
       "\t2 & NA06993 & CEU & CC & AT & CT & CG\\\\\n",
       "\t3 & NA06994 & CEU & CC & AA & TT & CG\\\\\n",
       "\t4 & NA07000 & CEU & CC & AT & TT & GG\\\\\n",
       "\t5 & NA07022 & CEU & CC & AA & CT & GG\\\\\n",
       "\t6 & NA07034 & CEU & CC & AA & TT & CG\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 6\n",
       "\n",
       "| <!--/--> | id &lt;fct&gt; | group &lt;fct&gt; | rs10399749 &lt;fct&gt; | rs11260616 &lt;fct&gt; | rs4648633 &lt;fct&gt; | rs6659552 &lt;fct&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 1 | NA06985 | CEU | CC | AA | TT | GG |\n",
       "| 2 | NA06993 | CEU | CC | AT | CT | CG |\n",
       "| 3 | NA06994 | CEU | CC | AA | TT | CG |\n",
       "| 4 | NA07000 | CEU | CC | AT | TT | GG |\n",
       "| 5 | NA07022 | CEU | CC | AA | CT | GG |\n",
       "| 6 | NA07034 | CEU | CC | AA | TT | CG |\n",
       "\n"
      ],
      "text/plain": [
       "  id      group rs10399749 rs11260616 rs4648633 rs6659552\n",
       "1 NA06985 CEU   CC         AA         TT        GG       \n",
       "2 NA06993 CEU   CC         AT         CT        CG       \n",
       "3 NA06994 CEU   CC         AA         TT        CG       \n",
       "4 NA07000 CEU   CC         AT         TT        GG       \n",
       "5 NA07022 CEU   CC         AA         CT        GG       \n",
       "6 NA07034 CEU   CC         AA         TT        CG       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(HapMap[1:6])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notes: The `HapMap` data used consists of two groups of population, namely, CEU (Utah residents with ancestors from Northern and Western Europe) and YRI (Yoruba in Ibadan, Nigeria). Each group is represented by 60 samples (totally 120 samples). The form shows the head of the `HapMap` data with the sample ID group and nucleotide pairs at the SNP sites."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'data.frame':\t9305 obs. of  3 variables:\n",
      " $ snp       : chr  \"rs10399749\" \"rs11260616\" \"rs4648633\" \"rs6659552\" ...\n",
      " $ chromosome: chr  \"chr1\" \"chr1\" \"chr1\" \"chr1\" ...\n",
      " $ position  : int  45162 1794167 2352864 2902617 3170389 3382844 3766548 4038743 4246260 4387131 ...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>snp</th><th scope=col>chromosome</th><th scope=col>position</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>rs10399749</td><td>chr1</td><td>  45162</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>rs11260616</td><td>chr1</td><td>1794167</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>rs4648633 </td><td>chr1</td><td>2352864</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>rs6659552 </td><td>chr1</td><td>2902617</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>rs7550396 </td><td>chr1</td><td>3170389</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>rs12239794</td><td>chr1</td><td>3382844</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & snp & chromosome & position\\\\\n",
       "  & <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & rs10399749 & chr1 &   45162\\\\\n",
       "\t2 & rs11260616 & chr1 & 1794167\\\\\n",
       "\t3 & rs4648633  & chr1 & 2352864\\\\\n",
       "\t4 & rs6659552  & chr1 & 2902617\\\\\n",
       "\t5 & rs7550396  & chr1 & 3170389\\\\\n",
       "\t6 & rs12239794 & chr1 & 3382844\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| <!--/--> | snp &lt;chr&gt; | chromosome &lt;chr&gt; | position &lt;int&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | rs10399749 | chr1 |   45162 |\n",
       "| 2 | rs11260616 | chr1 | 1794167 |\n",
       "| 3 | rs4648633  | chr1 | 2352864 |\n",
       "| 4 | rs6659552  | chr1 | 2902617 |\n",
       "| 5 | rs7550396  | chr1 | 3170389 |\n",
       "| 6 | rs12239794 | chr1 | 3382844 |\n",
       "\n"
      ],
      "text/plain": [
       "  snp        chromosome position\n",
       "1 rs10399749 chr1         45162 \n",
       "2 rs11260616 chr1       1794167 \n",
       "3 rs4648633  chr1       2352864 \n",
       "4 rs6659552  chr1       2902617 \n",
       "5 rs7550396  chr1       3170389 \n",
       "6 rs12239794 chr1       3382844 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "str(HapMap.SNPs.pos)\n",
    "head(HapMap.SNPs.pos)\n",
    "## The `HapMap.SNPs.pos` file has the genomic mapping information\n",
    "## (names of SNPs, chromosomes and genetic position) for the SNPs on the human chromosomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A WGassociation: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>comments</th><th scope=col>dominant</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>rs10399749</th><td>Monomorphic</td><td>          NA</td></tr>\n",
       "\t<tr><th scope=row>rs11260616</th><td>NA         </td><td>9.850780e-02</td></tr>\n",
       "\t<tr><th scope=row>rs4648633</th><td>NA         </td><td>6.550480e-08</td></tr>\n",
       "\t<tr><th scope=row>rs6659552</th><td>NA         </td><td>1.175992e-10</td></tr>\n",
       "\t<tr><th scope=row>rs7550396</th><td>NA         </td><td>1.000000e+00</td></tr>\n",
       "\t<tr><th scope=row>rs12239794</th><td>Monomorphic</td><td>          NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A WGassociation: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & comments & dominant\\\\\n",
       "  & <fct> & <dbl>\\\\\n",
       "\\hline\n",
       "\trs10399749 & Monomorphic &           NA\\\\\n",
       "\trs11260616 & NA          & 9.850780e-02\\\\\n",
       "\trs4648633 & NA          & 6.550480e-08\\\\\n",
       "\trs6659552 & NA          & 1.175992e-10\\\\\n",
       "\trs7550396 & NA          & 1.000000e+00\\\\\n",
       "\trs12239794 & Monomorphic &           NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A WGassociation: 6 × 2\n",
       "\n",
       "| <!--/--> | comments &lt;fct&gt; | dominant &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| rs10399749 | Monomorphic |           NA |\n",
       "| rs11260616 | NA          | 9.850780e-02 |\n",
       "| rs4648633 | NA          | 6.550480e-08 |\n",
       "| rs6659552 | NA          | 1.175992e-10 |\n",
       "| rs7550396 | NA          | 1.000000e+00 |\n",
       "| rs12239794 | Monomorphic |           NA |\n",
       "\n"
      ],
      "text/plain": [
       "           comments    dominant    \n",
       "rs10399749 Monomorphic           NA\n",
       "rs11260616 NA          9.850780e-02\n",
       "rs4648633  NA          6.550480e-08\n",
       "rs6659552  NA          1.175992e-10\n",
       "rs7550396  NA          1.000000e+00\n",
       "rs12239794 Monomorphic           NA"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           comments    dominant\n",
      "rs10399749 Monomorphic -       \n",
      "rs11260616 -           0.09851 \n",
      "rs4648633  -           0.00000 \n",
      "rs6659552  -           0.00000 \n",
      "rs7550396  -           1.00000 \n",
      "rs12239794 Monomorphic -       \n"
     ]
    }
   ],
   "source": [
    "## Create the `SNP` class object, run the `WGassociation` function on this data object for the genetic model of your interest \n",
    "myHapMap <- setupSNP(HapMap, colSNPs= 3:9307, sort=TRUE,\n",
    "                     info=HapMap.SNPs.pos,sep=\"\")\n",
    "myHapMapres <- WGassociation(group, data= myHapMap, model=\"dominant\")\n",
    "\n",
    "# The result is a data frame. Check the resulting data frame\n",
    "head(myHapMapres)\n",
    "print(head(myHapMapres))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 3.6.2",
   "language": "R",
   "name": "ir362"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
