{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The contents include:\n",
    "\n",
    "* 7. Overcoming batch effects in expression data\n",
    "* 8. An exploratory analysis of data with PCA\n",
    "* 9. Finding the differentially expressed genes\n",
    "* 10. Working with the data of multiple classes\n",
    "\n",
    "\n",
    "\n",
    "## 7. Overcoming batch effects in expression data\n",
    "\n",
    "As processing samples in different batches, the systematic errors will be caused, which is named batch effects. Systematic errors can be reduced, to some extent, by careful, experimental design but cannot be eliminated completely unless the study is performed under a single batch. It is difficult to combine data from different batches due to batch effects, which makes the statistical analysis of the data unsuccessful. Therefore, if we do preprocessing before the batches are combined, the analysis will be easy. <br>\n",
    "处理不同批次的样品会导致系统误差，这种称为批量效应。 在一定程度上，通过仔细严谨的实验设计可以减少系统误差，但是除非单一批次进行研究，否则不能完全消除。由于批处理效应而难以合并不同批次的数据，从而使得数据的统计分析不顺利。因此，如果我们在批次合并之前进行预处理，分析将变得简单许多。<br>\n",
    "\n",
    "In this section, we need a dataset that shows the batch effect without any preprocessing. Take the `bladderbatch` data as an example, which consists of five batches. The data is a part of the `bladderbatch` package. We will use `sva` library as well.<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Bioconductor version 3.9 (BiocManager 1.30.10), R 3.6.2 (2019-12-12)\n",
      "\n",
      "Installing package(s) 'sva', 'bladderbatch'\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "package 'sva' successfully unpacked and MD5 sums checked\n",
      "\n",
      "The downloaded binary packages are in\n",
      "\tC:\\Users\\Master\\AppData\\Local\\Temp\\RtmpSI9RIu\\downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing the source package 'bladderbatch'\n",
      "\n",
      "\n",
      "Old packages: 'BH', 'bit', 'caTools', 'cli', 'FactoMineR', 'fansi', 'farver',\n",
      "  'GetoptLong', 'ggridges', 'gplots', 'hms', 'knitr', 'leaps', 'multcomp',\n",
      "  'mvtnorm', 'PMA', 'precrec', 'prettyunits', 'pROC', 'RCurl', 'RSQLite',\n",
      "  'Rttf2pt1', 'shinyjs', 'stringi', 'tinytex', 'xfun', 'XML', 'xts', 'zoo'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## install and library packages\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "install.packages(\"BiocManager\")\n",
    "\n",
    "BiocManager::install(c(\"sva\", \"bladderbatch\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(sva)\n",
    "librar\n",
    "library(bladderbatch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 57 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>sample</th><th scope=col>outcome</th><th scope=col>batch</th><th scope=col>cancer</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>GSM71019.CEL</th><td> 1</td><td>Normal  </td><td>3</td><td>Normal</td></tr>\n",
       "\t<tr><th scope=row>GSM71020.CEL</th><td> 2</td><td>Normal  </td><td>2</td><td>Normal</td></tr>\n",
       "\t<tr><th scope=row>GSM71021.CEL</th><td> 3</td><td>Normal  </td><td>2</td><td>Normal</td></tr>\n",
       "\t<tr><th scope=row>GSM71022.CEL</th><td> 4</td><td>Normal  </td><td>3</td><td>Normal</td></tr>\n",
       "\t<tr><th scope=row>GSM71023.CEL</th><td> 5</td><td>Normal  </td><td>3</td><td>Normal</td></tr>\n",
       "\t<tr><th scope=row>GSM71024.CEL</th><td> 6</td><td>Normal  </td><td>3</td><td>Normal</td></tr>\n",
       "\t<tr><th scope=row>GSM71025.CEL</th><td> 7</td><td>Normal  </td><td>2</td><td>Normal</td></tr>\n",
       "\t<tr><th scope=row>GSM71026.CEL</th><td> 8</td><td>Normal  </td><td>2</td><td>Normal</td></tr>\n",
       "\t<tr><th scope=row>GSM71028.CEL</th><td> 9</td><td>sTCC+CIS</td><td>5</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71029.CEL</th><td>10</td><td>sTCC-CIS</td><td>2</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71030.CEL</th><td>11</td><td>sTCC-CIS</td><td>5</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71031.CEL</th><td>12</td><td>sTCC-CIS</td><td>2</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71032.CEL</th><td>13</td><td>sTCC+CIS</td><td>5</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71033.CEL</th><td>14</td><td>sTCC-CIS</td><td>2</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71034.CEL</th><td>15</td><td>sTCC+CIS</td><td>5</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71035.CEL</th><td>16</td><td>sTCC+CIS</td><td>5</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71036.CEL</th><td>17</td><td>sTCC-CIS</td><td>2</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71037.CEL</th><td>18</td><td>mTCC    </td><td>1</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71038.CEL</th><td>19</td><td>sTCC+CIS</td><td>5</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71039.CEL</th><td>20</td><td>mTCC    </td><td>1</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71040.CEL</th><td>21</td><td>mTCC    </td><td>2</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71041.CEL</th><td>22</td><td>mTCC    </td><td>1</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71042.CEL</th><td>23</td><td>sTCC-CIS</td><td>2</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71043.CEL</th><td>24</td><td>sTCC+CIS</td><td>5</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71044.CEL</th><td>25</td><td>sTCC-CIS</td><td>2</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71045.CEL</th><td>26</td><td>sTCC-CIS</td><td>2</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71046.CEL</th><td>27</td><td>sTCC+CIS</td><td>5</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71047.CEL</th><td>28</td><td>mTCC    </td><td>1</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71048.CEL</th><td>29</td><td>mTCC    </td><td>1</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71049.CEL</th><td>30</td><td>sTCC-CIS</td><td>2</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71050.CEL</th><td>31</td><td>mTCC    </td><td>1</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71051.CEL</th><td>32</td><td>mTCC    </td><td>1</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71052.CEL</th><td>33</td><td>mTCC    </td><td>1</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71053.CEL</th><td>34</td><td>sTCC+CIS</td><td>5</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71054.CEL</th><td>35</td><td>mTCC    </td><td>1</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71055.CEL</th><td>36</td><td>sTCC-CIS</td><td>2</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71056.CEL</th><td>37</td><td>sTCC-CIS</td><td>2</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71058.CEL</th><td>38</td><td>sTCC-CIS</td><td>2</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71059.CEL</th><td>39</td><td>sTCC-CIS</td><td>2</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71060.CEL</th><td>40</td><td>mTCC    </td><td>1</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71061.CEL</th><td>41</td><td>sTCC+CIS</td><td>5</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71062.CEL</th><td>42</td><td>sTCC+CIS</td><td>5</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71063.CEL</th><td>43</td><td>sTCC+CIS</td><td>5</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71064.CEL</th><td>44</td><td>sTCC-CIS</td><td>2</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71065.CEL</th><td>45</td><td>sTCC-CIS</td><td>5</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71066.CEL</th><td>46</td><td>mTCC    </td><td>1</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71067.CEL</th><td>47</td><td>sTCC-CIS</td><td>5</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71068.CEL</th><td>48</td><td>sTCC+CIS</td><td>5</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>GSM71069.CEL</th><td>49</td><td>Biopsy  </td><td>4</td><td>Biopsy</td></tr>\n",
       "\t<tr><th scope=row>GSM71070.CEL</th><td>50</td><td>Biopsy  </td><td>4</td><td>Biopsy</td></tr>\n",
       "\t<tr><th scope=row>GSM71071.CEL</th><td>51</td><td>Biopsy  </td><td>5</td><td>Biopsy</td></tr>\n",
       "\t<tr><th scope=row>GSM71072.CEL</th><td>52</td><td>Biopsy  </td><td>5</td><td>Biopsy</td></tr>\n",
       "\t<tr><th scope=row>GSM71073.CEL</th><td>53</td><td>Biopsy  </td><td>5</td><td>Biopsy</td></tr>\n",
       "\t<tr><th scope=row>GSM71074.CEL</th><td>54</td><td>Biopsy  </td><td>5</td><td>Biopsy</td></tr>\n",
       "\t<tr><th scope=row>GSM71075.CEL</th><td>55</td><td>Biopsy  </td><td>4</td><td>Biopsy</td></tr>\n",
       "\t<tr><th scope=row>GSM71076.CEL</th><td>56</td><td>Biopsy  </td><td>4</td><td>Biopsy</td></tr>\n",
       "\t<tr><th scope=row>GSM71077.CEL</th><td>57</td><td>Biopsy  </td><td>4</td><td>Biopsy</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 57 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & sample & outcome & batch & cancer\\\\\n",
       "  & <int> & <fct> & <dbl> & <fct>\\\\\n",
       "\\hline\n",
       "\tGSM71019.CEL &  1 & Normal   & 3 & Normal\\\\\n",
       "\tGSM71020.CEL &  2 & Normal   & 2 & Normal\\\\\n",
       "\tGSM71021.CEL &  3 & Normal   & 2 & Normal\\\\\n",
       "\tGSM71022.CEL &  4 & Normal   & 3 & Normal\\\\\n",
       "\tGSM71023.CEL &  5 & Normal   & 3 & Normal\\\\\n",
       "\tGSM71024.CEL &  6 & Normal   & 3 & Normal\\\\\n",
       "\tGSM71025.CEL &  7 & Normal   & 2 & Normal\\\\\n",
       "\tGSM71026.CEL &  8 & Normal   & 2 & Normal\\\\\n",
       "\tGSM71028.CEL &  9 & sTCC+CIS & 5 & Cancer\\\\\n",
       "\tGSM71029.CEL & 10 & sTCC-CIS & 2 & Cancer\\\\\n",
       "\tGSM71030.CEL & 11 & sTCC-CIS & 5 & Cancer\\\\\n",
       "\tGSM71031.CEL & 12 & sTCC-CIS & 2 & Cancer\\\\\n",
       "\tGSM71032.CEL & 13 & sTCC+CIS & 5 & Cancer\\\\\n",
       "\tGSM71033.CEL & 14 & sTCC-CIS & 2 & Cancer\\\\\n",
       "\tGSM71034.CEL & 15 & sTCC+CIS & 5 & Cancer\\\\\n",
       "\tGSM71035.CEL & 16 & sTCC+CIS & 5 & Cancer\\\\\n",
       "\tGSM71036.CEL & 17 & sTCC-CIS & 2 & Cancer\\\\\n",
       "\tGSM71037.CEL & 18 & mTCC     & 1 & Cancer\\\\\n",
       "\tGSM71038.CEL & 19 & sTCC+CIS & 5 & Cancer\\\\\n",
       "\tGSM71039.CEL & 20 & mTCC     & 1 & Cancer\\\\\n",
       "\tGSM71040.CEL & 21 & mTCC     & 2 & Cancer\\\\\n",
       "\tGSM71041.CEL & 22 & mTCC     & 1 & Cancer\\\\\n",
       "\tGSM71042.CEL & 23 & sTCC-CIS & 2 & Cancer\\\\\n",
       "\tGSM71043.CEL & 24 & sTCC+CIS & 5 & Cancer\\\\\n",
       "\tGSM71044.CEL & 25 & sTCC-CIS & 2 & Cancer\\\\\n",
       "\tGSM71045.CEL & 26 & sTCC-CIS & 2 & Cancer\\\\\n",
       "\tGSM71046.CEL & 27 & sTCC+CIS & 5 & Cancer\\\\\n",
       "\tGSM71047.CEL & 28 & mTCC     & 1 & Cancer\\\\\n",
       "\tGSM71048.CEL & 29 & mTCC     & 1 & Cancer\\\\\n",
       "\tGSM71049.CEL & 30 & sTCC-CIS & 2 & Cancer\\\\\n",
       "\tGSM71050.CEL & 31 & mTCC     & 1 & Cancer\\\\\n",
       "\tGSM71051.CEL & 32 & mTCC     & 1 & Cancer\\\\\n",
       "\tGSM71052.CEL & 33 & mTCC     & 1 & Cancer\\\\\n",
       "\tGSM71053.CEL & 34 & sTCC+CIS & 5 & Cancer\\\\\n",
       "\tGSM71054.CEL & 35 & mTCC     & 1 & Cancer\\\\\n",
       "\tGSM71055.CEL & 36 & sTCC-CIS & 2 & Cancer\\\\\n",
       "\tGSM71056.CEL & 37 & sTCC-CIS & 2 & Cancer\\\\\n",
       "\tGSM71058.CEL & 38 & sTCC-CIS & 2 & Cancer\\\\\n",
       "\tGSM71059.CEL & 39 & sTCC-CIS & 2 & Cancer\\\\\n",
       "\tGSM71060.CEL & 40 & mTCC     & 1 & Cancer\\\\\n",
       "\tGSM71061.CEL & 41 & sTCC+CIS & 5 & Cancer\\\\\n",
       "\tGSM71062.CEL & 42 & sTCC+CIS & 5 & Cancer\\\\\n",
       "\tGSM71063.CEL & 43 & sTCC+CIS & 5 & Cancer\\\\\n",
       "\tGSM71064.CEL & 44 & sTCC-CIS & 2 & Cancer\\\\\n",
       "\tGSM71065.CEL & 45 & sTCC-CIS & 5 & Cancer\\\\\n",
       "\tGSM71066.CEL & 46 & mTCC     & 1 & Cancer\\\\\n",
       "\tGSM71067.CEL & 47 & sTCC-CIS & 5 & Cancer\\\\\n",
       "\tGSM71068.CEL & 48 & sTCC+CIS & 5 & Cancer\\\\\n",
       "\tGSM71069.CEL & 49 & Biopsy   & 4 & Biopsy\\\\\n",
       "\tGSM71070.CEL & 50 & Biopsy   & 4 & Biopsy\\\\\n",
       "\tGSM71071.CEL & 51 & Biopsy   & 5 & Biopsy\\\\\n",
       "\tGSM71072.CEL & 52 & Biopsy   & 5 & Biopsy\\\\\n",
       "\tGSM71073.CEL & 53 & Biopsy   & 5 & Biopsy\\\\\n",
       "\tGSM71074.CEL & 54 & Biopsy   & 5 & Biopsy\\\\\n",
       "\tGSM71075.CEL & 55 & Biopsy   & 4 & Biopsy\\\\\n",
       "\tGSM71076.CEL & 56 & Biopsy   & 4 & Biopsy\\\\\n",
       "\tGSM71077.CEL & 57 & Biopsy   & 4 & Biopsy\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 57 × 4\n",
       "\n",
       "| <!--/--> | sample &lt;int&gt; | outcome &lt;fct&gt; | batch &lt;dbl&gt; | cancer &lt;fct&gt; |\n",
       "|---|---|---|---|---|\n",
       "| GSM71019.CEL |  1 | Normal   | 3 | Normal |\n",
       "| GSM71020.CEL |  2 | Normal   | 2 | Normal |\n",
       "| GSM71021.CEL |  3 | Normal   | 2 | Normal |\n",
       "| GSM71022.CEL |  4 | Normal   | 3 | Normal |\n",
       "| GSM71023.CEL |  5 | Normal   | 3 | Normal |\n",
       "| GSM71024.CEL |  6 | Normal   | 3 | Normal |\n",
       "| GSM71025.CEL |  7 | Normal   | 2 | Normal |\n",
       "| GSM71026.CEL |  8 | Normal   | 2 | Normal |\n",
       "| GSM71028.CEL |  9 | sTCC+CIS | 5 | Cancer |\n",
       "| GSM71029.CEL | 10 | sTCC-CIS | 2 | Cancer |\n",
       "| GSM71030.CEL | 11 | sTCC-CIS | 5 | Cancer |\n",
       "| GSM71031.CEL | 12 | sTCC-CIS | 2 | Cancer |\n",
       "| GSM71032.CEL | 13 | sTCC+CIS | 5 | Cancer |\n",
       "| GSM71033.CEL | 14 | sTCC-CIS | 2 | Cancer |\n",
       "| GSM71034.CEL | 15 | sTCC+CIS | 5 | Cancer |\n",
       "| GSM71035.CEL | 16 | sTCC+CIS | 5 | Cancer |\n",
       "| GSM71036.CEL | 17 | sTCC-CIS | 2 | Cancer |\n",
       "| GSM71037.CEL | 18 | mTCC     | 1 | Cancer |\n",
       "| GSM71038.CEL | 19 | sTCC+CIS | 5 | Cancer |\n",
       "| GSM71039.CEL | 20 | mTCC     | 1 | Cancer |\n",
       "| GSM71040.CEL | 21 | mTCC     | 2 | Cancer |\n",
       "| GSM71041.CEL | 22 | mTCC     | 1 | Cancer |\n",
       "| GSM71042.CEL | 23 | sTCC-CIS | 2 | Cancer |\n",
       "| GSM71043.CEL | 24 | sTCC+CIS | 5 | Cancer |\n",
       "| GSM71044.CEL | 25 | sTCC-CIS | 2 | Cancer |\n",
       "| GSM71045.CEL | 26 | sTCC-CIS | 2 | Cancer |\n",
       "| GSM71046.CEL | 27 | sTCC+CIS | 5 | Cancer |\n",
       "| GSM71047.CEL | 28 | mTCC     | 1 | Cancer |\n",
       "| GSM71048.CEL | 29 | mTCC     | 1 | Cancer |\n",
       "| GSM71049.CEL | 30 | sTCC-CIS | 2 | Cancer |\n",
       "| GSM71050.CEL | 31 | mTCC     | 1 | Cancer |\n",
       "| GSM71051.CEL | 32 | mTCC     | 1 | Cancer |\n",
       "| GSM71052.CEL | 33 | mTCC     | 1 | Cancer |\n",
       "| GSM71053.CEL | 34 | sTCC+CIS | 5 | Cancer |\n",
       "| GSM71054.CEL | 35 | mTCC     | 1 | Cancer |\n",
       "| GSM71055.CEL | 36 | sTCC-CIS | 2 | Cancer |\n",
       "| GSM71056.CEL | 37 | sTCC-CIS | 2 | Cancer |\n",
       "| GSM71058.CEL | 38 | sTCC-CIS | 2 | Cancer |\n",
       "| GSM71059.CEL | 39 | sTCC-CIS | 2 | Cancer |\n",
       "| GSM71060.CEL | 40 | mTCC     | 1 | Cancer |\n",
       "| GSM71061.CEL | 41 | sTCC+CIS | 5 | Cancer |\n",
       "| GSM71062.CEL | 42 | sTCC+CIS | 5 | Cancer |\n",
       "| GSM71063.CEL | 43 | sTCC+CIS | 5 | Cancer |\n",
       "| GSM71064.CEL | 44 | sTCC-CIS | 2 | Cancer |\n",
       "| GSM71065.CEL | 45 | sTCC-CIS | 5 | Cancer |\n",
       "| GSM71066.CEL | 46 | mTCC     | 1 | Cancer |\n",
       "| GSM71067.CEL | 47 | sTCC-CIS | 5 | Cancer |\n",
       "| GSM71068.CEL | 48 | sTCC+CIS | 5 | Cancer |\n",
       "| GSM71069.CEL | 49 | Biopsy   | 4 | Biopsy |\n",
       "| GSM71070.CEL | 50 | Biopsy   | 4 | Biopsy |\n",
       "| GSM71071.CEL | 51 | Biopsy   | 5 | Biopsy |\n",
       "| GSM71072.CEL | 52 | Biopsy   | 5 | Biopsy |\n",
       "| GSM71073.CEL | 53 | Biopsy   | 5 | Biopsy |\n",
       "| GSM71074.CEL | 54 | Biopsy   | 5 | Biopsy |\n",
       "| GSM71075.CEL | 55 | Biopsy   | 4 | Biopsy |\n",
       "| GSM71076.CEL | 56 | Biopsy   | 4 | Biopsy |\n",
       "| GSM71077.CEL | 57 | Biopsy   | 4 | Biopsy |\n",
       "\n"
      ],
      "text/plain": [
       "             sample outcome  batch cancer\n",
       "GSM71019.CEL  1     Normal   3     Normal\n",
       "GSM71020.CEL  2     Normal   2     Normal\n",
       "GSM71021.CEL  3     Normal   2     Normal\n",
       "GSM71022.CEL  4     Normal   3     Normal\n",
       "GSM71023.CEL  5     Normal   3     Normal\n",
       "GSM71024.CEL  6     Normal   3     Normal\n",
       "GSM71025.CEL  7     Normal   2     Normal\n",
       "GSM71026.CEL  8     Normal   2     Normal\n",
       "GSM71028.CEL  9     sTCC+CIS 5     Cancer\n",
       "GSM71029.CEL 10     sTCC-CIS 2     Cancer\n",
       "GSM71030.CEL 11     sTCC-CIS 5     Cancer\n",
       "GSM71031.CEL 12     sTCC-CIS 2     Cancer\n",
       "GSM71032.CEL 13     sTCC+CIS 5     Cancer\n",
       "GSM71033.CEL 14     sTCC-CIS 2     Cancer\n",
       "GSM71034.CEL 15     sTCC+CIS 5     Cancer\n",
       "GSM71035.CEL 16     sTCC+CIS 5     Cancer\n",
       "GSM71036.CEL 17     sTCC-CIS 2     Cancer\n",
       "GSM71037.CEL 18     mTCC     1     Cancer\n",
       "GSM71038.CEL 19     sTCC+CIS 5     Cancer\n",
       "GSM71039.CEL 20     mTCC     1     Cancer\n",
       "GSM71040.CEL 21     mTCC     2     Cancer\n",
       "GSM71041.CEL 22     mTCC     1     Cancer\n",
       "GSM71042.CEL 23     sTCC-CIS 2     Cancer\n",
       "GSM71043.CEL 24     sTCC+CIS 5     Cancer\n",
       "GSM71044.CEL 25     sTCC-CIS 2     Cancer\n",
       "GSM71045.CEL 26     sTCC-CIS 2     Cancer\n",
       "GSM71046.CEL 27     sTCC+CIS 5     Cancer\n",
       "GSM71047.CEL 28     mTCC     1     Cancer\n",
       "GSM71048.CEL 29     mTCC     1     Cancer\n",
       "GSM71049.CEL 30     sTCC-CIS 2     Cancer\n",
       "GSM71050.CEL 31     mTCC     1     Cancer\n",
       "GSM71051.CEL 32     mTCC     1     Cancer\n",
       "GSM71052.CEL 33     mTCC     1     Cancer\n",
       "GSM71053.CEL 34     sTCC+CIS 5     Cancer\n",
       "GSM71054.CEL 35     mTCC     1     Cancer\n",
       "GSM71055.CEL 36     sTCC-CIS 2     Cancer\n",
       "GSM71056.CEL 37     sTCC-CIS 2     Cancer\n",
       "GSM71058.CEL 38     sTCC-CIS 2     Cancer\n",
       "GSM71059.CEL 39     sTCC-CIS 2     Cancer\n",
       "GSM71060.CEL 40     mTCC     1     Cancer\n",
       "GSM71061.CEL 41     sTCC+CIS 5     Cancer\n",
       "GSM71062.CEL 42     sTCC+CIS 5     Cancer\n",
       "GSM71063.CEL 43     sTCC+CIS 5     Cancer\n",
       "GSM71064.CEL 44     sTCC-CIS 2     Cancer\n",
       "GSM71065.CEL 45     sTCC-CIS 5     Cancer\n",
       "GSM71066.CEL 46     mTCC     1     Cancer\n",
       "GSM71067.CEL 47     sTCC-CIS 5     Cancer\n",
       "GSM71068.CEL 48     sTCC+CIS 5     Cancer\n",
       "GSM71069.CEL 49     Biopsy   4     Biopsy\n",
       "GSM71070.CEL 50     Biopsy   4     Biopsy\n",
       "GSM71071.CEL 51     Biopsy   5     Biopsy\n",
       "GSM71072.CEL 52     Biopsy   5     Biopsy\n",
       "GSM71073.CEL 53     Biopsy   5     Biopsy\n",
       "GSM71074.CEL 54     Biopsy   5     Biopsy\n",
       "GSM71075.CEL 55     Biopsy   4     Biopsy\n",
       "GSM71076.CEL 56     Biopsy   4     Biopsy\n",
       "GSM71077.CEL 57     Biopsy   4     Biopsy"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Load `bladderdata`. And extract the expression matrix and pheno data from it.\n",
    "data(bladderdata)\n",
    "\n",
    "pheno <- pData(bladderEset)\n",
    "edata <- exprs(bladderEset)\n",
    "pheno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ExpressionSet (storageMode: lockedEnvironment)\n",
       "assayData: 22283 features, 8 samples \n",
       "  element names: exprs, se.exprs \n",
       "protocolData: none\n",
       "phenoData\n",
       "  sampleNames: GSM71019.CEL GSM71020.CEL ... GSM71026.CEL (8 total)\n",
       "  varLabels: sample outcome batch cancer\n",
       "  varMetadata: labelDescription\n",
       "featureData: none\n",
       "experimentData: use 'experimentData(object)'\n",
       "Annotation: hgu133a "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mydata <- bladderEset[,sampleNames(bladderEset)[1:8]]\n",
    "mydata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And then use `arrayQualityMetrics` to perform a quality check on the data, and then we will have a look at the batch effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(arrayQualityMetrics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The directory 'F:/qc_be' has been created.\n",
      "\n",
      "Warning message in svgStyleAttributes(style, svgdev):\n",
      "\"Removing non-SVG style attribute name(s): subscripts, group.number, group.value\"\n",
      "Warning message in svgStyleAttributes(style, svgdev):\n",
      "\"Removing non-SVG style attribute name(s): subscripts, group.number, group.value\"\n",
      "Warning message in svgStyleAttributes(style, svgdev):\n",
      "\"Removing non-SVG style attribute name(s): subscripts, group.number, group.value\"\n",
      "Warning message in svgStyleAttributes(style, svgdev):\n",
      "\"Removing non-SVG style attribute name(s): subscripts, group.number, group.value\"\n",
      "Warning message in svgStyleAttributes(style, svgdev):\n",
      "\"Removing non-SVG style attribute name(s): subscripts, group.number, group.value\"\n",
      "Warning message in svgStyleAttributes(style, svgdev):\n",
      "\"Removing non-SVG style attribute name(s): subscripts, group.number, group.value\"\n",
      "Warning message in svgStyleAttributes(style, svgdev):\n",
      "\"Removing non-SVG style attribute name(s): subscripts, group.number, group.value\"\n",
      "Warning message in svgStyleAttributes(style, svgdev):\n",
      "\"Removing non-SVG style attribute name(s): subscripts, group.number, group.value\"\n",
      "(loaded the KernSmooth namespace)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "arrayQualityMetrics(mydata, outdir=\"F:/qc_be\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Take a look at the heatmap and other plots produced for the samples to check for the batch effect. \n",
    "## In the filefolder, there are files called 'hm' and 'ma'.\n",
    "library(hexbin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "browseURL(\"F:/qc_be/index.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A matrix: 8 × 2 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>(Intercept)</th><th scope=col>as.factor(cancer)Normal</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>GSM71019.CEL</th><td>1</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>GSM71020.CEL</th><td>1</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>GSM71021.CEL</th><td>1</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>GSM71022.CEL</th><td>1</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>GSM71023.CEL</th><td>1</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>GSM71024.CEL</th><td>1</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>GSM71025.CEL</th><td>1</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>GSM71026.CEL</th><td>1</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 8 × 2 of type dbl\n",
       "\\begin{tabular}{r|ll}\n",
       "  & (Intercept) & as.factor(cancer)Normal\\\\\n",
       "\\hline\n",
       "\tGSM71019.CEL & 1 & 1\\\\\n",
       "\tGSM71020.CEL & 1 & 1\\\\\n",
       "\tGSM71021.CEL & 1 & 1\\\\\n",
       "\tGSM71022.CEL & 1 & 1\\\\\n",
       "\tGSM71023.CEL & 1 & 1\\\\\n",
       "\tGSM71024.CEL & 1 & 1\\\\\n",
       "\tGSM71025.CEL & 1 & 1\\\\\n",
       "\tGSM71026.CEL & 1 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 8 × 2 of type dbl\n",
       "\n",
       "| <!--/--> | (Intercept) | as.factor(cancer)Normal |\n",
       "|---|---|---|\n",
       "| GSM71019.CEL | 1 | 1 |\n",
       "| GSM71020.CEL | 1 | 1 |\n",
       "| GSM71021.CEL | 1 | 1 |\n",
       "| GSM71022.CEL | 1 | 1 |\n",
       "| GSM71023.CEL | 1 | 1 |\n",
       "| GSM71024.CEL | 1 | 1 |\n",
       "| GSM71025.CEL | 1 | 1 |\n",
       "| GSM71026.CEL | 1 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "             (Intercept) as.factor(cancer)Normal\n",
       "GSM71019.CEL 1           1                      \n",
       "GSM71020.CEL 1           1                      \n",
       "GSM71021.CEL 1           1                      \n",
       "GSM71022.CEL 1           1                      \n",
       "GSM71023.CEL 1           1                      \n",
       "GSM71024.CEL 1           1                      \n",
       "GSM71025.CEL 1           1                      \n",
       "GSM71026.CEL 1           1                      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Create the model matrix for the dataset \n",
    "## (note that only the first and third columns have been used from the model matrix as the data has only one condition).\n",
    "mod1 <- model.matrix(~as.factor(cancer), data=pData(mydata))[,c(1,3)]\n",
    "mod1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "batch <- pData(mydata)$batch  ## define the batches\n",
    "\n",
    "\n",
    "## Extract the expression matrix from the expression set object myData, where the batch effect has to be removed\n",
    "edata <- exprs(mydata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Found2batches\n",
      "\n",
      "Adjusting for0covariate(s) or covariate level(s)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Standardizing Data across genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting L/S model and finding priors\n",
      "\n",
      "Finding parametric adjustments\n",
      "\n",
      "Adjusting the Data\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "combat_edata <- ComBat(dat=edata, batch=batch, mod=mod1, par.prior=TRUE) # numCovs=NULL?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create an ExpressionSet object with everything as the original input data, \n",
    "## except the expression matrix—which is replaced by the matrix received as a result of the `ComBat` function in the last step.\n",
    "\n",
    "mydata2 <- mydata\n",
    "exprs(mydata2) <- combat_edata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The directory 'F:/qc_nbe' has been created.\n",
      "\n",
      "Warning message in svgStyleAttributes(style, svgdev):\n",
      "\"Removing non-SVG style attribute name(s): subscripts, group.number, group.value\"\n",
      "Warning message in svgStyleAttributes(style, svgdev):\n",
      "\"Removing non-SVG style attribute name(s): subscripts, group.number, group.value\"\n",
      "Warning message in svgStyleAttributes(style, svgdev):\n",
      "\"Removing non-SVG style attribute name(s): subscripts, group.number, group.value\"\n",
      "Warning message in svgStyleAttributes(style, svgdev):\n",
      "\"Removing non-SVG style attribute name(s): subscripts, group.number, group.value\"\n",
      "Warning message in svgStyleAttributes(style, svgdev):\n",
      "\"Removing non-SVG style attribute name(s): subscripts, group.number, group.value\"\n",
      "Warning message in svgStyleAttributes(style, svgdev):\n",
      "\"Removing non-SVG style attribute name(s): subscripts, group.number, group.value\"\n",
      "Warning message in svgStyleAttributes(style, svgdev):\n",
      "\"Removing non-SVG style attribute name(s): subscripts, group.number, group.value\"\n",
      "Warning message in svgStyleAttributes(style, svgdev):\n",
      "\"Removing non-SVG style attribute name(s): subscripts, group.number, group.value\"\n"
     ]
    }
   ],
   "source": [
    "## Rerun the `arrayQualityMetrics` function to check for the elimination of batch effects with this new object as the input.\n",
    "arrayQualityMetrics(mydata2, outdir=\"F:/qc_nbe\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "browseURL(\"F:/qc_nbe/index.html\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8. An exploratory analysis of data with PCA\n",
    "\n",
    "If we measure too many human genes(>2000) in many samples, we will get a multiple column-row matrix. That will makes it difficult to build a pattern. So it is better to smplify the multidimensional pattern to a lower dimensions in order to explain and graphically represent the data.<br>\n",
    "Principal Components Analysis (PCA) is a method that achieves this by performing a covariance analysis between factors. This finds the orthogonal components that represent the data and each component (called principal components) that represents the dimension where the features are more extended. Thus, PCA projects data onto a lower dimensional space and can be used as an exploratory method to serve purposes such as finding patterns in data and noise reduction. In this section, we will deal with PCA for microarray data.<br>\n",
    "如果我们在很多样本中测量了太多的人类基因（>2000）,我们将得到一个多列-多行矩阵。这会让人难以构建出一种模式。因此，将多维模式缩小到更低的维度，能够很好的解释并图形化这些数据。<br>\n",
    "主成分分析（PCA）是一种通过执行因素之间的协方差分析来实现这一点的方法。这将会找到表现数据的正交分量和每个组件（称为主要组件），这些组件都有着被极大扩展的维度特征。因此，PCA将数据投影到较低维的空间上，又可以被用作探索性方法来提供诸如查找数据模式和降低噪声等的用途。 在本节中，我们将处理PCA的微阵列数据。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(affy)\n",
    "mydata <- ReadAffy(celfile.path= \"F:/GSE24460_RAW/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "## select your data for the analysis.\n",
    "mydata.pca <- exprs(mydata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Transpose the matrix to get the genes (features) as columns\n",
    "mypca <- prcomp(mydata.pca, scale=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Importance of components:\n",
       "                          PC1     PC2     PC3     PC4\n",
       "Standard deviation     1.9305 0.46011 0.17821 0.17182\n",
       "Proportion of Variance 0.9317 0.05293 0.00794 0.00738\n",
       "Cumulative Proportion  0.9317 0.98468 0.99262 1.00000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## check the principal components, take a look at the summary of created objects\n",
    "summary(mypca)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create a vector of colors for every sample in the data\n",
    "colors <- c(\"green\",\"cyan\",\"violet\",\"magenta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAPFBMVEUAAAAA/wAA//9NTU1o\naGh8fHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enugu7w8PD/AP////9qtqz4AAAA\nCXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO2di2KbOgyGsdttPWu3uub93/UgySTpysUQ\nX6T0/09P1jRAhMWHZfnCMEIQdLeG3gZA0CMIIEFQAQEkCCoggARBBQSQIKiAABIEFRBAgqAC\nAkgQVEAACYIKCCBBUAEBJAgqIIAEQQUEkCCogAASBBUQQIKgAgJIEFRAAAmCCgggQVABASQI\nKiCABEEFBJAgqIAAEgQVEECCoAICSBBUQAAJggoIIEFQAQEkCCoggARBBQSQIKiAABIEFRBA\ngqACAkgQVEAACYIKCCBBUAEBJAgqIIAEQQUEkCCogAASBBUQQIKgAgJIEFRAAAmCCgggQVAB\nAaQbDZOu/2xscf4AUuL737DzRXa1e34ZThh3Pt8pu739T+kxvXVOUsDzPxtbnD8AOXDzMDmH\nsKzd88twwjDmFOG6CTv7n9MjOusu3QPS/gHyQKrgZ1XavVFsk3YfSLv7n9PDOuusKoPEnwKk\nO0Aa7wRpd/9zelhnnVV3kIadOsu6cs4PIJlXVgvmrgPsbCBefmCQxv3z2yENIClWyuHc66OM\nA2Sh+pAgXfJka+e354RCTgJI1TV8ftn4/NwBBtZ3Bemi/cgNNZJpDdfXcz7aPcD4nWsk0v75\n7ZchQFKuVF+c7+vLOEByIzpk184vxwk3x9n4inUTcjY6qsf0FgQ1FkCCoAICSBBUQAAJggoI\nIEFQAQEkCCoggARBBQSQIKiAABIEFRBAgqACAkj/aLdA9kusxSFM6+4Cur+EixfwY3vshABS\nfQGkbyCAVF8A6RsIINUXQPoGAkj1BZC+gQBSfQGkbyCAVF8A6RsIINUXQNo4jn7Zs3iwbrEF\nk0sBoOw49fTFx12sOKR/QepjxRHZK2RtANgrMf0WA6QG0gaAvRLTbHGM/M8SSOkjpVovZK12\nawNA82UpMgRS5J9FkOaPlGq1kNXarQ0AxZdlkh2Q4vzyFaTLR0q1Vsh67dYGgN7LchZAaiCA\npOU49WQHJIR2DaUNAMWXZZIhkJBsaCdtAGi+LEWWQEpC+ru+tAFgr8T0WwyQGmjRwhOjHgCS\nYgGk+lqycNj68MBxzsheiem3GCA1EEA6Kns+BkgNBJCOyp6PAVIDAaSjsudjgNRASDYclT0f\nA6QGKjUvCSApFkCqr7Ua6ehDZgGSYgGk+lprIw1rHx44zhnZKzH9FgOkBgJIR2XPxwCpgQDS\nUdnzMUBqIIB0VPZ8DJAaCMmGo7LnY4DUQNoAsFdi+i0GSA2kDQB7JabfYoDUQNoAsFdi+i0G\nSA2kDQB7JabfYoDUQNoAsFdi+i0GSA2kDQB7JabfYoDUQNoAsFdi+i0GSA2kDQB7JabfYoDU\nQNoAsFdi+i0GSA2kDQB7JabfYoDUQNoAsFdi+i0GSA2kDQB7JabfYoDUQNoAsFdi+i0GSA2k\nDQB7JabfYoDUQNoAsFdi+i0GSA2kDQB7JabfYoDUQNoAsFdi+i0GSA2kDQB7JabfYoDUQNoA\nsFdi+i0GSA2kDQB7JabfYoDUQNoAsFdi+i0GSA2kDQB7JabfYoDUQIsW4mkUG7LnY4DUQEsW\nDlsfHjjOGdkrMf0WA6QGAkhHZc/HAKmBANJR2fMxQGoggHRU9nwMkBoIyYajsudjgNRAePTl\nUdnzMUBqoOUaaRzxNIo12fMxQGqgtTbSsPbhgeOckb0S028xQGoggHRU9nwMkBoIIB2VPR8D\npAZaTX8PKx8eOM4Z2Ssx/RYDpAZC+vuo7PkYIDWQNgDslZh+iwFSA2kDwF6J6bcYIDWQNgDs\nlZh+iwFSA2kDwF6J6bcYIDWQNgDslZh+iwFSA2kDwF6J6bcYIDWQNgDslZh+iwFSA2kDwF6J\n6bcYIDWQNgDslZh+iwFSA2kDwF6J6bcYIDWQNgDslZh+iwFSA2kDwF6J6bcYIDWQNgDslZh+\niwFSA2kDwF6J6bcYIDWQNgDOH+eyjMvzy9/LH99enqe//Hy93fD3fbba8zFAaqD+AJQ6znCj\n3/Knv8/zH57eLtu937likj0fA6QG6g9AqeNc+Pj4PQxcJ71NddHbx/TLn5dhmEl6fwJIfaw4\nInuF3B+AUse54eP38Gt6/TMMl5Du48fwzr8QRwBJvewVcn8ASh3nho8P/v15uGka/WG2JsSG\nF00geX/P3rlSAtKBk9UFUo7h/QEodZxbPuj3t+HH7ceScKDmkyKQPP9Ulw6QjpysKpCyDO8P\nQKnjfKqRnsbx1/D6daOf76MikPzlpa5UgHToZDWBlGd4fwBKHeeGj5fhhSK7990NT33Pzvtc\nheC/F0jeT6ecubEqkLIMX7TQ5HJcF3P//hqePjZ40QHS5Bom6duEdtPlOP1kbqwLpBzDlywc\ntj48cJwzKtKPxL1GukHiW9zkoG+TbPBytpkkKQIp0/AHBOnnfx/yfnXD09/Bu++8z1MC6S5L\nsqUEpPSSI10g5Rj+SCB9fv9Ddxsp+Bim2PsuU3KlAaSRzjeXI0UgxcgtpG8V2n1+//I5a/ff\nn7UNj37PzvtcTe6JMfvKuksqQIrUZp/OOEsqQCJjYzJ8d+MHTDYk/fncj3R9pwOkGKerqhFJ\nCkBihNKVmSENICWDwxhDhtWP8+jLL6fwPFxHqo6/r2+0gMT+id8DJLoa+b6RSZICkGJ6oVvA\nVJXubb5h4SHjFYL05zpSdXwdntc3PPg9O+/zxJdVCM7dZUumuoMU+MZBJ2wNJIbIuXMgDSeq\nKYUgTa0kGf39/vo8PL1vbHjse3be5yn6OHnHhbEFSUpA8tPluH9FshSANMZUIVHTbt9LK22k\n1Y8OHeeESoI01UOzfrxvbnjoe3be58hTXTS5h5pIDUjqDhKFdn6KYl3IzFN2BIkslOxCnPiZ\nOHLTDWCfpBUL6WKzD9L48frzaRiefr3tbXjke3beZ4i7yic5aid9C5A42x9i8E47SIFsDSMN\nZZg849juyVFnQeL47sj39weple73sadbsw/T7dnlBA33SwFIfpyuRv7JIqkbSGSdlxf2zOQk\n6kc6GdrtfnTnxg2OU093+5jGbhFIgZMNmR0rd0kDSIEbGyHmxXY9QaIKiZ3E2foQuOv8fI10\nUNqOU0/3+thTB5+ji4paSpn9KvdJAUgj575CVt/m2DW0C0SSF/CpYTfZ7TMCcG0APD5IUaoj\nR6EdN5RKGbYuBSAFSVJmjbYZ+4I03eUomosjpRnEbqmQNu8B2gAwCtKB/iA/tVwDR3cU6cTs\nUTN3SAFIAhGf+CZIqSDvLeTzmiofqoloxLejOklgIpLIWev7aQPAJkhuzE++UW0U6aKiFx+z\nxp/cqf4gcW3k+WVzxPtckPcW8h0Kkl+g6I5ucwmkwH2yGyRpA8AkSJLfydudU8COKyQvDdmY\n2Ul5XgpAcpxF5obhxmaXgryzkO8RpRY4qKObHIcNUyPJp7w9QCqnu3xMFRC3ZLmd5AN3rlSv\nkrqD5MaE0XR1bpGkACQegeG5J4l9JL0UU7uJaqONW542AEyClB11TA1YzgJTcpUD8QmjyUW1\nSeoPkpf6lznaqZL6hnY85DskDzFHUYahOO5RXt9RGwA2QcpsB0daoYH797nVzTP7pvg7d27B\naXUHyUtehV789rTgzsmGSMbx4FrJgKcUnqM0ybiZFtIGgFGQ8iRuStGCl1wDhTqPHtrJLSP1\nx2ZNr++V/pbbGlWgFIzyLxzX7S+uoQ2AhwZplK7yuReJZ4y2mJKkACQRNd+z0v29QHJRZk6I\nexxn7MnqiaMdkrQB8NggcdTAYZ3nftlIKeHMcZx3qDdII9e8njKUMW9UVCeQHA/5llpIGkXO\ns5Oo33ybJG0APDhItHAQNZAik0QNBhrHWc64ZXUHiZcQoRs7XZM5JHUDKUpI59hM4oj6kggm\ngFRad/nY8zwkyl1NfuLBkDThpTZJCkCi8YWUodQNEoV2VBtNrzwtlvqSovT1IbQrrXt8TNVR\nlC5Zar9yxUTRXUn7FtQdJKqQ6AZP2YaYc77daiSC3PGkeI7BeaAQmS7Jho3UvTYAHhwkihdk\nfpuM5aKI3Lva/SP9QXJ8MXIKPKsG7gMS2TVRJD6K3Ovl3RzdOU6/r5KkDYDHBmnkmW0xSvce\npxu8m3z16CA5aRCGNPJmf4d+ICWShHuKwCcHUcqBJ/dukKQNgAcHaZTJfEEcJM85cNUbSb1B\nojMNDBPd2XPGu/fL2k2NI643Y5QEeGRHcT3Ks88BUindl/52qYvfOekypzDP1+6z7w9S5I4z\nutHTiKj9HuhuyQZpI3GkQHkhholbd9ERSetdFdoAeCiQvrRNQxSMZEQx98xSIyk+OEjSNHRz\nn+wNSGut95IgHVtgPYY0kkk8lcYXU66RGklINpRTvo/Dl4g6yCzR1HGegoeM8Sd3qjNIKaeS\nxkVNFfCVo7U2R0GQVr9jZXPOqs4rAcyTPxxlWzfbstoAeCCQliawpKEnaaKLLAdCfSx1SeoN\n0sjrNaQhQkRS0voUn3Ig7Uwj+rp5mkbBFl+6kZ2XGb7rOy5aaHIR/VY6B5LczLiTL8jVJEts\ncD/S/kiu+9Q7tOMkmGSUqbF0IUkjSJGfLJbWKQ7SHSuT/LYf27dk4bD14YHjnNEDgXQbVMhc\nmukKCilO4MGQPBol8kCHhwZJxn5yQ0NuHJcPFIZ2Y6qNPIMU5N5HgbhbT9mtWAiQtnQm2ZCq\nIx4HzatSyR06pDFd8bFDO352pJP2oePBq9dP1CUb+CHZkq7zMsKBe5V4ctJWlQSQjuqMj2eQ\nPK//wb0TUh9R35/MSaqp/iBxWEdDbyhiUjzWjhRTC4ljOo5GI/efh00kAdJRnfKxm3+oSpIZ\nzCGtUkOhXeXHjfUGaZynYCWQ9nfoCRINOHEShKZhKFIfEUnreyHZcFTnfMxxHfWn+LQQF/fy\n8QwdH2uvW9wdJBro7iTRELPWO+8IkuNlynlMk9SiHDgEHoG3sdvjPLGvlU77mId70y1OWt5R\n+pRk0uhjd8imhwzxmfKii7s79ASJLJR0HdvKbxw/z+rwSqszWAjtFnTWx96P86KD/C8PO+FJ\nBYRW1T7Z3iCle0aaIusz1vHrCFJM9zieeykxXZTx34dBukAEkBZ0wMe3ozMpjBs5SmAf+SCT\nRqPcqd21c6WCeoOUVnmJUZLfmpMNUTgKyUtyx+M0ntu+AawnGwaAtKh8H8fbwajycIOQ+vg5\nCSxxTpCF02JFkrqDJCGdPLOLYqXdPXqBFCW7wHn6KDMvQ5rJTBwdXNdumP8BSAvK9nH8tKQh\n398kaeXS2C0vU8hk9iiRVAul3kOEpDbipxRS3ZuR/+4EEmW6hSNZCoCs5Z4vXsvzLEiHnhEJ\nkL6I29T8wo8lHdMIrnmRGvEZX2Q8IpIGclUiqTdIY1qIXpYsPfOM8FYgRc4sBnd1FE944fX/\nTywQOWx9eOA4J/RAIPHiTrxMGi8eRN0QfD/mHlluFfFkTP5/7l2pQ1JfkJgjXqw03es3l/4V\ndQvt5oXzrw+wkuWKOSd0tEY6I23HqacDPqblBilbNyFF66nKaLN0OTFKc7KB5l/y/e4xQQqy\nfPb1Nq8ZpDT2ZHYT16CeE0ObExK1AfBQIBFJ6QGk9HJ5tkGaU0DRHT97noYI0U3vUUM7eaJ5\nuJ6+WpA4axcuMeh8x5PRXJsTe7UB8Fgg0TWU4jrHPeYS1aXYwfMDrGQNcPo480HFd5vcuox9\n6nxOLQ7dY+0EpMuDSVPNxAH4pne0AfBgII0+PctFBsiE+TFBLiUc0hoo/CireHDA/2mTm4NE\n2TAfLnkW9R2y17Bubs8SR9vO0QbAo4HEARsv5SYLDqbsd3rqlvM8QpJQ4jWfNme8FDO5fWjn\nLtGSnPvuLv1A4sdQpCZsai251FmxuZ82AB4NpFFGmlAfEQ/amlFieV6Ki//hX0dfZ0H93iBx\nfTyHtXwT2dulG0i0UNC8fuen2GFvETFtADwcSCPD4530kAc/P0tRpiXRJ7LUk2OItlYFKGZy\n86wdrZh0ySZTa3GXpH4gyWI0Mjoo9cxK7bTzXEVtADweSGlMg4xkCOyflAzm8Vvc4S8gue3l\nNYqZ3KaM5zs4dQBIDOvkmqRRAnun2TG045kuYnKcH/DC3ecr2e90ntoAeECQxnlcXRooE2VU\nEM8c4+5+GagvoV0Lk5uU8aXXJfqRh9h4ua8HrpnVguRlyqWsnh9jeqITd5v7RZDm89QGwCOC\nNPIgu/lhW3GunuQpSSEt2hAl2VBFHUCKlxd61rScskxKcjGj3u0HEq+XJqFdTDOopKdiubf8\ncp7aAHhIkOQBfS6tGjRXUDwUnO95kX+plGn4amJjkLiVyFNOeUawbpCEJD8v9z9P/eBkA0Aq\nqjM+lsRckDXzBR1Z6I0dxAMcaIWuGuYumNgmtLssFuRlPlJ6Pkrk6fZ7e3cBSUbfy1QxxzTF\nNDdp44l9CO3O6pSP+Tku8xJp8hQ4GfYdJOwJDzcfyV/v4NLccF4GufudIQKsHiD5S/AW0lN+\nJVhItq+5B8mGkzrn45QB5oDBpaasd/LgS7nC9h5Seoc6gPTpAXccxM4tj6ynQXUAiSwWg2Ua\nr/Tt0W2Oo4W90RjaAHhUkJgkudHJg+A4ZycJYR4i5CouW9wJJBpjmN6lu4dPHc+76gMSrbI6\n8iAhfrCYn0ec5MyO1wbAw4I0yjq4cfZQlKeTBhnZwDfs0paumdgmtBv5hIQkJ4l/ebRaTlOw\nS2gXBCaajcmPy/ayfhqHD7t7awPgcUEanTx9wkvozU7iATPyFLuai6126ZD1afl6akOkXIs0\nkXJuGF2SDTP5UYY3iI+4N/bEMNuz0nacejrvY1m2IPo5GxzTnBd+il3NtVY7DRHi0eyc1eLm\noONhh8HtjP5k9Ul/y/B7abymzr4okxGPp0fOSttx6ukOH/MSGsLR/JjSy7yXojb+o15j7UJI\n/Sxpobh5RvD+np36kUIQ7HlVJ+FIVogM497KNNoAeGiQRh4pw4uyy3oNPj3dpaR9C+o4aFVA\n8lF6Y2SsQMZu/TpkuQr1shqXZL9l8N3eMgDaAHhskEbue+WHG6RRrL52bUTqOfo7ylTzmDji\nVWX39+oIEls8dyDJoJP0RKdNu7UB8OAgMUnMUUwReO3HXpJ6gkTJBs70S2BH3bTKQRolAE0c\nyXqrAKm87vQxd5LL6Ib4LUDi5V+8rPLC7UK1WbuLuFsipgdR0KAuPyK0K657fSyr9qbIIeR0\nUdytviCNPHzazXf4rPPtDpK7hKJBupCRbCitu31MA1edPBgpq6vvfnUGaZSZwT6/idQZpFHG\nPkbJjvicfL06AL4BSGNIq9lx0q6MUdvqDdIoD1WLMj1O68iGW0VZLi0tYXwaJDyxb0MFfCzd\nSUxSEZP21B0kfqpNTNOELYA0cmo1Pd/3VDD66W8HzAdIB8RTxpxfnipWQQpAomkVjgcJZc26\n6g4Sr3wb5InMWdsvPfoSIG2phI+93OkacaQCpDHNts+bvdgfJEGfaqW8rVEjHVURH0tgV8Ca\nLKkAacxO2Y06QBIn5TZiAdJRlfGxz7+o7pcOkMYDp6wCpDG/PkKy4bh0+PiQlIB0QPYKWRsA\n9kpMv8UAqYG0AWCvxPRbDJAaSBsAg37Zs/hfkAzIXiGXAqDQcSDoWwsgQVABASQIKiCABEEF\nBJAgqIAAEgQVEECCoAICSBBUQAAJggoIIEFQAQEkCCoggARBBQSQIKiAABIEFRBAgqACOgTS\nxhSOnjNKMmXPYsxHaqAjAGyxcWbbhZ3012z2Jm9ihmwDAaSjsudjgNRAAClbaTEpez5eBKnh\namAntF7IWu0GSLmK6TmOjwHSfDZKtVrIau3usmbDRvtM7WUZ55eHAOlyNkq1Vsh67day+Enp\n5EdxAaSWAkjZG1urkRDatRRCu9xth5WdFF+WSDa0E5INudsaBCnpQUDSLXuFDJCOyp6PAVID\ndUt/X2g6f5w+sudjgNRASH8flT0fA6QG0pL+Ln2cerLnY4DUQNoAsFdi+i0GSA2kDQB7Jabf\nYoDUQNoAsFdi+i0GSA2kDQB7JabfYoDUQNoAsFdi+i0GSA2kDQB7JabfYoDUQNoAsFdi+i0G\nSA2kDQB7JabfYoDUQNoAsFdi+i0GSA2kDQB7JabfYoDUQNoAsFdi+i0GSA2kDQB7JabfYoDU\nQNoAsFdi+i0GSA2kDQB7JabfYoDUQNoAsFdi+i0GSA2kDQB7JabfYoDUQNoAsFdi+i0GSA2k\nDQB7JabfYoDUQNoAsFdi+i0GSA3UBQAsftJWAKm+eoA0fPnl3HH6yJ6PAVIDAaSjsudjgNRA\nAOmo7PkYIDUQQDoqez4GSA2EZMNR2fMxQGogLWlr9Q8au8iejwFSA3UEaXEXeyWm32KA1EBd\n2kgb1Y+9EtNvMUBqoE5tpNVd7JWYfosBUgN1Cu0IJYDUSgCpvrq1kVbyCvZKTL/FAKmBkGw4\nKns+BkgNpCX9Xfo49WTPxwCpgbQBYK/E9FsMkBpIGwD2Sky/xQCpgbQBYK/E9FsMkBpIGwD2\nSky/xQCpgbQBYK/E9FsMkBpIGwD2Sky/xQCpgbQBYK/E9FsMkBpIGwD2Sky/xQCpgbQBYK/E\n9FsMkBpIGwD2Sky/xQCpgbQBYK/E9FsMkBpIGwD2Sky/xQCpgbQBYK/E9FsMkBpIGwD2Sky/\nxQCpgbQBYK/E9FsMkBpIGwD2Sky/xQCpgbQBYK/E9FsMkBpIGwD2Sky/xQCpgbQBYK/E9FsM\nkBpIGwD2Sky/xQCpgboAgEX02wog1VcPkIYvv5w7Th/Z8zFAaiCAdFT2fAyQGgggHZU9HwOk\nBgJIR2XPxwCpgZBsOCp7PgZIDaQlbf0dn9h3Oefnl7+XP769PE9/+fk6v/+g988v76e/5auJ\n+ssYIGVuPKw+IcleiRUAadJv+dPf5/kPT2/8h7f5/dvprwFILdStjTQs72SvxO4BKf3y8XsY\nuE6asPn59jH98udF0Pk7DFQZvf9KG5z8os23GgWQ8rcdANJ119/Dr+n1zzBcQ7ofw0TQr/kP\nv3iDs1+0+VajANKBbQeAdPn1g39/vnJEVE3oPA2fNjj7RZtvNQogHdn2uz+M+fb86fe34cft\nxz9f1zY+/EWbbzUKIB3aGCDN+hiebuK4BX0Mz6e/ByA1kJb0d+nj1FMVkF6GF4rs1pPcv2mD\n01+0+VajAJKW49RTBZD+/hqePrajtyfa4PQXbb7VKICk5Tj1VKUfiXuNNkD6gX4k5dIGwPpx\nYiz0FXeqAkg///uQ92tbvtwT2OkCKc+NWkDKv+jMgBS9j6HQl9ylOlk70o+1NtKdHGkCKTrv\nMzZTAlIcs6wlWQEp+jEwSdlnVkv1QHr5nLX778/l7/dxpAikiaPRZXhQB0iRLrbM680ISAmi\n6d/sM6uleiD9+dyPNL/7dS9HekDyzocQM0jSAtJkbcy73myAFDiuo1qJ3vUlqR5I4/NtSuG3\nvHl/nge0npcWkKg2mjxpBSQ/krWPA9J0Nt45OiMXuPH3sCD9uRnk/So9sO9PT/fk69IXbb6t\npIUY3CeO9lvwXUD612If5MLL2lk/SJ7L300nFYIbp9NyOUF2PVUEaWoNyejv99fn4YkyDx9P\nT3fNREpftPm2jpZicEfX5eRHnTXSV4s9GZxhLUk9SCFMJ0O1kWMnxDidWVeSaoI01UOzfjBA\nv276mk5/Tw+Q2EPhc5qVoorJiS4jWOoAEhn12eDJ0sjs5+yuHSSuXSd4onBEbiBfdCSpKkjj\nx+vPp2F4+vU2b2IYpJDatEmBXCd+3N29D0hhiuZu/zLbm7O7cpDCVPyObwvMjwv0Ns7n1iUV\nrqIdfEx9QjuOia4XZiC/Uc7OZTQ6uoR2QWC6vCd76eVUz9dZ1QJpAidEaiJJRUQxXpz+Nnlk\nivmmn6nFRD8NBZDyJDf3G5DkDkgv+zt3KeTP5HPsk6qksD8UQDdIxBGRROm6IOG1hNn0npIq\n1ICa/CKuibHFMCKAlKvAP7NPJo+Rg6JTC1IymCwmeAh8+ZmusvT3dakGaeIocO6bGn0xBE6e\nUosp8AlO1ZOXOop9E+lM65MEkLJFN/I4ik8iB+XkN5cx0qtTIXPNM5nLlWfkJBdZTJdduJzJ\nsjSDNN3DCB4KB6bbA3M0n1fkuslH6WLi9HigJtOEWCFDsixceq9QHTtkY3ohD5I/maT93ToW\nsnA0MvOTvZS6o8ssxHRbWJFikCK3gqTFx9Wro2AuSnwQEkwc+/kU9Y28fSFLMixcfK9Q3UHi\ny5Ai9CDh3e5unUEKkhehn0RS4JDIJEjRC0aRct6TBxzHeNxGkmg7cFMwptiPECM3Ua8tpx9S\nA7F4Zg8gHRIHRJGas45f+fa3u1fPQo4jDQekpFZIBrPhPozjRlCqFiQaEOQpZ8f1TmSOKN9A\nFa7wxGdIaQfvpMUktxCKHgIFEnza5Qe5AqRjiswR9aKTP31QDxJzxKM7JQriq2oiiZsaqztp\nBYkqU06dpNxCFCdIZ3PgCshJG4prpZAqJKmSOMdHdxaBqCxJAOmgpivSsyd9SrTqnkZB4acP\nc+ObrjqGynP3sjmQYhpOEiQBLs0jagtxLRQl0+CkK4l7l6hConsIvWfwqGWYfAaQNt9WlwR2\n3GCXTov9XfoV8nQHZnKiDEijS2yOSt1WvlEnSH5OKgQhiWukkGI7SaFKAiLBFOWeESQ/wT0A\nIe2D0K47SCNXScmnPPh4d49uhczp4UAXlpd0fWoweL6aNkhSCZLYH4UZSqDQWxkf5CXBQKAI\nWhzcCTT0v4x4oF3oLY/VK2TZJws33itUb5Acp4okYUdu1QxSTAnGhJFLbQWe/bHVB9YFpJ3n\nI8V070od4VLDyrAgHrUhoZy0BoO0Y4Wo9Ae+d1CygRq1xdd5AEiHJXe15Fa/1WSf1auQKSEc\n5Grzco1xwphy3zyNx61eTz1AGr788ukzrmoYJOkL45sCdydJoMdpuhTAcvVDIS2NGqI3XCvR\nHYS7ocbPw3mLCN/RE3IAACAASURBVCAdFt3WwuzWNDtzW50K2Y1cHcloTs/wR84cO75tUx+l\ndpBuJgtE6QPjakcGNjA7IclxlCCkcbOJRz9Itjx4Gc4aUnjry3MEkE4o3MToPHRyb4d+IKUK\niXtgpVc2JY95xOd6gKMFpJs/ydifNICBK1X5lUeCzyTJzc3zgIcotPFnUkFR8pvuexlBxGEB\npMMK7CuplbjjTytI0jpPg2i4qcDNBCf9mU5bP9IOSGPKkFAnLGcRUiOJf+iFM6mXztkLPBzl\ncbaSQrxqa+ABpMPiTvOU+Eot9509+oHkJJiTrFVM+SsegiZ9KmvSmGwYeXoEd4xJfMZxXqqN\nJHyV4UOCV0jxbODOJu61LZ9i+MfCjfcK1bsfiX1CXpNbOwV6O7t0KmTu7Q+pdSAWzxcYRT5b\nV5XK9PcoCzRwUiFNleDgWoI7TjdwasGlYZCB20cu9ZzRhOGKU2cB0lFxN5KTYMJxkLQbL/QC\nafQhXu7Yc6M8pMFqm0brBGmURInj0QqS6w4S3DmJ8KK0hIK0BqMk9GT5JJk4W08A6bCidMEk\nF7ITd/boFtqlbhZ3aUnIzZvn9loEafQytttL50MQkFxIZ8mpca5/QhoMLifNU9J93aUcANJR\n8eQeIWiOmfZ26Za1u9RCYe665IY3N502Z41qBWmcF4XkoT7S+xquzSRuM7k0IIhrJer1SymI\nQqbsWbjyXqH6ghRlASEOjySo0NqPRC23uTqaOZJxAGlck8kZspKI404xHqbqL0G2c2kQ3vzC\nHIU0obb2FFmAdFCRFhJJgz9ZGdOY+4E037BTIynNRUzpK5MgjX5eqIE7WHlqlU/NJGo4eRno\nENMyTzyDDIufLKp7aCd3Q4kYprhOK0jUIcvR3Qw89yXJqM2wvWSDZpC4keqDdCnz6GGXVhWS\neRJEGfch0XxYroiq10ZfLFx4r1Ddkw0yQCjKjT6Do37JBp7GlhaBk5lJTrom9x46phokroQu\nw9ilRSSDoXg0ITecHD8uwHMA0WANIYCUq5vrjgcHBRlR7LPWp+kG0ihjPYNMM5dFuej9xi5y\nOqpB4gdRyMCf6NLsJA7xeCiDG6nnleeTU5outuEIIOXpNhJy1xjd500P61bIURbdmTtUJOG1\nZfJ8nrpBGqXweZae44eNycRffmDSvyusNnrILEDKUby8kHg5Qp61HZWDNKaVTlIDXBZ/2+iX\nvJyncpAkLEgTKoijUW5tdJPo9HRmgJSjW5C8PPkghRNZu3cs5DCmtSVCWu6AGxVrq1HZAWmk\nuE7WKqZTG9NCkWFsvOT3RQApS9fQjobyy9wXlz2tpStIYxqv6mTpBl7OwI8rD6MxEtqRaKyQ\nLE08cmvJy7SQy5LfjQWQ8hQvHI289qDAlLlzx0KmDB1PjWeSpOvfp2TWkkwkG0TpuZfT2biQ\n5lQkiHqQBJCOyfOaUDxQJX8QZL9CpmWveSlIJomXEpouvP3eFQsgjTICPK1+meABSPnqClKk\n9oX0BKYnp2SoWyHL48Z4VS4vc3Jo5HpaaNQ8SKMsRCEzQuboAKFdtjoPEUpP056v0Rz1BCk9\naZDiIC8PGKTrzO/0rtgAie8OngeyXyftIdmQq95DhHhAF41SyV5jsGdoN14W3gsS2M1P39rc\nzwhIo6wvyCdVcfZrjgDSYc151vz5LT2TDVcjg9y/s3azAhIvS+H/fb5vDwGkw/IhY72TT1JS\nyDQYbS3v/Y/MgJTGAFWd/JolJT4+os4gzWsMHpCSQqYxnmMeSaZAijJrudBXnZQSHx9Rb5CO\nj8zvXsjzw1S9vOzLEEh8Vt7Fm2RdjXXr9tTdx8fVPbTjWmnpBrjmwJKFfOYikZSw3LgVh3Z7\ny3GtidcvHq+LOW0/Q62SANJxSe74K0mrDixYyGcukrmTki42xcmG4csv2ceR55Ynj4TLS0sB\npMOKkjv+AtK6A8sV8qmL5NLbL9dbjoyBlCpdOTmAlKne/UgjPeZhuULSCdLc2/95MsimzIEU\nnbucG0K7PCkAabn7XGtoN5v7yCB9mnqJZEOW+od2a1ej1mTDrJ0VT25kKtkg6jShbxZAOq7D\nLlNTyNmWW0p/65AaH+erO0iHZa+Qu9VIKzvZKzH9FgOkBurWRho+7zRshHu6ZM/HAKmB+iUb\nBtRIrQSQ6qtj1m4ASI0EkOqrZ/p7KY6zV2L6LQZIDdQn2bC+k70S028xQGogbWlreyWm32KA\n1ED9AFjew16J6bcYIDUQQDoqez4GSA0EkI7Kno8BUgNpa9vYKzH9FgOkBtIGgL0S028xQGog\nbQDYKzH9FgOkBtIGgL0S028xQGogbQDYKzH9FgOkBtIGwH3HabEcuD0fWwIpeXCxkDst9p4n\nHQAUOk6TB1QApIqaPbhUyL0eP5InFQAUOk6bRyYBpHq6eHChkLs9ECtPGgAodRyAtCKAVF8a\nACh2HIR2yzIDEkI7JcdBsmFRdkBCskHbceoJIDWQvULWBoC9EtNvMUBqIG0A2Csx/RYDpAbS\nBsCgX/Ys/hckA7JXyKUAKHQcCPrWAkgQVEAACYIKCCBBUAEBJAgqIIAEQQUEkCCogAASBBUQ\nQIKgAgJIEFRAAAmCCgggQVABASQIKiCABEEFhGkUmmXdYgsmlwJA2XHqyd6cM0zsayBtANgr\nMf0WA6QG0gaA0VWEVC9w8yggqS5kHQAUOk63de10L7n2ICDpLmQVABQ6TreVVpUvAvoYICkv\nZA0AlDoOQFoRQKovDQAUOw5Cu2U9BEjKC1kFAMWOg2TDoh4DJN2FrAOA8sepJ3uZ2UcBSbW0\nAWCvxPRbDJAaSBsA9kpMv8UAqYG0AWCvxPRbDJAaSBsA9kpMv8UAqYG0AWCvxPRbDJAaaNHC\nEyPDAZJiAaT6WrJw2PrwwHHOyF6J6bcYIDUQQDoqez4GSA0EkI7Kno8BUgMBpKOy52OA1EC2\nkw0hFPraAzrg4x7mLel7g9TGC6XWbugCUuCfxsr3cRfzlvStQWrkBW0h2ZHjhMtLS2X7uI95\nS/rOILXyAkA6KoDUQI8B0pm1uhDafRVCu9N6kNDuhNlINiwIyYazMphsOPDX48c5Lv1OtpeZ\n/eYgtZE2AOyVmH6LAVIDaQPAXonptxggNZA2AOyVmH6LAVIDaQPAXonptxggNZA2ADaSyTpT\nYGfOPDQ+F4B0RCGc8Y8VkJyTs4uTPn/y7/vautPHk7lhOoemJAGkfNEFFlw4fFXZACk4UphO\nj+4W0y+0WKCca5SfdjTd5+NIJ+Gc/D+dRzjusuMCSNmK5Bm+0OKxakkzSN7zP4SQ99F7gogu\nxOCoWppImjYQmOivhQw4ZOHS+03FIJHDdELkLzed13Q2031inP4y/a2gnesmAqRVEUdhvs7Y\nNYFvdfuLvCoFyTk/Tv/FOF1vdPPm2zddeW76iYGhcnIxTifPv8TRE07Vmx93+Zg4IqM9u2lC\nJ/ItMFKF612sRBJAylSqj+ji4ru354su0F18zzM6QZowoutqql+jD8yRnFJCKcTpGgz0n2BF\nn4bphf6dR1ZVi5jO+3iqQINwJGeSbgd85yOO6ESqkASQMuXmCilIY8IJUXSDZs9s3KVVgsR1\nDV1nju/a1JyIwpGTH3rLjQxmituGXB1Husfz6UrTqYZO+3gKVD2bL/cDuvlRZMetW656p9MA\nSKI+IDkJ5eRGxwC5dOeLFCBtjn9VCRLdqqOkT6jl4PlKc4mjRFJ0USJZ3piuSKmKp//n+K4O\nSad8zJXRKLRcSfLTXUBI8hzdBYR2SZ1AGq/wuDl2EJACeZC2WSNJI0jccEjy3JDghLHcy+Uy\nDBTtOSHJ0fXIbUMv7Xhu0Y8E0vRPyliU0xkfk2Ec16X7nEs3iime83KTIJKmure0sYsmAqQ1\nUTvB3YDEgQ6FeJH9R5sYAolSDCneYZT43YyRk2YghXEc2FEFRGkv/ktIxUAXJ+VaqHrikKqk\nTvjYs2kp6+jSLY5Pb7odpIT49NtY3NZlEwHSqrgtnqolapuHKM0kbs9aC+0owXABiajynLK7\nnB5FckxKyk66kKI84ojrKUFpuruMFUg6A9LE9JiyI9LY4zYtnaNnZ1FoR1VnHY4AUr7iTdTA\nv8eUeaCw3FiygSOdFNk5yjx6d01L8mlxLm/uNaOwjnDjez5XUJJ3oBEEI9fH3UGaiHaSR/Wp\nJcvJfLlHzHcMXz4KXTERIG0oSvtBulmSpzhgoBbExn4aVxGSeihxRBebj5JrSG0/Tzfy1B0T\nHLeMKD/OGXFpPTFaI9VGwfcM7eZ+PLoZpAx9Oo0oOfy5Ypr+WtbKLRMB0oakoS39llE4ks5L\njvrW91NYI/FYNM7COY59JpI4WTwnwL1kFfiEpSdm+lu4Rk6cykt9UJTJK2TZrYVb72/kuGeC\nnEOnkqqjIM28GCXJH+SOV7UbGSDli0dt8e1arj/puaQ8EEd7q/tpBImTBKlHzMVLkJe6x1xM\n44X4LH2qo6hT1l+SDRIZjlWCpXwf0/1Luo69mMhZEn+5IXBnWWT66w7HAEgHRCRRqMBNBr5H\nM1Re2ueru6kEaUyJLO5g5XZelL7L1M3MA4dSdjL6uauJAzoeKuRrrsF0ACSCmTiiDIjjni/J\nMnonI5/onkDRax1DV00ESJuSO/GcqqNbNP8igd7qXjpBorOhDiSKzKLUSpwDTxlxqY3kI0KO\nP5WmfLoqK67BlO/jKONqecQto8O1D3FEDTnJMIQGHAGkg4oySDrGdOWFFNfFjUaSWpAiB0GR\nB/3w+Ia5X4xTdJ77mrmSkiFEXga4XmKkesFSto+jjL1ltwSuebh2lQBU4lK6V8RaubpVEwHS\njiL140v2KkpL20mP0sZFpRWkUUax892arjYe4e2pw8yPadIE9ZDxxSkDVKslj9csXHl/UbKd\nsofcaKMRxJHnTkXu6JIhgl8mKlYQQDoknndJV59UTSlu4ETr+k5qQaLgbh55StOOJGadLsvp\nwovzFs2nx46HQjv64QYcWU+dWgQ9OyrVmNzeqzW6dtVEgLSpmCa4kVskGOIu9Z05b3pBIuPH\nK0l8vdFMvha38C0d8PGlpox0Mv7mj9dN6K+1TwggHZBcdXN4wz3lo5C06SbFII1XkChC6ovP\nVad8PPvj6weXl4oCSAe04hHvt72kGaSKk4ruUGEftzhHgHRE5zyiGqQeTaBdlfYxkg0L6pxs\nOLGTbpBGPcvZXVTFx3XPEiBl6g43aAdJzTOGLqrh48pnCZDydI8bFi3U8zBmOi8a8KMoxjvv\n49WTmM+ylgBSlgLlgw67IXl1ycLhyy/7qgcSjbHRlHU47eP1k0hneYdR2wJIOZpuZf44SLNX\nlYM08lT5Jp0tuTrr461Ed6gwAfFGAClDDNHhuODiVe0gydpajw6StHIB0kV9QDqRbLAD0igD\nNNRwVCO0G+Usz1q0K4CUo3Mu2ArtFCUbSDzGRg1HNZINpJpjbgFSls65YCPZcEbajlNPXfsK\nzwkg1Zc2AOyVmH6LAVIDaVxFSLfs+RggNdBisuEEXQBJsQBSfS0nGwod54TslZh+iwFSAy1b\neNxugKRYAKm+tAFgr8T0WwyQGkgbAPZKTL/FAKmB1i08ZjtAUiyAVF8A6ajs+RggNRBAOip7\nPgZIDaQNAHslpt/igiBdehifX/5e/vj28jz95efr/P7j5Wl4+vV38QC537PzXp+0AWCvxPRb\nXAOkSb/lT3+f5z88vfEf3tLb/85/jcFC1gaAvRLTb3FRkNIvH7+HgeucCZufbx/TL39ehoFI\n+nga/pvevw7Dnzu+Z+e9PmkDwF6J6be4Bkjj+Hv4Nb3+GYZrSPdjeB/Hl1QVvQ4/7vienff6\npA0AeyWm3+I6IH3w789XjoiqX/SXjy/bHv+enff6pA0AeyWm3+I6IPHvb5+rnZ+vK9se/p6d\n9/qkDQB7Jabf4lo10tM4/rqtkD7rlUO/s9+z816ftAFgr8T0W1wHpJfhheK49+UN31+Gp487\nvmfnvT5pA8Beiem3uAZIf38xKCvh229Kht/BkcFC1gaAvRLTb3GdfiTuNVoD6efPp+EZ6e8H\nOE492fNxDZB+/vch71c3fR3Wwr6c79l5r0/aALBXYvotrpS1I/3YgOX3HdkGe4WMxU+Oyp6P\nK4L08jlr99+frY2PfM/Oe33SBoC9EtNvcUWQ/nzuR/rnHUCyf5x6sufjiiCNz8Pb9c1vevM0\nj2x4H57Pf8/Oe33SBoC9EtNvcU2Q/gxXkl6ZHO5fIv2ax4ef+Z6d9/qkDQB7Jabf4pogTdzI\n6O/31+fhiTIP78Pwm97/opEPp79n570+aQPAXonpt7gqSJTmTvohGby3pzT373z222AhawPA\nXonpt7guSOPH688JnadflxDvg2bM/jgf140WC1kbAPZKTL/FBUFqJXuFrA0AeyWm32KA1EDa\nALBXYvotBkgNpA0AeyWm32KA1EDaALBXYvotBkgNpA0AeyWm32KA1EDaALBXYvotBkgNtGih\nrqeaK5M9HwOkBlqycNj68MBxzsheiem3GCA1EEA6Kns+BkgNBJCOyp6PAVIDAaSjsudjgNRA\nSDYclT0fA6QG0gaAvRLTbzFAaqC1GmlEjbQiez4GSA20tIrQkP5HG2lJ9nwMkBpoLdkAkNZk\nz8cAqYEA0lHZ8zFAaiCkv4/Kno8BUgMtJxs2PjtynBOyV2L6LQZIDaQNAHslpt9igNRA6xYe\nsx0gKRZAqi+AdFT2fAyQGgggHZU9HwOkBtIGgL0S028xQGogbQDYKzH9FgOkBtIGgL0S028x\nQGogbQDYKzH9FgOkBtIGgL0S028xQGogbQDYKzH9FgOkBtIGgL0S028xQGogbQDYKzH9FgOk\nBtIGgL0S028xQGogbQDYKzH9FgOkBtIGgL0S028xQGogbQDYKzH9FgOkBlpa/OTUcYpYY6LE\ndt4rFECqL20A2Csx/RYDpAbSBoC9EtNvMUBqIG0A2Csx/RYDpAbSBoC9EtNvMUBqIG0A2Csx\n/RYDpAbSBoC9EtNvMUBqIG0A2Csx/RYDpAbSBoC9EtNvMUBqIG0A2Csx/RYDpAbSBoC9EtNv\nMUBqIG0A2Csx/RYDpAbSBoC9EtNvMUBqIG0A2Csx/RYDpAZatBBPNd+QPR8DpAZasnDY+vDA\ncc7IXonptxggNRBAOip7PgZIDQSQjsqejwFSAwGko7LnY4DUQEg2HJU9HwOkBtIGgL0S028x\nQGqg1dAONdKK7PkYIDXQ0ipCwzjDdOA4pewpdJx6sudjgNRAazUSQFqTPR8DpAYCSEdlz8cA\nqYEWQZoCPYD0RTHyP5Z8vGyyZouT1gs5nZI6rRQqN5qQbLhV5B9TIK2YrNjiWauFPJ+SOmkD\nQK2T4/xiB6Q1k/VafNFaIV9OSZ3WC/VYcQMkfQJIDQWQcoXQrqEeKbQDSP8IyYZ2ephkg/nj\n1JMlkJIeCSSt0gaAvRLTbzFAaiBtANgrMf0WA6QG0gbAoF/2LP4XJAOyV8ilACh0HAj61gJI\nEFRAAAmCCgggQVABASQIKiCABEEFBJAgqIAAEgQVEECCoAICSBBUQAAJggoIIEFQAQEkCCog\ngARBBQSQIKiAMB9Js6xbbMHkUgAoO0492Zu8iRmyDaQNAHslpt9igNRAXQDYqA1XjqNoEabz\nPu52El1AuutsAdKhbRd2Wj6OpmUBT/u430n0AOm+swVIh7bNBUnVQrVnfdzxJDqAdOfZAqRD\n2wKkNgJI9WUCJIR29wmhXX0h2XBUSDbkCcmGLscp3a9VT/Z8jPR3A2kBqfRx6smejwFSA2kD\nwF6J6bcYIDVQl2TDRhxnr8T0WwyQGqhPsqHQcbrIno8BUgP1CcnWt7ZXYvotBkgNpK1tY6/E\n9FsMkBpIGwD2Sky/xQCpgbQBYK/E9FsMkBpIGwD2Sky/xQCpgbQBYK/E9FsMkBpIGwD2Sky/\nxQCpgbQBYK/E9FsMkBpIGwD2Sky/xQCpgbQBcOI4IRT67jyV8nFDs/WAlHvSAKn9cQL/tFMh\nH7c0Ww1I2ScNkFoc53bKmA/OXb0TG9zmy/g4jN65QHfoBtP9+oM0nWuI4qYsDwGk6seJMU1i\nDpMmBzk/utk5MdBPZd3rYwEnTDcAIolRKmPYunqDFCeM+IfeAKS+x0k37si/RIIo+DHE6WIk\nkuSzML/U1CkfX6sduQ0E78J0Ei6GFiR1Bok4isF5JgmhXfvjeP/p1yht1alKIo6IJPIN3+1k\nI70gxcvJpOoo+omh6eKic5n+K2ziv+oLUvR0inTjmO4eSDa0P47nH/k10q/TbT06qpKCp+tw\nco7n4I444utUa2hHHE3XU/p1FIKmK8v7eGk61FRXkCavUc3ryGku91QBUrnj+MsLXYXUNo8+\n8gU43dk8/TM5Z3o7tZKcQMfXZSE7cixcfr8guQ2MiSSqjhzdpj29eu8fO7TzdOdzdMrTjYNp\nytgJIJU7zhUkbgg5uu4cRUJTPTRdhBzfEUbTxeg85YU8u+vT0WpwdSq081IrsUnTuTgKcqZT\noDopP945rb4gUQwx3fo8kTS56NKk3RJAKngc/5kjiuE4qqMwgVrqchVG+tfT5TldkUyS41dq\n3lPb9qahVdzCxfeLivPJUI7ESZhDt+fAWYfsiOek+raRPHEUPN8/AtfA+yQBpJLHYUCm4E2A\nSSDRXZzfTXf04IUt+it5yPHL5CgviSI+wliYpJM+9nQFUdabKlPPZ8NnF6RSKmvjtomtr0p2\nCt8/HPsPIDU9DvHDKeLpqvPcrKCOF7qLe4mK6F0Ick9PVM13PW5KUX6cQCpM0lkfE9JB8nSc\ntQ+Omnhkd2QrK6p3PxJVwBQvOIolppPf3wMglToOXWJcIzl2Ard/xBNcQcVLEpzfOArz+Nqk\nCow/cJP76NL1vUHi2JL6jcnYwEEpnwil7ukvkaAvHoBumdj8qqSbH1dHfO/zGbkVgFToOJHa\nOYQLc+Tl/s2DAdJtXKChK5O2C1J5cf85JyNGuj6l5dQ5tPMcy3BwGlNrLqRTCSkqpdSjr4lS\nd5DkfKOcdU63GUAqcxypiFj0L7eB3AUl7n0J0sfH2e/0B7lKqceC7vdRWvg9kw0xUs5q5Fad\n56aRBKt8Y5YmQ+CMvjSZpq0K27psYoc2UuQMUUjnvr8LQCpyHKlMOFXnua+IAzl3cytntIJE\ncamPghtTc1ctf0BXcCGz/rVw4/1VzBBlD73Eo3Q9yUkETjNwooHxEpAow1/e3q8mNgeJwvTA\n7UMhad8rAKnEcTjYifI6x23iA8kuOD87hesh7pON0oSigI/DcL7rVenozPaxPGvLM0xSsXKe\njk9CTo9DOr4BJJQqkdR7rF2U3jLyZ+BW7u4uAClv483nI6Wb9EyS4xwdpxnm9sV09XFcxxwx\nWJxgnW/0dK3We6zXIZCkXcTpbh8l6eiSmUEqIuldlvR+eFCQOLbggVGSagFIZY4zfPnl02cU\npnEsJyhJrvuSS+DbOafzOJ/AtzjuWeKUGGeFRuqoiLXm+RwL7egy4jYSpU5cyiimeklyKTxY\ng0docL9zA5M7pL9Dur/JCK/9PQDSoW1vdrp5QAWP55xBovu5xEHSRuLozfOoupRw4M7yNCfB\n83CH6XCu3nMmjyUbZpLYZsc9/HQWUUZ+x3n0LYPkK3HUHaRRegO5SqIm7f4OAOnQtss1knDj\nL6/UjIgpupMePZ8qqSiBAv/KyWTplJkh6hzaiXiokgxQnZt4EttdOaK5IDRksNpk2e4gpTPl\nu6HL6X0GSIe2XQaJSbrlyPNQBj93JQVOf1HlcwmS5kGsdNOjKsnpAYnFncozR477leNlcEbq\nG/s0/6qseoMkTd2UUXE+Y6oLQMrbeOdhzKkXNnI8wMmswH9wqfeFR00LWT7lGxwP0ZdJCRQ7\nqAjtxstgdCfBqPRM0m2ZQj2fzo1zDnWMXTSxebJhpP6+a1d0RkcSQCpzHArjnHRVCkNhHuAg\nLQweyhoSWqmTM43Ckx6kio8QP+BjN87jMydQZJzFnHrk5hL3wvKwwKocdQeJx0E66Z+Ql71d\nAFKZ43AyLrXGvaS3PY/unkc3BGlxRJnWw3kHx4s4SLVUcwxovo/d5YWrJLklz6Od0qgZTo3X\nHWnXHSQnc1+cm5NF+4OEAFKh43C8wzP9eWy0REI+zPkGx30wadBqSgcRXGEeXleRpFMgcefX\ndXTQPARKBn/nzHS7S31B8jzFhXuj07DjuDsAHCCVOk5aLoM7iWSKMvcduTnnJTMR+ENOhLt5\nRgWP9665DMKZ0I4WZmBj53wJv/KaDTSG6KFBkr7A1BMt5x4DQGp2HO76l4qJe4fmMQFzIylI\nPcXdMLx0Q5Dpl5RGHnP6/O62cOX9rdyFo5FnTPjLMKf0iwznrGbrsolNr8ooHYPShZ5q4/1G\nEkAqdxwamhClq1WCtjS4W4LskEbbeBnT6qRDVmDK6jsvYOHy+wWR/WO4NpJklNM8pMk98uIn\n0g8xN2Ov/Wg7uwGkkseJiabIt3Oedz5zxHN8LnNO3dzHKWsgaEk2XCVNITkHN3cryyqXnGGs\nYOaGie1BknVxJS6XgUI7uwGkose5dAZ5ifOCLL/FTSRuX3hax47u6MSThOM0xVxLG2mWzI4d\ng9wT0jxRDklDxbFMqyY2D+1GXvQlpCn2NAULNVLb48ydQV7SdWOa0cepBqaFV1d01FznTTmE\nIidpA2mcF/mWkZsxDV2PoWKH11Wdkw38D0cPacDu/ikDpOLHkXCaJ5vzLT3KaBP5jDiawgS5\nrc+rGld+YMqp0E5gijz8iVORTkbdzhvUXdquF0ifzorXlPWpf3BvT4BU+jgzFd6n5gSjNEdD\n9CcKtx2vZuxv/lpPp3w8r/zthHWfBrTPH7dlv9VV+eWsuA3LZ763K0AqfBxyxOfUwS1HrLnb\nMyNgKKK7fBxljgefxLX76OtZllUXkPzCQ1zYdzkzxQBS4ePwqnT/ZrP/cQRfklkrShfRfT52\nadbheGvx8vT45AAACEtJREFU4lkWVA+Q0gn9e3vgVT739wZIpY/Dya6dMZ3tICLd6eNFY3PO\n8g51AIkh8gsBa56zAFLx43CDp/KgzkOq4uO6Z9kJpDsaqwCp/HGKL5V6p+r4uOpZ9gnt7jkh\ngFThOLUnGRxUJR8/3Eqrd50QQNJynHqy5+O+HbKnZK+QtQFgr8T0WwyQGkgbAPZKTL/FAKmB\nugCwt/iJatnzMUBqoB4gDV9+OXecPrLnY4DUQADpqOz5GCA1EEA6Kns+BkgNBJCOyp6PAVID\nIdlwVPZ8DJAaSFva2l6J6bcYIDWQNgDslZh+i781SJfnBT2//L388e3lefrLz9dPW/5ZDJGy\nv+eOfUseZ9gI93QJIDVQBZAm/ZY//X2e//D0drPlszmQhg1q9DsZIDVQSZDSLx+/h4HrpLep\nLnr7mH758zIMV5J+33cX75NsKHScLgJIDVQBJELl18gR3CWk+/gxvKdf34cneyBtbK3fyQCp\ngaqA9MG/Pw83TaM/zBbp5/BqEKT6x6kngNRAVUDi39+GH7cfzwkH+jNAaiqA1ECVaqSncfw1\nvC5t9jTFeEZBWt5Dv5MBUgNVAelleKHI7n1hq5fhvxEgtRVAaqAKIP39NTx9rNDyl+oqgNRW\nAKmBqvQjca/RIi0/hj9rH2V/zx37aj5OPQGkBqoA0s//PuT9121+Dz/XPsr/njv21XycegJI\nDVQna0f68bWN9PE0rDKW/z137Kv5OPUEkBqoHkgvn7N2//2hRN7vxU2Pfc8d+2o+Tj0BpAaq\nB9Kfz/1I9O52ON757zm9p+7j1BNAaqB6II3PN+PrptbRG0DqJIDUQBVB+nMzUvV1eN7a9ND3\n3LGv5uPUE0BqoIogTa0kGf39/vo8PL1vbnrke+7YV/Nx6gkgNVBNkKZ6aNaP951ND3zPHftq\nPk49AaQGqgrS+PH682kYnn697W+a/z137Kv5OPUEkBrIXiFrA8Beiem3GCA1kDYA7JWYfosB\nUgNpA8Beiem3GCA1kDYA7JWYfosBUgNpA8Beiem3GCA1kDYAjh3n/GOzz+uAj3uYtyS9IK2V\nEEDK27jQ2t+Bfxor38ddzFuSWpBWSwggHdp2YacjxwmXl5bK9nEf85akFaT1EgJIh7YFSG0E\nkOrLMkgI7TKlFSSEdvcdpxhISDbkSS1ISDbcdxw8aKyt9IK0JnuFbDv93UP2fAyQGqhbjbSy\nk70S028xQGqgbm2k4fNOeNBYRQGk+uqXbBhQI7USQKqvjlm7ASA1EkCqr57pbzz6spEAUn31\nSTas72SvxPRbDJAaSFva2l6J6bcYIDVQPwCW97BXYvotBkgNBJCOyp6PAVIDAaSjsudjgNRA\n2to29kpMv8UAqYG0AWCvxPRbDJAaSBsA9kpMv8UAqYG0AWCvxPRbDJAaSBsA9kpMv8UAqYG0\nAWCvxPRbDJAaSBsA9kpMv8UAqYG0AWCvxPRbDJAaSBsA9kpMv8UAqYG0AWCvxPRbDJAaSBsA\n9kpMv8UAqYG0AWCvxPRbDJAaSBsA9kpMv8UAqYG0AWCvxPRbDJAaSBsA9kpMv8UAqYG0AWCv\nxPRbDJAaqAsAWPu7rQBSffUAafjyy7nj9JE9HwOkBgJIR2XPxwCpgQDSUdnzMUBqIIB0VPZ8\nDJAaCMmGo7LnY4DUQNrS1vZKTL/FAKmBetVII2qkdgJI9dWrjTT8sxMeNFZRAKm+tIB05jh9\nZM/HAKmBANJR2fMxQGogpL+Pyp6PAVID9Uk2rO9jr8T0WwyQGkhb2tpeiem3GCA1UD8Alvew\nV2L6LQZIDQSQjsqejwFSAwGkbMXI/1jy8bLJw81HSmWpkEXa2jZ6Syzyjykfr5g83HykVIYK\nOUkbAGpLLM4vdny8ZvJw85FS2SnkWdoAUFtiAKml7BTyLG0A6C0xhHYNZaiQk7QBcN9xnCtk\nxpK2kg1Vv/i87CUbUkECpK7HcfxTWUs+bvLF52Un/T0XJEDqeRx3eampBR+3+eLzMgPSpSAB\nUs/jAKQVAaT60gBAseMgtFuWGZAQ2ik5Tos2v6VkQ5IdkJBs0HacerLnY0sgJdkrZG0A2Csx\n/RYDpAbSBoC9EtNvMUBqIG0ADPplz+J/QTIge4VcCoBCx8k/XJFNynyPZS2f3oG/1tn02wog\nGRVA0iWAZFQASZcAklEBJF0CSEYFkHQJIBkVQNIlgGRUAEmXAJJRASRdQmlAUAEBJAgqIIAE\nQQUEkCCogAASBBUQQIKgAgJIEFRAAAmCCgggQVABASQIKiCABEEFVBKkvRnw6fOdzXY2yThI\n3vcYVlprIOsEDxXGcLNHmQ2/jQqWxLBzvPT5zmZ8haxvknGQvO+xrOH6uneChwpDkMvYNHvD\n76OGIKWNCoC0/V0A6ev2B3yTsWn2ht9I2kDizwHSpjLK4N8dMrfN5gMg/St7IA07ddYcuz+u\nn4dDJ3ho2xEgnVRjkHYZ2NskReeokY6cIGqk+moL0m7574MkL98WpEuiLOMED2ybNgVIp9UU\npGFvs7SILEDaF2okXWoJ0pC1GWqkPR05wWOFAZBOq2RB7PW05vUjDttHQofssRNEh2wboSQg\nqIAAEgQVEECCoAICSBBUQAAJggoIIEFQAQEkCCoggARBBQSQIKiAABIEFRBAgqACAkgQVEAA\nCYIKCCBBUAEBJAgqIIAEQQUEkCCogAASBBUQQIKgAgJIEFRAAAmCCgggQVABASQIKiCABEEF\nBJAgqIAAEgQVEECCoAICSBBUQAAJggoIIEFQAQEkCCoggARBBQSQIKiAABIEFRBAgqACAkgQ\nVEAACYIKCCBBUAEBJAgqIIAEQQUEkCCogAASBBXQ/0c0zIVmmssdAAAAAElFTkSuQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "## plot the principal components, use the `pairs` function.(note that it might take some time depending on the size of the data)\n",
    "pairs(mypca$x, col=colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mydata.loess <- normalize.AffyBatch.loess(mydata)\n",
    "mydata.loess\n",
    "boxplot(mydata.loess)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The PCA computation via the `prcomp` function performs the principal component analysis on the data matrix. It returns the principal components, their standard deviations (the square roots of Eigen values), and the rotation (containing the Eigen vectors)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 9. Finding the differentially expressed genes(查找差异表达的基因)\n",
    "\n",
    "At the genome level, we know cells have similar genes. But what makes them different? As we all known, only a fraction of a genome is expressed in each cell, and this phenomenon of selective expression of genes based on cell types is the baseline behind the concept of differential gene expression. Thus, it is important to find which genes show differential expression in a particular cell. This will be performed by comparing the cell under study with a reference, usually called control. We will present how to find the DE genes for a cell based on the expression levels of the control and treatment cells.<br>\n",
    "在基因组水平上，我们知道细胞具有相似的基因。但是因为什么才使得他们有所不同？众所周知，每个细胞只有一小部分基因组表达，这种基于细胞类型选择性表达基因的现象是差异基因表达概念的基础。因此，找出哪些基因在特定细胞中表现出差异表达很重要。这将通过比较实验组细胞和对照组细胞来进行参考。我们将根据对照组和实验组细胞的表达水平来演示如何找到细胞的DE基因。<br>\n",
    "\n",
    "1、The normalized expression data for treatment and control samples. Take the `quantile` normalized data as an example.实验组和对照组样本的标准化表达数据.以`quantile`标准化数据为例.<br>\n",
    "2、The experiment and phenotype details, which are part of the `affyBatch` or `ExpressionSet` object.需要`affyBatch`或`ExpressionSet`对象的实验和表型细节部分.<br>\n",
    "4、The R library, `limma`, that houses one of the most popular methods in R for differential gene expression analysis.R 库,limma是R中用于差异基因表达分析的最流行的方法之一.<br>\n",
    "5、For demonstration, use normal colon cancer preprocessed `affy` data from the `antiProfilesData` package.为了演示,使用来自`antiProfilesData`软件包的正常结肠癌的预处理`affy`数据.<br>\n",
    "\n",
    "The `limma` stands for linear models for microarray dat, which is used for analyzing gene expression microarray data, especially the use of linear models for analyzing gene expression data. It implements several methods of linear modeling for microarray data that can be used to identify DE genes. At first, it fits a linear model for each gene in the data given a set of arrays. Thereafter, it uses an empirical Bayes method to assess differential expression. This computes the statistical test and corresponding score in the form of p-values, log fold change, and so on.<br>\n",
    "“limma”代表微阵列数据的线性模型，用于分析基因表达微阵列数据，尤其是使用线性模型分析基因表达数据。它实现了可用于识别DE基因的微阵列数据的多种线性建模方法。首先，它对给定一组数组的数据中的每个基因进行线性模型的拟合。此后，它使用经验贝叶斯方法来评估差异表达。这以p值，log倍数变化等形式计算统计测试和相应分数。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Bioconductor version 3.9 (BiocManager 1.30.10), R 3.6.2 (2019-12-12)\n",
      "\n",
      "Installing package(s) 'limma', 'antiProfilesData'\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "package 'limma' successfully unpacked and MD5 sums checked\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"cannot remove prior installation of package 'limma'\"\n",
      "Warning message in file.copy(savedcopy, lib, recursive = TRUE):\n",
      "\"problem copying D:\\R-3.6.2\\library\\00LOCK\\limma\\libs\\x64\\limma.dll to D:\\R-3.6.2\\library\\limma\\libs\\x64\\limma.dll: Permission denied\"\n",
      "Warning message:\n",
      "\"restored 'limma'\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded binary packages are in\n",
      "\tC:\\Users\\Master\\AppData\\Local\\Temp\\RtmpSI9RIu\\downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing the source package 'antiProfilesData'\n",
      "\n",
      "\n",
      "Old packages: 'BH', 'bit', 'caTools', 'cli', 'FactoMineR', 'fansi', 'farver',\n",
      "  'GetoptLong', 'ggridges', 'gplots', 'hms', 'knitr', 'leaps', 'multcomp',\n",
      "  'mvtnorm', 'PMA', 'precrec', 'prettyunits', 'pROC', 'RCurl', 'RSQLite',\n",
      "  'Rttf2pt1', 'shinyjs', 'stringi', 'tinytex', 'xfun', 'XML', 'xts', 'zoo'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "BiocManager::install(c(\"limma\", \"antiProfilesData\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(affy)\n",
    "library(antiProfilesData)\n",
    "library(affyPLM)\n",
    "library(limma)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ExpressionSet (storageMode: lockedEnvironment)\n",
       "assayData: 5339 features, 16 samples \n",
       "  element names: exprs \n",
       "protocolData: none\n",
       "phenoData\n",
       "  sampleNames: GSM95473 GSM95474 ... GSM95488 (16 total)\n",
       "  varLabels: filename DB_ID ... Status (7 total)\n",
       "  varMetadata: labelDescription\n",
       "featureData: none\n",
       "experimentData: use 'experimentData(object)'\n",
       "Annotation: hgu133plus2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Get the data (the colon cancer data) from the antiProfilesData package \n",
    "## and subset the first 16 samples that represent the normal and tumor samples (eight each)\n",
    "data(apColonData)\n",
    "mydata <- apColonData[, sampleNames(apColonData)[1:16]]\n",
    "mydata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ExpressionSet (storageMode: lockedEnvironment)\n",
       "assayData: 5339 features, 16 samples \n",
       "  element names: exprs \n",
       "protocolData: none\n",
       "phenoData\n",
       "  sampleNames: GSM95473 GSM95474 ... GSM95488 (16 total)\n",
       "  varLabels: filename DB_ID ... Status (7 total)\n",
       "  varMetadata: labelDescription\n",
       "featureData: none\n",
       "experimentData: use 'experimentData(object)'\n",
       "Annotation: hgu133plus2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mydata_quantile <- normalize.ExpressionSet.quantiles(mydata)\n",
    "mydata_quantile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A matrix: 16 × 1 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>pData(mydata)$Status</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>0</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>0</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>0</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>0</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>0</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>0</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>0</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>0</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>1</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>1</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>1</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>1</td></tr>\n",
       "\t<tr><th scope=row>13</th><td>1</td></tr>\n",
       "\t<tr><th scope=row>14</th><td>1</td></tr>\n",
       "\t<tr><th scope=row>15</th><td>1</td></tr>\n",
       "\t<tr><th scope=row>16</th><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 16 × 1 of type dbl\n",
       "\\begin{tabular}{r|l}\n",
       "  & pData(mydata)\\$Status\\\\\n",
       "\\hline\n",
       "\t1 & 0\\\\\n",
       "\t2 & 0\\\\\n",
       "\t3 & 0\\\\\n",
       "\t4 & 0\\\\\n",
       "\t5 & 0\\\\\n",
       "\t6 & 0\\\\\n",
       "\t7 & 0\\\\\n",
       "\t8 & 0\\\\\n",
       "\t9 & 1\\\\\n",
       "\t10 & 1\\\\\n",
       "\t11 & 1\\\\\n",
       "\t12 & 1\\\\\n",
       "\t13 & 1\\\\\n",
       "\t14 & 1\\\\\n",
       "\t15 & 1\\\\\n",
       "\t16 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 16 × 1 of type dbl\n",
       "\n",
       "| <!--/--> | pData(mydata)$Status |\n",
       "|---|---|\n",
       "| 1 | 0 |\n",
       "| 2 | 0 |\n",
       "| 3 | 0 |\n",
       "| 4 | 0 |\n",
       "| 5 | 0 |\n",
       "| 6 | 0 |\n",
       "| 7 | 0 |\n",
       "| 8 | 0 |\n",
       "| 9 | 1 |\n",
       "| 10 | 1 |\n",
       "| 11 | 1 |\n",
       "| 12 | 1 |\n",
       "| 13 | 1 |\n",
       "| 14 | 1 |\n",
       "| 15 | 1 |\n",
       "| 16 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "   pData(mydata)$Status\n",
       "1  0                   \n",
       "2  0                   \n",
       "3  0                   \n",
       "4  0                   \n",
       "5  0                   \n",
       "6  0                   \n",
       "7  0                   \n",
       "8  0                   \n",
       "9  1                   \n",
       "10 1                   \n",
       "11 1                   \n",
       "12 1                   \n",
       "13 1                   \n",
       "14 1                   \n",
       "15 1                   \n",
       "16 1                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Prepare a design matrix based on the experiment details and `phenoData`\n",
    "## The design matrix describes the experiment condition in each of its column. In our case,\n",
    "## we have only two conditions, and hence, the single column design matrix with appropriate experiment indication works as well.\n",
    "design <- model.matrix(~0 + pData(mydata)$Status)   # design <- model.matrix(~-1 + factor(pData(myData_quantile)$type))\n",
    "design"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Your code contains a unicode char which cannot be displayed in your\n",
      "current locale and R will silently convert it to an escaped form when the\n",
      "R kernel executes this code. This can lead to subtle errors if you use\n",
      "such chars to do comparisons. For more information, please see\n",
      "https://github.com/IRkernel/repr/wiki/Problems-with-unicode-on-windows"
     ]
    },
    {
     "data": {
      "text/plain": [
       "An object of class \"MArrayLM\"\n",
       "$coefficients\n",
       "             pData(mydata)$Status\n",
       "1555078_at             -0.3891325\n",
       "238493_at               0.3318299\n",
       "1562133_x_at           -0.3822317\n",
       "1559616_x_at           -0.9623383\n",
       "235687_at              -0.6461947\n",
       "5334 more rows ...\n",
       "\n",
       "$rank\n",
       "[1] 1\n",
       "\n",
       "$assign\n",
       "[1] 1\n",
       "\n",
       "$qr\n",
       "$qr\n",
       "  pData(mydata)$Status\n",
       "1            -2.828427\n",
       "2             0.000000\n",
       "3             0.000000\n",
       "4             0.000000\n",
       "5             0.000000\n",
       "11 more rows ...\n",
       "\n",
       "$qraux\n",
       "[1] 1\n",
       "\n",
       "$pivot\n",
       "[1] 1\n",
       "\n",
       "$tol\n",
       "[1] 1e-07\n",
       "\n",
       "$rank\n",
       "[1] 1\n",
       "\n",
       "\n",
       "$df.residual\n",
       "[1] 15 15 15 15 15\n",
       "5334 more elements ...\n",
       "\n",
       "$sigma\n",
       "  1555078_at    238493_at 1562133_x_at 1559616_x_at    235687_at \n",
       "   0.3893793    1.3899476    0.4058717    0.6304099    0.6146187 \n",
       "5334 more elements ...\n",
       "\n",
       "$cov.coefficients\n",
       "                     pData(mydata)$Status\n",
       "pData(mydata)$Status                0.125\n",
       "\n",
       "$stdev.unscaled\n",
       "             pData(mydata)$Status\n",
       "1555078_at              0.3535534\n",
       "238493_at               0.3535534\n",
       "1562133_x_at            0.3535534\n",
       "1559616_x_at            0.3535534\n",
       "235687_at               0.3535534\n",
       "5334 more rows ...\n",
       "\n",
       "$pivot\n",
       "[1] 1\n",
       "\n",
       "$Amean\n",
       "  1555078_at    238493_at 1562133_x_at 1559616_x_at    235687_at \n",
       "  -0.2382232    0.6460257   -0.4434705   -0.8408450   -0.6992017 \n",
       "5334 more elements ...\n",
       "\n",
       "$method\n",
       "[1] \"ls\"\n",
       "\n",
       "$design\n",
       "  pData(mydata)$Status\n",
       "1                    0\n",
       "2                    0\n",
       "3                    0\n",
       "4                    0\n",
       "5                    0\n",
       "11 more rows ...\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Fit a linear model using the expression data and design matrix.使用表达式数据和设计矩阵拟合线性模型.\n",
    "fit <- lmFit(mydata_quantile,design)\n",
    "fit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "An object of class \"MArrayLM\"\n",
       "$coefficients\n",
       "             pData(mydata)$Status\n",
       "1555078_at             -0.3891325\n",
       "238493_at               0.3318299\n",
       "1562133_x_at           -0.3822317\n",
       "1559616_x_at           -0.9623383\n",
       "235687_at              -0.6461947\n",
       "5334 more rows ...\n",
       "\n",
       "$rank\n",
       "[1] 1\n",
       "\n",
       "$assign\n",
       "[1] 1\n",
       "\n",
       "$qr\n",
       "$qr\n",
       "  pData(mydata)$Status\n",
       "1            -2.828427\n",
       "2             0.000000\n",
       "3             0.000000\n",
       "4             0.000000\n",
       "5             0.000000\n",
       "11 more rows ...\n",
       "\n",
       "$qraux\n",
       "[1] 1\n",
       "\n",
       "$pivot\n",
       "[1] 1\n",
       "\n",
       "$tol\n",
       "[1] 1e-07\n",
       "\n",
       "$rank\n",
       "[1] 1\n",
       "\n",
       "\n",
       "$df.residual\n",
       "[1] 15 15 15 15 15\n",
       "5334 more elements ...\n",
       "\n",
       "$sigma\n",
       "  1555078_at    238493_at 1562133_x_at 1559616_x_at    235687_at \n",
       "   0.3893793    1.3899476    0.4058717    0.6304099    0.6146187 \n",
       "5334 more elements ...\n",
       "\n",
       "$cov.coefficients\n",
       "                     pData(mydata)$Status\n",
       "pData(mydata)$Status                0.125\n",
       "\n",
       "$stdev.unscaled\n",
       "             pData(mydata)$Status\n",
       "1555078_at              0.3535534\n",
       "238493_at               0.3535534\n",
       "1562133_x_at            0.3535534\n",
       "1559616_x_at            0.3535534\n",
       "235687_at               0.3535534\n",
       "5334 more rows ...\n",
       "\n",
       "$pivot\n",
       "[1] 1\n",
       "\n",
       "$Amean\n",
       "  1555078_at    238493_at 1562133_x_at 1559616_x_at    235687_at \n",
       "  -0.2382232    0.6460257   -0.4434705   -0.8408450   -0.6992017 \n",
       "5334 more elements ...\n",
       "\n",
       "$method\n",
       "[1] \"ls\"\n",
       "\n",
       "$design\n",
       "  pData(mydata)$Status\n",
       "1                    0\n",
       "2                    0\n",
       "3                    0\n",
       "4                    0\n",
       "5                    0\n",
       "11 more rows ...\n",
       "\n",
       "$df.prior\n",
       "[1] 1.316078\n",
       "\n",
       "$s2.prior\n",
       "[1] 0.2580634\n",
       "\n",
       "$var.prior\n",
       "[1] 9.156446\n",
       "\n",
       "$proportion\n",
       "[1] 0.01\n",
       "\n",
       "$s2.post\n",
       "  1555078_at    238493_at 1562133_x_at 1559616_x_at    235687_at \n",
       "   0.1602024    1.7969358    0.1722601    0.3861762    0.3681016 \n",
       "5334 more elements ...\n",
       "\n",
       "$t\n",
       "             pData(mydata)$Status\n",
       "1555078_at              -2.749843\n",
       "238493_at                0.700155\n",
       "1562133_x_at            -2.604830\n",
       "1559616_x_at            -4.380060\n",
       "235687_at               -3.012482\n",
       "5334 more rows ...\n",
       "\n",
       "$df.total\n",
       "[1] 16.31608 16.31608 16.31608 16.31608 16.31608\n",
       "5334 more elements ...\n",
       "\n",
       "$p.value\n",
       "             pData(mydata)$Status\n",
       "1555078_at           0.0140512222\n",
       "238493_at            0.4936939029\n",
       "1562133_x_at         0.0189345321\n",
       "1559616_x_at         0.0004465372\n",
       "235687_at            0.0081226177\n",
       "5334 more rows ...\n",
       "\n",
       "$lods\n",
       "             pData(mydata)$Status\n",
       "1555078_at             -3.5057863\n",
       "238493_at              -6.4960522\n",
       "1562133_x_at           -3.7865081\n",
       "1559616_x_at           -0.1540363\n",
       "235687_at              -2.9844557\n",
       "5334 more rows ...\n",
       "\n",
       "$F\n",
       "[1]  7.561638  0.490217  6.785137 19.184923  9.075050\n",
       "5334 more elements ...\n",
       "\n",
       "$F.p.value\n",
       "[1] 0.0140512222 0.4936939029 0.0189345321 0.0004465372 0.0081226177\n",
       "5334 more elements ...\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## After having a linear model fit, compute the moderated statistics for it by `eBayes` function\n",
    "fitE <- eBayes(fit)\n",
    "fitE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 5339 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>logFC</th><th scope=col>AveExpr</th><th scope=col>t</th><th scope=col>P.Value</th><th scope=col>adj.P.Val</th><th scope=col>B</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>210372_s_at</th><td> 5.200189</td><td> 2.9369060</td><td> 18.586820</td><td>2.082126e-12</td><td>1.111647e-08</td><td>17.909406</td></tr>\n",
       "\t<tr><th scope=row>205470_s_at</th><td> 6.255458</td><td> 3.4624353</td><td> 14.523453</td><td>9.365043e-11</td><td>2.499998e-07</td><td>14.665745</td></tr>\n",
       "\t<tr><th scope=row>204855_at</th><td>13.342429</td><td> 8.3325984</td><td> 12.511677</td><td>8.820153e-10</td><td>1.260731e-06</td><td>12.634755</td></tr>\n",
       "\t<tr><th scope=row>219795_at</th><td> 6.773997</td><td> 3.2696607</td><td> 12.454113</td><td>9.445448e-10</td><td>1.260731e-06</td><td>12.571572</td></tr>\n",
       "\t<tr><th scope=row>220133_at</th><td> 3.223162</td><td> 1.5617431</td><td> 12.127986</td><td>1.399296e-09</td><td>1.494168e-06</td><td>12.207825</td></tr>\n",
       "\t<tr><th scope=row>204259_at</th><td> 7.111360</td><td> 3.6682683</td><td> 10.809794</td><td>7.511799e-09</td><td>6.684249e-06</td><td>10.631112</td></tr>\n",
       "\t<tr><th scope=row>203585_at</th><td> 3.814912</td><td> 2.3970818</td><td> 10.575189</td><td>1.030032e-08</td><td>7.014783e-06</td><td>10.331384</td></tr>\n",
       "\t<tr><th scope=row>203485_at</th><td>-1.477512</td><td>-0.7942994</td><td>-10.526023</td><td>1.101229e-08</td><td>7.014783e-06</td><td>10.267797</td></tr>\n",
       "\t<tr><th scope=row>219752_at</th><td> 2.948165</td><td> 1.8787889</td><td> 10.473843</td><td>1.182488e-08</td><td>7.014783e-06</td><td>10.200014</td></tr>\n",
       "\t<tr><th scope=row>232151_at</th><td> 5.490196</td><td> 3.4301028</td><td> 10.321526</td><td>1.457882e-08</td><td>7.783631e-06</td><td>10.000392</td></tr>\n",
       "\t<tr><th scope=row>209309_at</th><td> 6.996555</td><td> 4.1278207</td><td>  9.848596</td><td>2.835113e-08</td><td>1.376061e-05</td><td> 9.363467</td></tr>\n",
       "\t<tr><th scope=row>230398_at</th><td> 5.665631</td><td> 3.3003175</td><td>  9.766528</td><td>3.189540e-08</td><td>1.419079e-05</td><td> 9.250240</td></tr>\n",
       "\t<tr><th scope=row>218412_s_at</th><td> 7.497466</td><td> 5.2023793</td><td>  9.423892</td><td>5.256847e-08</td><td>2.158947e-05</td><td> 8.768637</td></tr>\n",
       "\t<tr><th scope=row>203962_s_at</th><td>10.751804</td><td> 7.5600927</td><td>  9.370139</td><td>5.692160e-08</td><td>2.166592e-05</td><td> 8.691761</td></tr>\n",
       "\t<tr><th scope=row>212531_at</th><td>17.064917</td><td>11.9881353</td><td>  9.317129</td><td>6.158685e-08</td><td>2.166592e-05</td><td> 8.615594</td></tr>\n",
       "\t<tr><th scope=row>241782_at</th><td> 2.195624</td><td> 1.1963186</td><td>  9.253373</td><td>6.773548e-08</td><td>2.166592e-05</td><td> 8.523515</td></tr>\n",
       "\t<tr><th scope=row>235245_at</th><td> 2.334599</td><td> 1.4756777</td><td>  9.227164</td><td>7.044712e-08</td><td>2.166592e-05</td><td> 8.485513</td></tr>\n",
       "\t<tr><th scope=row>208209_s_at</th><td> 3.822005</td><td> 2.4020938</td><td>  9.203031</td><td>7.304486e-08</td><td>2.166592e-05</td><td> 8.450445</td></tr>\n",
       "\t<tr><th scope=row>1566766_a_at</th><td> 4.146974</td><td> 2.6918098</td><td>  9.136375</td><td>8.075648e-08</td><td>2.269257e-05</td><td> 8.353196</td></tr>\n",
       "\t<tr><th scope=row>244180_at</th><td>-1.344145</td><td>-0.8307531</td><td> -9.061854</td><td>9.040027e-08</td><td>2.413235e-05</td><td> 8.243797</td></tr>\n",
       "\t<tr><th scope=row>205566_at</th><td> 3.226118</td><td> 1.9002566</td><td>  9.012534</td><td>9.744220e-08</td><td>2.477352e-05</td><td> 8.171000</td></tr>\n",
       "\t<tr><th scope=row>206941_x_at</th><td>-1.005474</td><td>-0.6635295</td><td> -8.962673</td><td>1.051493e-07</td><td>2.551782e-05</td><td> 8.097083</td></tr>\n",
       "\t<tr><th scope=row>205349_at</th><td> 3.212065</td><td> 1.6030105</td><td>  8.791665</td><td>1.368203e-07</td><td>3.176015e-05</td><td> 7.841098</td></tr>\n",
       "\t<tr><th scope=row>202917_s_at</th><td> 9.110590</td><td> 4.6962185</td><td>  8.695275</td><td>1.589515e-07</td><td>3.340168e-05</td><td> 7.695107</td></tr>\n",
       "\t<tr><th scope=row>202357_s_at</th><td> 8.271597</td><td> 5.6286043</td><td>  8.689031</td><td>1.605088e-07</td><td>3.340168e-05</td><td> 7.685609</td></tr>\n",
       "\t<tr><th scope=row>205977_s_at</th><td> 4.119680</td><td> 2.8835908</td><td>  8.680509</td><td>1.626604e-07</td><td>3.340168e-05</td><td> 7.672635</td></tr>\n",
       "\t<tr><th scope=row>209409_at</th><td> 5.113576</td><td> 3.5591380</td><td>  8.444412</td><td>2.360515e-07</td><td>4.667700e-05</td><td> 7.309324</td></tr>\n",
       "\t<tr><th scope=row>218704_at</th><td> 7.976579</td><td> 5.7665953</td><td>  8.404005</td><td>2.517568e-07</td><td>4.800462e-05</td><td> 7.246389</td></tr>\n",
       "\t<tr><th scope=row>221815_at</th><td> 4.468376</td><td> 3.0759494</td><td>  8.351318</td><td>2.738977e-07</td><td>5.042551e-05</td><td> 7.163994</td></tr>\n",
       "\t<tr><th scope=row>205476_at</th><td>14.466160</td><td>10.4372247</td><td>  8.248990</td><td>3.229343e-07</td><td>5.747153e-05</td><td> 7.002877</td></tr>\n",
       "\t<tr><th scope=row>...</th><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td></tr>\n",
       "\t<tr><th scope=row>1562211_a_at</th><td>-3.627688e-03</td><td> 0.033859078</td><td>-0.0281154771</td><td>0.9779111</td><td>0.9832518</td><td>-6.748436</td></tr>\n",
       "\t<tr><th scope=row>220373_at</th><td> 5.740751e-03</td><td>-0.258401883</td><td> 0.0269088363</td><td>0.9788588</td><td>0.9840194</td><td>-6.748470</td></tr>\n",
       "\t<tr><th scope=row>231626_at</th><td> 2.757279e-02</td><td> 1.857394629</td><td> 0.0253280978</td><td>0.9801005</td><td>0.9850070</td><td>-6.748514</td></tr>\n",
       "\t<tr><th scope=row>220343_at</th><td>-2.956402e-03</td><td> 0.004189963</td><td>-0.0251884424</td><td>0.9802102</td><td>0.9850070</td><td>-6.748517</td></tr>\n",
       "\t<tr><th scope=row>1559429_a_at</th><td> 3.890462e-03</td><td>-0.040891499</td><td> 0.0217618683</td><td>0.9829018</td><td>0.9875259</td><td>-6.748601</td></tr>\n",
       "\t<tr><th scope=row>221690_s_at</th><td> 1.831913e-02</td><td> 0.659423984</td><td> 0.0214927376</td><td>0.9831133</td><td>0.9875525</td><td>-6.748608</td></tr>\n",
       "\t<tr><th scope=row>1553059_at</th><td>-3.434531e-03</td><td>-0.163317401</td><td>-0.0210531485</td><td>0.9834586</td><td>0.9877136</td><td>-6.748617</td></tr>\n",
       "\t<tr><th scope=row>206834_at</th><td>-4.011657e-03</td><td>-0.167439369</td><td>-0.0198192016</td><td>0.9844279</td><td>0.9885012</td><td>-6.748644</td></tr>\n",
       "\t<tr><th scope=row>234361_at</th><td> 3.016104e-03</td><td> 0.044745530</td><td> 0.0192995924</td><td>0.9848362</td><td>0.9887251</td><td>-6.748654</td></tr>\n",
       "\t<tr><th scope=row>224390_s_at</th><td>-1.756399e-03</td><td>-0.001132483</td><td>-0.0188067738</td><td>0.9852233</td><td>0.9889247</td><td>-6.748664</td></tr>\n",
       "\t<tr><th scope=row>240967_at</th><td> 2.842829e-03</td><td>-0.164153264</td><td> 0.0185750184</td><td>0.9854054</td><td>0.9889247</td><td>-6.748669</td></tr>\n",
       "\t<tr><th scope=row>243231_at</th><td>-4.517521e-03</td><td> 0.252822841</td><td>-0.0168659091</td><td>0.9867481</td><td>0.9900861</td><td>-6.748700</td></tr>\n",
       "\t<tr><th scope=row>1552526_at</th><td>-1.509388e-03</td><td> 0.042989550</td><td>-0.0147194442</td><td>0.9884345</td><td>0.9915918</td><td>-6.748736</td></tr>\n",
       "\t<tr><th scope=row>213780_at</th><td>-1.830945e-03</td><td>-0.012559300</td><td>-0.0120276478</td><td>0.9905494</td><td>0.9935268</td><td>-6.748774</td></tr>\n",
       "\t<tr><th scope=row>229152_at</th><td> 7.274305e-03</td><td> 0.785249728</td><td> 0.0094557174</td><td>0.9925702</td><td>0.9953667</td><td>-6.748803</td></tr>\n",
       "\t<tr><th scope=row>242883_at</th><td> 1.284877e-03</td><td>-0.202721125</td><td> 0.0086463667</td><td>0.9932061</td><td>0.9958174</td><td>-6.748810</td></tr>\n",
       "\t<tr><th scope=row>221658_s_at</th><td>-1.067008e-03</td><td>-0.026685009</td><td>-0.0081065141</td><td>0.9936303</td><td>0.9960556</td><td>-6.748815</td></tr>\n",
       "\t<tr><th scope=row>231386_at</th><td> 8.382728e-04</td><td> 0.027066565</td><td> 0.0077754172</td><td>0.9938905</td><td>0.9961049</td><td>-6.748818</td></tr>\n",
       "\t<tr><th scope=row>1555643_s_at</th><td>-1.162331e-03</td><td> 0.030274093</td><td>-0.0075690004</td><td>0.9940526</td><td>0.9961049</td><td>-6.748819</td></tr>\n",
       "\t<tr><th scope=row>207972_at</th><td>-9.074408e-04</td><td>-0.120894936</td><td>-0.0060321130</td><td>0.9952602</td><td>0.9971208</td><td>-6.748830</td></tr>\n",
       "\t<tr><th scope=row>210402_at</th><td>-8.584665e-04</td><td> 0.164386348</td><td>-0.0058033702</td><td>0.9954400</td><td>0.9971208</td><td>-6.748832</td></tr>\n",
       "\t<tr><th scope=row>1552948_at</th><td>-9.712917e-04</td><td> 0.037384033</td><td>-0.0055466975</td><td>0.9956417</td><td>0.9971358</td><td>-6.748833</td></tr>\n",
       "\t<tr><th scope=row>205685_at</th><td>-8.604690e-04</td><td> 0.230873696</td><td>-0.0047693364</td><td>0.9962525</td><td>0.9975604</td><td>-6.748837</td></tr>\n",
       "\t<tr><th scope=row>231794_at</th><td>-6.487468e-04</td><td> 0.098780981</td><td>-0.0034424825</td><td>0.9972950</td><td>0.9984065</td><td>-6.748843</td></tr>\n",
       "\t<tr><th scope=row>242809_at</th><td>-5.042769e-04</td><td>-0.022999856</td><td>-0.0029907888</td><td>0.9976500</td><td>0.9984065</td><td>-6.748845</td></tr>\n",
       "\t<tr><th scope=row>203603_s_at</th><td> 2.719045e-03</td><td> 1.850278399</td><td> 0.0027853887</td><td>0.9978114</td><td>0.9984065</td><td>-6.748845</td></tr>\n",
       "\t<tr><th scope=row>208402_at</th><td>-3.644787e-04</td><td>-0.131042892</td><td>-0.0026030083</td><td>0.9979547</td><td>0.9984065</td><td>-6.748846</td></tr>\n",
       "\t<tr><th scope=row>1553534_at</th><td>-2.792301e-04</td><td>-0.044046503</td><td>-0.0025039235</td><td>0.9980325</td><td>0.9984065</td><td>-6.748846</td></tr>\n",
       "\t<tr><th scope=row>237320_at</th><td> 4.826854e-05</td><td>-0.213631604</td><td> 0.0002714129</td><td>0.9997867</td><td>0.9999727</td><td>-6.748849</td></tr>\n",
       "\t<tr><th scope=row>207067_s_at</th><td>-7.696790e-06</td><td> 0.334753187</td><td>-0.0000347936</td><td>0.9999727</td><td>0.9999727</td><td>-6.748849</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5339 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & logFC & AveExpr & t & P.Value & adj.P.Val & B\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t210372\\_s\\_at &  5.200189 &  2.9369060 &  18.586820 & 2.082126e-12 & 1.111647e-08 & 17.909406\\\\\n",
       "\t205470\\_s\\_at &  6.255458 &  3.4624353 &  14.523453 & 9.365043e-11 & 2.499998e-07 & 14.665745\\\\\n",
       "\t204855\\_at & 13.342429 &  8.3325984 &  12.511677 & 8.820153e-10 & 1.260731e-06 & 12.634755\\\\\n",
       "\t219795\\_at &  6.773997 &  3.2696607 &  12.454113 & 9.445448e-10 & 1.260731e-06 & 12.571572\\\\\n",
       "\t220133\\_at &  3.223162 &  1.5617431 &  12.127986 & 1.399296e-09 & 1.494168e-06 & 12.207825\\\\\n",
       "\t204259\\_at &  7.111360 &  3.6682683 &  10.809794 & 7.511799e-09 & 6.684249e-06 & 10.631112\\\\\n",
       "\t203585\\_at &  3.814912 &  2.3970818 &  10.575189 & 1.030032e-08 & 7.014783e-06 & 10.331384\\\\\n",
       "\t203485\\_at & -1.477512 & -0.7942994 & -10.526023 & 1.101229e-08 & 7.014783e-06 & 10.267797\\\\\n",
       "\t219752\\_at &  2.948165 &  1.8787889 &  10.473843 & 1.182488e-08 & 7.014783e-06 & 10.200014\\\\\n",
       "\t232151\\_at &  5.490196 &  3.4301028 &  10.321526 & 1.457882e-08 & 7.783631e-06 & 10.000392\\\\\n",
       "\t209309\\_at &  6.996555 &  4.1278207 &   9.848596 & 2.835113e-08 & 1.376061e-05 &  9.363467\\\\\n",
       "\t230398\\_at &  5.665631 &  3.3003175 &   9.766528 & 3.189540e-08 & 1.419079e-05 &  9.250240\\\\\n",
       "\t218412\\_s\\_at &  7.497466 &  5.2023793 &   9.423892 & 5.256847e-08 & 2.158947e-05 &  8.768637\\\\\n",
       "\t203962\\_s\\_at & 10.751804 &  7.5600927 &   9.370139 & 5.692160e-08 & 2.166592e-05 &  8.691761\\\\\n",
       "\t212531\\_at & 17.064917 & 11.9881353 &   9.317129 & 6.158685e-08 & 2.166592e-05 &  8.615594\\\\\n",
       "\t241782\\_at &  2.195624 &  1.1963186 &   9.253373 & 6.773548e-08 & 2.166592e-05 &  8.523515\\\\\n",
       "\t235245\\_at &  2.334599 &  1.4756777 &   9.227164 & 7.044712e-08 & 2.166592e-05 &  8.485513\\\\\n",
       "\t208209\\_s\\_at &  3.822005 &  2.4020938 &   9.203031 & 7.304486e-08 & 2.166592e-05 &  8.450445\\\\\n",
       "\t1566766\\_a\\_at &  4.146974 &  2.6918098 &   9.136375 & 8.075648e-08 & 2.269257e-05 &  8.353196\\\\\n",
       "\t244180\\_at & -1.344145 & -0.8307531 &  -9.061854 & 9.040027e-08 & 2.413235e-05 &  8.243797\\\\\n",
       "\t205566\\_at &  3.226118 &  1.9002566 &   9.012534 & 9.744220e-08 & 2.477352e-05 &  8.171000\\\\\n",
       "\t206941\\_x\\_at & -1.005474 & -0.6635295 &  -8.962673 & 1.051493e-07 & 2.551782e-05 &  8.097083\\\\\n",
       "\t205349\\_at &  3.212065 &  1.6030105 &   8.791665 & 1.368203e-07 & 3.176015e-05 &  7.841098\\\\\n",
       "\t202917\\_s\\_at &  9.110590 &  4.6962185 &   8.695275 & 1.589515e-07 & 3.340168e-05 &  7.695107\\\\\n",
       "\t202357\\_s\\_at &  8.271597 &  5.6286043 &   8.689031 & 1.605088e-07 & 3.340168e-05 &  7.685609\\\\\n",
       "\t205977\\_s\\_at &  4.119680 &  2.8835908 &   8.680509 & 1.626604e-07 & 3.340168e-05 &  7.672635\\\\\n",
       "\t209409\\_at &  5.113576 &  3.5591380 &   8.444412 & 2.360515e-07 & 4.667700e-05 &  7.309324\\\\\n",
       "\t218704\\_at &  7.976579 &  5.7665953 &   8.404005 & 2.517568e-07 & 4.800462e-05 &  7.246389\\\\\n",
       "\t221815\\_at &  4.468376 &  3.0759494 &   8.351318 & 2.738977e-07 & 5.042551e-05 &  7.163994\\\\\n",
       "\t205476\\_at & 14.466160 & 10.4372247 &   8.248990 & 3.229343e-07 & 5.747153e-05 &  7.002877\\\\\n",
       "\t... & ... & ... & ... & ... & ... & ...\\\\\n",
       "\t1562211\\_a\\_at & -3.627688e-03 &  0.033859078 & -0.0281154771 & 0.9779111 & 0.9832518 & -6.748436\\\\\n",
       "\t220373\\_at &  5.740751e-03 & -0.258401883 &  0.0269088363 & 0.9788588 & 0.9840194 & -6.748470\\\\\n",
       "\t231626\\_at &  2.757279e-02 &  1.857394629 &  0.0253280978 & 0.9801005 & 0.9850070 & -6.748514\\\\\n",
       "\t220343\\_at & -2.956402e-03 &  0.004189963 & -0.0251884424 & 0.9802102 & 0.9850070 & -6.748517\\\\\n",
       "\t1559429\\_a\\_at &  3.890462e-03 & -0.040891499 &  0.0217618683 & 0.9829018 & 0.9875259 & -6.748601\\\\\n",
       "\t221690\\_s\\_at &  1.831913e-02 &  0.659423984 &  0.0214927376 & 0.9831133 & 0.9875525 & -6.748608\\\\\n",
       "\t1553059\\_at & -3.434531e-03 & -0.163317401 & -0.0210531485 & 0.9834586 & 0.9877136 & -6.748617\\\\\n",
       "\t206834\\_at & -4.011657e-03 & -0.167439369 & -0.0198192016 & 0.9844279 & 0.9885012 & -6.748644\\\\\n",
       "\t234361\\_at &  3.016104e-03 &  0.044745530 &  0.0192995924 & 0.9848362 & 0.9887251 & -6.748654\\\\\n",
       "\t224390\\_s\\_at & -1.756399e-03 & -0.001132483 & -0.0188067738 & 0.9852233 & 0.9889247 & -6.748664\\\\\n",
       "\t240967\\_at &  2.842829e-03 & -0.164153264 &  0.0185750184 & 0.9854054 & 0.9889247 & -6.748669\\\\\n",
       "\t243231\\_at & -4.517521e-03 &  0.252822841 & -0.0168659091 & 0.9867481 & 0.9900861 & -6.748700\\\\\n",
       "\t1552526\\_at & -1.509388e-03 &  0.042989550 & -0.0147194442 & 0.9884345 & 0.9915918 & -6.748736\\\\\n",
       "\t213780\\_at & -1.830945e-03 & -0.012559300 & -0.0120276478 & 0.9905494 & 0.9935268 & -6.748774\\\\\n",
       "\t229152\\_at &  7.274305e-03 &  0.785249728 &  0.0094557174 & 0.9925702 & 0.9953667 & -6.748803\\\\\n",
       "\t242883\\_at &  1.284877e-03 & -0.202721125 &  0.0086463667 & 0.9932061 & 0.9958174 & -6.748810\\\\\n",
       "\t221658\\_s\\_at & -1.067008e-03 & -0.026685009 & -0.0081065141 & 0.9936303 & 0.9960556 & -6.748815\\\\\n",
       "\t231386\\_at &  8.382728e-04 &  0.027066565 &  0.0077754172 & 0.9938905 & 0.9961049 & -6.748818\\\\\n",
       "\t1555643\\_s\\_at & -1.162331e-03 &  0.030274093 & -0.0075690004 & 0.9940526 & 0.9961049 & -6.748819\\\\\n",
       "\t207972\\_at & -9.074408e-04 & -0.120894936 & -0.0060321130 & 0.9952602 & 0.9971208 & -6.748830\\\\\n",
       "\t210402\\_at & -8.584665e-04 &  0.164386348 & -0.0058033702 & 0.9954400 & 0.9971208 & -6.748832\\\\\n",
       "\t1552948\\_at & -9.712917e-04 &  0.037384033 & -0.0055466975 & 0.9956417 & 0.9971358 & -6.748833\\\\\n",
       "\t205685\\_at & -8.604690e-04 &  0.230873696 & -0.0047693364 & 0.9962525 & 0.9975604 & -6.748837\\\\\n",
       "\t231794\\_at & -6.487468e-04 &  0.098780981 & -0.0034424825 & 0.9972950 & 0.9984065 & -6.748843\\\\\n",
       "\t242809\\_at & -5.042769e-04 & -0.022999856 & -0.0029907888 & 0.9976500 & 0.9984065 & -6.748845\\\\\n",
       "\t203603\\_s\\_at &  2.719045e-03 &  1.850278399 &  0.0027853887 & 0.9978114 & 0.9984065 & -6.748845\\\\\n",
       "\t208402\\_at & -3.644787e-04 & -0.131042892 & -0.0026030083 & 0.9979547 & 0.9984065 & -6.748846\\\\\n",
       "\t1553534\\_at & -2.792301e-04 & -0.044046503 & -0.0025039235 & 0.9980325 & 0.9984065 & -6.748846\\\\\n",
       "\t237320\\_at &  4.826854e-05 & -0.213631604 &  0.0002714129 & 0.9997867 & 0.9999727 & -6.748849\\\\\n",
       "\t207067\\_s\\_at & -7.696790e-06 &  0.334753187 & -0.0000347936 & 0.9999727 & 0.9999727 & -6.748849\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5339 × 6\n",
       "\n",
       "| <!--/--> | logFC &lt;dbl&gt; | AveExpr &lt;dbl&gt; | t &lt;dbl&gt; | P.Value &lt;dbl&gt; | adj.P.Val &lt;dbl&gt; | B &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 210372_s_at |  5.200189 |  2.9369060 |  18.586820 | 2.082126e-12 | 1.111647e-08 | 17.909406 |\n",
       "| 205470_s_at |  6.255458 |  3.4624353 |  14.523453 | 9.365043e-11 | 2.499998e-07 | 14.665745 |\n",
       "| 204855_at | 13.342429 |  8.3325984 |  12.511677 | 8.820153e-10 | 1.260731e-06 | 12.634755 |\n",
       "| 219795_at |  6.773997 |  3.2696607 |  12.454113 | 9.445448e-10 | 1.260731e-06 | 12.571572 |\n",
       "| 220133_at |  3.223162 |  1.5617431 |  12.127986 | 1.399296e-09 | 1.494168e-06 | 12.207825 |\n",
       "| 204259_at |  7.111360 |  3.6682683 |  10.809794 | 7.511799e-09 | 6.684249e-06 | 10.631112 |\n",
       "| 203585_at |  3.814912 |  2.3970818 |  10.575189 | 1.030032e-08 | 7.014783e-06 | 10.331384 |\n",
       "| 203485_at | -1.477512 | -0.7942994 | -10.526023 | 1.101229e-08 | 7.014783e-06 | 10.267797 |\n",
       "| 219752_at |  2.948165 |  1.8787889 |  10.473843 | 1.182488e-08 | 7.014783e-06 | 10.200014 |\n",
       "| 232151_at |  5.490196 |  3.4301028 |  10.321526 | 1.457882e-08 | 7.783631e-06 | 10.000392 |\n",
       "| 209309_at |  6.996555 |  4.1278207 |   9.848596 | 2.835113e-08 | 1.376061e-05 |  9.363467 |\n",
       "| 230398_at |  5.665631 |  3.3003175 |   9.766528 | 3.189540e-08 | 1.419079e-05 |  9.250240 |\n",
       "| 218412_s_at |  7.497466 |  5.2023793 |   9.423892 | 5.256847e-08 | 2.158947e-05 |  8.768637 |\n",
       "| 203962_s_at | 10.751804 |  7.5600927 |   9.370139 | 5.692160e-08 | 2.166592e-05 |  8.691761 |\n",
       "| 212531_at | 17.064917 | 11.9881353 |   9.317129 | 6.158685e-08 | 2.166592e-05 |  8.615594 |\n",
       "| 241782_at |  2.195624 |  1.1963186 |   9.253373 | 6.773548e-08 | 2.166592e-05 |  8.523515 |\n",
       "| 235245_at |  2.334599 |  1.4756777 |   9.227164 | 7.044712e-08 | 2.166592e-05 |  8.485513 |\n",
       "| 208209_s_at |  3.822005 |  2.4020938 |   9.203031 | 7.304486e-08 | 2.166592e-05 |  8.450445 |\n",
       "| 1566766_a_at |  4.146974 |  2.6918098 |   9.136375 | 8.075648e-08 | 2.269257e-05 |  8.353196 |\n",
       "| 244180_at | -1.344145 | -0.8307531 |  -9.061854 | 9.040027e-08 | 2.413235e-05 |  8.243797 |\n",
       "| 205566_at |  3.226118 |  1.9002566 |   9.012534 | 9.744220e-08 | 2.477352e-05 |  8.171000 |\n",
       "| 206941_x_at | -1.005474 | -0.6635295 |  -8.962673 | 1.051493e-07 | 2.551782e-05 |  8.097083 |\n",
       "| 205349_at |  3.212065 |  1.6030105 |   8.791665 | 1.368203e-07 | 3.176015e-05 |  7.841098 |\n",
       "| 202917_s_at |  9.110590 |  4.6962185 |   8.695275 | 1.589515e-07 | 3.340168e-05 |  7.695107 |\n",
       "| 202357_s_at |  8.271597 |  5.6286043 |   8.689031 | 1.605088e-07 | 3.340168e-05 |  7.685609 |\n",
       "| 205977_s_at |  4.119680 |  2.8835908 |   8.680509 | 1.626604e-07 | 3.340168e-05 |  7.672635 |\n",
       "| 209409_at |  5.113576 |  3.5591380 |   8.444412 | 2.360515e-07 | 4.667700e-05 |  7.309324 |\n",
       "| 218704_at |  7.976579 |  5.7665953 |   8.404005 | 2.517568e-07 | 4.800462e-05 |  7.246389 |\n",
       "| 221815_at |  4.468376 |  3.0759494 |   8.351318 | 2.738977e-07 | 5.042551e-05 |  7.163994 |\n",
       "| 205476_at | 14.466160 | 10.4372247 |   8.248990 | 3.229343e-07 | 5.747153e-05 |  7.002877 |\n",
       "| ... | ... | ... | ... | ... | ... | ... |\n",
       "| 1562211_a_at | -3.627688e-03 |  0.033859078 | -0.0281154771 | 0.9779111 | 0.9832518 | -6.748436 |\n",
       "| 220373_at |  5.740751e-03 | -0.258401883 |  0.0269088363 | 0.9788588 | 0.9840194 | -6.748470 |\n",
       "| 231626_at |  2.757279e-02 |  1.857394629 |  0.0253280978 | 0.9801005 | 0.9850070 | -6.748514 |\n",
       "| 220343_at | -2.956402e-03 |  0.004189963 | -0.0251884424 | 0.9802102 | 0.9850070 | -6.748517 |\n",
       "| 1559429_a_at |  3.890462e-03 | -0.040891499 |  0.0217618683 | 0.9829018 | 0.9875259 | -6.748601 |\n",
       "| 221690_s_at |  1.831913e-02 |  0.659423984 |  0.0214927376 | 0.9831133 | 0.9875525 | -6.748608 |\n",
       "| 1553059_at | -3.434531e-03 | -0.163317401 | -0.0210531485 | 0.9834586 | 0.9877136 | -6.748617 |\n",
       "| 206834_at | -4.011657e-03 | -0.167439369 | -0.0198192016 | 0.9844279 | 0.9885012 | -6.748644 |\n",
       "| 234361_at |  3.016104e-03 |  0.044745530 |  0.0192995924 | 0.9848362 | 0.9887251 | -6.748654 |\n",
       "| 224390_s_at | -1.756399e-03 | -0.001132483 | -0.0188067738 | 0.9852233 | 0.9889247 | -6.748664 |\n",
       "| 240967_at |  2.842829e-03 | -0.164153264 |  0.0185750184 | 0.9854054 | 0.9889247 | -6.748669 |\n",
       "| 243231_at | -4.517521e-03 |  0.252822841 | -0.0168659091 | 0.9867481 | 0.9900861 | -6.748700 |\n",
       "| 1552526_at | -1.509388e-03 |  0.042989550 | -0.0147194442 | 0.9884345 | 0.9915918 | -6.748736 |\n",
       "| 213780_at | -1.830945e-03 | -0.012559300 | -0.0120276478 | 0.9905494 | 0.9935268 | -6.748774 |\n",
       "| 229152_at |  7.274305e-03 |  0.785249728 |  0.0094557174 | 0.9925702 | 0.9953667 | -6.748803 |\n",
       "| 242883_at |  1.284877e-03 | -0.202721125 |  0.0086463667 | 0.9932061 | 0.9958174 | -6.748810 |\n",
       "| 221658_s_at | -1.067008e-03 | -0.026685009 | -0.0081065141 | 0.9936303 | 0.9960556 | -6.748815 |\n",
       "| 231386_at |  8.382728e-04 |  0.027066565 |  0.0077754172 | 0.9938905 | 0.9961049 | -6.748818 |\n",
       "| 1555643_s_at | -1.162331e-03 |  0.030274093 | -0.0075690004 | 0.9940526 | 0.9961049 | -6.748819 |\n",
       "| 207972_at | -9.074408e-04 | -0.120894936 | -0.0060321130 | 0.9952602 | 0.9971208 | -6.748830 |\n",
       "| 210402_at | -8.584665e-04 |  0.164386348 | -0.0058033702 | 0.9954400 | 0.9971208 | -6.748832 |\n",
       "| 1552948_at | -9.712917e-04 |  0.037384033 | -0.0055466975 | 0.9956417 | 0.9971358 | -6.748833 |\n",
       "| 205685_at | -8.604690e-04 |  0.230873696 | -0.0047693364 | 0.9962525 | 0.9975604 | -6.748837 |\n",
       "| 231794_at | -6.487468e-04 |  0.098780981 | -0.0034424825 | 0.9972950 | 0.9984065 | -6.748843 |\n",
       "| 242809_at | -5.042769e-04 | -0.022999856 | -0.0029907888 | 0.9976500 | 0.9984065 | -6.748845 |\n",
       "| 203603_s_at |  2.719045e-03 |  1.850278399 |  0.0027853887 | 0.9978114 | 0.9984065 | -6.748845 |\n",
       "| 208402_at | -3.644787e-04 | -0.131042892 | -0.0026030083 | 0.9979547 | 0.9984065 | -6.748846 |\n",
       "| 1553534_at | -2.792301e-04 | -0.044046503 | -0.0025039235 | 0.9980325 | 0.9984065 | -6.748846 |\n",
       "| 237320_at |  4.826854e-05 | -0.213631604 |  0.0002714129 | 0.9997867 | 0.9999727 | -6.748849 |\n",
       "| 207067_s_at | -7.696790e-06 |  0.334753187 | -0.0000347936 | 0.9999727 | 0.9999727 | -6.748849 |\n",
       "\n"
      ],
      "text/plain": [
       "             logFC         AveExpr      t             P.Value      adj.P.Val   \n",
       "210372_s_at   5.200189      2.9369060    18.586820    2.082126e-12 1.111647e-08\n",
       "205470_s_at   6.255458      3.4624353    14.523453    9.365043e-11 2.499998e-07\n",
       "204855_at    13.342429      8.3325984    12.511677    8.820153e-10 1.260731e-06\n",
       "219795_at     6.773997      3.2696607    12.454113    9.445448e-10 1.260731e-06\n",
       "220133_at     3.223162      1.5617431    12.127986    1.399296e-09 1.494168e-06\n",
       "204259_at     7.111360      3.6682683    10.809794    7.511799e-09 6.684249e-06\n",
       "203585_at     3.814912      2.3970818    10.575189    1.030032e-08 7.014783e-06\n",
       "203485_at    -1.477512     -0.7942994   -10.526023    1.101229e-08 7.014783e-06\n",
       "219752_at     2.948165      1.8787889    10.473843    1.182488e-08 7.014783e-06\n",
       "232151_at     5.490196      3.4301028    10.321526    1.457882e-08 7.783631e-06\n",
       "209309_at     6.996555      4.1278207     9.848596    2.835113e-08 1.376061e-05\n",
       "230398_at     5.665631      3.3003175     9.766528    3.189540e-08 1.419079e-05\n",
       "218412_s_at   7.497466      5.2023793     9.423892    5.256847e-08 2.158947e-05\n",
       "203962_s_at  10.751804      7.5600927     9.370139    5.692160e-08 2.166592e-05\n",
       "212531_at    17.064917     11.9881353     9.317129    6.158685e-08 2.166592e-05\n",
       "241782_at     2.195624      1.1963186     9.253373    6.773548e-08 2.166592e-05\n",
       "235245_at     2.334599      1.4756777     9.227164    7.044712e-08 2.166592e-05\n",
       "208209_s_at   3.822005      2.4020938     9.203031    7.304486e-08 2.166592e-05\n",
       "1566766_a_at  4.146974      2.6918098     9.136375    8.075648e-08 2.269257e-05\n",
       "244180_at    -1.344145     -0.8307531    -9.061854    9.040027e-08 2.413235e-05\n",
       "205566_at     3.226118      1.9002566     9.012534    9.744220e-08 2.477352e-05\n",
       "206941_x_at  -1.005474     -0.6635295    -8.962673    1.051493e-07 2.551782e-05\n",
       "205349_at     3.212065      1.6030105     8.791665    1.368203e-07 3.176015e-05\n",
       "202917_s_at   9.110590      4.6962185     8.695275    1.589515e-07 3.340168e-05\n",
       "202357_s_at   8.271597      5.6286043     8.689031    1.605088e-07 3.340168e-05\n",
       "205977_s_at   4.119680      2.8835908     8.680509    1.626604e-07 3.340168e-05\n",
       "209409_at     5.113576      3.5591380     8.444412    2.360515e-07 4.667700e-05\n",
       "218704_at     7.976579      5.7665953     8.404005    2.517568e-07 4.800462e-05\n",
       "221815_at     4.468376      3.0759494     8.351318    2.738977e-07 5.042551e-05\n",
       "205476_at    14.466160     10.4372247     8.248990    3.229343e-07 5.747153e-05\n",
       "...          ...           ...          ...           ...          ...         \n",
       "1562211_a_at -3.627688e-03  0.033859078 -0.0281154771 0.9779111    0.9832518   \n",
       "220373_at     5.740751e-03 -0.258401883  0.0269088363 0.9788588    0.9840194   \n",
       "231626_at     2.757279e-02  1.857394629  0.0253280978 0.9801005    0.9850070   \n",
       "220343_at    -2.956402e-03  0.004189963 -0.0251884424 0.9802102    0.9850070   \n",
       "1559429_a_at  3.890462e-03 -0.040891499  0.0217618683 0.9829018    0.9875259   \n",
       "221690_s_at   1.831913e-02  0.659423984  0.0214927376 0.9831133    0.9875525   \n",
       "1553059_at   -3.434531e-03 -0.163317401 -0.0210531485 0.9834586    0.9877136   \n",
       "206834_at    -4.011657e-03 -0.167439369 -0.0198192016 0.9844279    0.9885012   \n",
       "234361_at     3.016104e-03  0.044745530  0.0192995924 0.9848362    0.9887251   \n",
       "224390_s_at  -1.756399e-03 -0.001132483 -0.0188067738 0.9852233    0.9889247   \n",
       "240967_at     2.842829e-03 -0.164153264  0.0185750184 0.9854054    0.9889247   \n",
       "243231_at    -4.517521e-03  0.252822841 -0.0168659091 0.9867481    0.9900861   \n",
       "1552526_at   -1.509388e-03  0.042989550 -0.0147194442 0.9884345    0.9915918   \n",
       "213780_at    -1.830945e-03 -0.012559300 -0.0120276478 0.9905494    0.9935268   \n",
       "229152_at     7.274305e-03  0.785249728  0.0094557174 0.9925702    0.9953667   \n",
       "242883_at     1.284877e-03 -0.202721125  0.0086463667 0.9932061    0.9958174   \n",
       "221658_s_at  -1.067008e-03 -0.026685009 -0.0081065141 0.9936303    0.9960556   \n",
       "231386_at     8.382728e-04  0.027066565  0.0077754172 0.9938905    0.9961049   \n",
       "1555643_s_at -1.162331e-03  0.030274093 -0.0075690004 0.9940526    0.9961049   \n",
       "207972_at    -9.074408e-04 -0.120894936 -0.0060321130 0.9952602    0.9971208   \n",
       "210402_at    -8.584665e-04  0.164386348 -0.0058033702 0.9954400    0.9971208   \n",
       "1552948_at   -9.712917e-04  0.037384033 -0.0055466975 0.9956417    0.9971358   \n",
       "205685_at    -8.604690e-04  0.230873696 -0.0047693364 0.9962525    0.9975604   \n",
       "231794_at    -6.487468e-04  0.098780981 -0.0034424825 0.9972950    0.9984065   \n",
       "242809_at    -5.042769e-04 -0.022999856 -0.0029907888 0.9976500    0.9984065   \n",
       "203603_s_at   2.719045e-03  1.850278399  0.0027853887 0.9978114    0.9984065   \n",
       "208402_at    -3.644787e-04 -0.131042892 -0.0026030083 0.9979547    0.9984065   \n",
       "1553534_at   -2.792301e-04 -0.044046503 -0.0025039235 0.9980325    0.9984065   \n",
       "237320_at     4.826854e-05 -0.213631604  0.0002714129 0.9997867    0.9999727   \n",
       "207067_s_at  -7.696790e-06  0.334753187 -0.0000347936 0.9999727    0.9999727   \n",
       "             B        \n",
       "210372_s_at  17.909406\n",
       "205470_s_at  14.665745\n",
       "204855_at    12.634755\n",
       "219795_at    12.571572\n",
       "220133_at    12.207825\n",
       "204259_at    10.631112\n",
       "203585_at    10.331384\n",
       "203485_at    10.267797\n",
       "219752_at    10.200014\n",
       "232151_at    10.000392\n",
       "209309_at     9.363467\n",
       "230398_at     9.250240\n",
       "218412_s_at   8.768637\n",
       "203962_s_at   8.691761\n",
       "212531_at     8.615594\n",
       "241782_at     8.523515\n",
       "235245_at     8.485513\n",
       "208209_s_at   8.450445\n",
       "1566766_a_at  8.353196\n",
       "244180_at     8.243797\n",
       "205566_at     8.171000\n",
       "206941_x_at   8.097083\n",
       "205349_at     7.841098\n",
       "202917_s_at   7.695107\n",
       "202357_s_at   7.685609\n",
       "205977_s_at   7.672635\n",
       "209409_at     7.309324\n",
       "218704_at     7.246389\n",
       "221815_at     7.163994\n",
       "205476_at     7.002877\n",
       "...          ...      \n",
       "1562211_a_at -6.748436\n",
       "220373_at    -6.748470\n",
       "231626_at    -6.748514\n",
       "220343_at    -6.748517\n",
       "1559429_a_at -6.748601\n",
       "221690_s_at  -6.748608\n",
       "1553059_at   -6.748617\n",
       "206834_at    -6.748644\n",
       "234361_at    -6.748654\n",
       "224390_s_at  -6.748664\n",
       "240967_at    -6.748669\n",
       "243231_at    -6.748700\n",
       "1552526_at   -6.748736\n",
       "213780_at    -6.748774\n",
       "229152_at    -6.748803\n",
       "242883_at    -6.748810\n",
       "221658_s_at  -6.748815\n",
       "231386_at    -6.748818\n",
       "1555643_s_at -6.748819\n",
       "207972_at    -6.748830\n",
       "210402_at    -6.748832\n",
       "1552948_at   -6.748833\n",
       "205685_at    -6.748837\n",
       "231794_at    -6.748843\n",
       "242809_at    -6.748845\n",
       "203603_s_at  -6.748845\n",
       "208402_at    -6.748846\n",
       "1553534_at   -6.748846\n",
       "237320_at    -6.748849\n",
       "207067_s_at  -6.748849"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## The output can be ranked and top-ranking genes extracted from it\n",
    "tested <- topTable(fitE, adjust=\"fdr\", sort.by=\"B\", number=Inf)\n",
    "tested"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Your code contains a unicode char which cannot be displayed in your\n",
      "current locale and R will silently convert it to an escaped form when the\n",
      "R kernel executes this code. This can lead to subtle errors if you use\n",
      "such chars to do comparisons. For more information, please see\n",
      "https://github.com/IRkernel/repr/wiki/Problems-with-unicode-on-windows"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>1211</li>\n",
       "\t<li>6</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1211\n",
       "\\item 6\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1211\n",
       "2. 6\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 1211    6"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## To add the conditions of the p-values or other conditions to get the DE genes.添加p值或其他条件的条件以获得DE基因.\n",
    "DE <- tested[tested$adj.P.Val<0.01,]\n",
    "dim(DE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>558</li>\n",
       "\t<li>6</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 558\n",
       "\\item 6\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 558\n",
       "2. 6\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 558   6"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "DE <- tested[tested$adj.P.Val< 0.01 & abs(tested$logFC) >2, ]\n",
    "dim(DE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 10. Working with the data of multiple classes\n",
    "\n",
    "We know how to make a analysis with two experimental groups, namely, treatment and control. But if we faces three or more, how can we compare them systematically against each other?<br>\n",
    "我们能够将对照组和实验组数据进行分析，但是如果我们遇到三个甚至更多的组别的时候该怎么将他们进行对比呢？<br>\n",
    "\n",
    "Take another dataset from the `leukemiasEset` package as an example. The data is from 60 bone marrow samples of patients with one of the four main types of leukemia (ALL, AML, CLL, and CML) and non-leukemia controls. However, for demonstration purposes, we just take three samples from each of these categories. <br>\n",
    "\n",
    "We need to principally focus on creating the design matrix for the comparison, which is the key difference. And then present it on imaginary data with three conditions and pairwise comparisons.主要需要专注于创建用于比较的设计矩阵，这是关键区别。 用三个条件和成对比较在虚数数据上显示"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Bioconductor version 3.9 (BiocManager 1.30.10), R 3.6.2 (2019-12-12)\n",
      "\n",
      "Installing package(s) 'leukemiasEset'\n",
      "\n",
      "installing the source package 'leukemiasEset'\n",
      "\n",
      "\n",
      "Old packages: 'BH', 'bit', 'caTools', 'cli', 'FactoMineR', 'fansi', 'farver',\n",
      "  'GetoptLong', 'ggridges', 'gplots', 'hms', 'knitr', 'leaps', 'multcomp',\n",
      "  'mvtnorm', 'PMA', 'precrec', 'prettyunits', 'pROC', 'RCurl', 'RSQLite',\n",
      "  'Rttf2pt1', 'shinyjs', 'stringi', 'tinytex', 'xfun', 'XML', 'xts', 'zoo'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "BiocManager::install(\"leukemiasEset\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(leukemiasEset)\n",
    "\n",
    "## Load the dataset wanted to start with from the `leukemiasEset` library.\n",
    "data(leukemiasEset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define the three conditions with two replicates for each.\n",
    "\n",
    "pheno <- pData(leukemiasEset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "## select three samples from each set (use the corresponding indexes here).\n",
    "\n",
    "mydata <- leukemiasEset[, sampleNames(leukemiasEset)[c(1:3, 13:15, 25:27, 49:51)]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create the design matrix based on the condition variables.\n",
    "\n",
    "design <- model.matrix(~0 + factor(pData(mydata)$LeukemiaType))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Rename the columns of the design matrix.\n",
    "\n",
    "colnames(design) <- unique(as.character(pData(mydata)$LeukemiaType))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A matrix: 12 × 4 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ALL</th><th scope=col>AML</th><th scope=col>CLL</th><th scope=col>NoL</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>1</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>1</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>0</td><td>1</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>0</td><td>1</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>0</td><td>1</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>0</td><td>0</td><td>1</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>0</td><td>0</td><td>1</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>0</td><td>0</td><td>1</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>0</td><td>0</td><td>0</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>0</td><td>0</td><td>0</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>0</td><td>0</td><td>0</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 12 × 4 of type dbl\n",
       "\\begin{tabular}{r|llll}\n",
       "  & ALL & AML & CLL & NoL\\\\\n",
       "\\hline\n",
       "\t1 & 1 & 0 & 0 & 0\\\\\n",
       "\t2 & 1 & 0 & 0 & 0\\\\\n",
       "\t3 & 1 & 0 & 0 & 0\\\\\n",
       "\t4 & 0 & 1 & 0 & 0\\\\\n",
       "\t5 & 0 & 1 & 0 & 0\\\\\n",
       "\t6 & 0 & 1 & 0 & 0\\\\\n",
       "\t7 & 0 & 0 & 1 & 0\\\\\n",
       "\t8 & 0 & 0 & 1 & 0\\\\\n",
       "\t9 & 0 & 0 & 1 & 0\\\\\n",
       "\t10 & 0 & 0 & 0 & 1\\\\\n",
       "\t11 & 0 & 0 & 0 & 1\\\\\n",
       "\t12 & 0 & 0 & 0 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 12 × 4 of type dbl\n",
       "\n",
       "| <!--/--> | ALL | AML | CLL | NoL |\n",
       "|---|---|---|---|---|\n",
       "| 1 | 1 | 0 | 0 | 0 |\n",
       "| 2 | 1 | 0 | 0 | 0 |\n",
       "| 3 | 1 | 0 | 0 | 0 |\n",
       "| 4 | 0 | 1 | 0 | 0 |\n",
       "| 5 | 0 | 1 | 0 | 0 |\n",
       "| 6 | 0 | 1 | 0 | 0 |\n",
       "| 7 | 0 | 0 | 1 | 0 |\n",
       "| 8 | 0 | 0 | 1 | 0 |\n",
       "| 9 | 0 | 0 | 1 | 0 |\n",
       "| 10 | 0 | 0 | 0 | 1 |\n",
       "| 11 | 0 | 0 | 0 | 1 |\n",
       "| 12 | 0 | 0 | 0 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "   ALL AML CLL NoL\n",
       "1  1   0   0   0  \n",
       "2  1   0   0   0  \n",
       "3  1   0   0   0  \n",
       "4  0   1   0   0  \n",
       "5  0   1   0   0  \n",
       "6  0   1   0   0  \n",
       "7  0   0   1   0  \n",
       "8  0   0   1   0  \n",
       "9  0   0   1   0  \n",
       "10 0   0   0   1  \n",
       "11 0   0   0   1  \n",
       "12 0   0   0   1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## See the created design matrix.\n",
    "\n",
    "design # First three columns being different types of leukemiaand the fourth one being the control non leukemia samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Proceed by fitting a linear model to the data.\n",
    "\n",
    "library(limma)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "An object of class \"MArrayLM\"\n",
       "$coefficients\n",
       "                     ALL      AML      CLL      NoL\n",
       "ENSG00000000003 3.478096 3.435959 3.608932 3.860124\n",
       "ENSG00000000005 3.540522 3.332321 3.379161 3.278787\n",
       "ENSG00000000419 9.083140 9.124083 9.639916 9.551689\n",
       "ENSG00000000457 4.865089 6.217431 5.838287 5.503014\n",
       "ENSG00000000460 4.294320 3.979242 3.612462 5.218094\n",
       "20167 more rows ...\n",
       "\n",
       "$rank\n",
       "[1] 4\n",
       "\n",
       "$assign\n",
       "[1] 1 1 1 1\n",
       "\n",
       "$qr\n",
       "$qr\n",
       "         ALL        AML       CLL       NoL\n",
       "1 -1.7320508  0.0000000  0.000000  0.000000\n",
       "2  0.5773503 -1.7320508  0.000000  0.000000\n",
       "3  0.5773503  0.0000000 -1.732051  0.000000\n",
       "4  0.0000000  0.5773503  0.000000 -1.732051\n",
       "5  0.0000000  0.5773503  0.000000  0.000000\n",
       "7 more rows ...\n",
       "\n",
       "$qraux\n",
       "[1] 1.57735 1.00000 1.00000 1.00000\n",
       "\n",
       "$pivot\n",
       "[1] 1 2 3 4\n",
       "\n",
       "$tol\n",
       "[1] 1e-07\n",
       "\n",
       "$rank\n",
       "[1] 4\n",
       "\n",
       "\n",
       "$df.residual\n",
       "[1] 8 8 8 8 8\n",
       "20167 more elements ...\n",
       "\n",
       "$sigma\n",
       "ENSG00000000003 ENSG00000000005 ENSG00000000419 ENSG00000000457 ENSG00000000460 \n",
       "      0.2263650       0.1942866       0.6732108       0.4570501       0.6726086 \n",
       "20167 more elements ...\n",
       "\n",
       "$cov.coefficients\n",
       "          ALL       AML       CLL       NoL\n",
       "ALL 0.3333333 0.0000000 0.0000000 0.0000000\n",
       "AML 0.0000000 0.3333333 0.0000000 0.0000000\n",
       "CLL 0.0000000 0.0000000 0.3333333 0.0000000\n",
       "NoL 0.0000000 0.0000000 0.0000000 0.3333333\n",
       "\n",
       "$stdev.unscaled\n",
       "                      ALL       AML       CLL       NoL\n",
       "ENSG00000000003 0.5773503 0.5773503 0.5773503 0.5773503\n",
       "ENSG00000000005 0.5773503 0.5773503 0.5773503 0.5773503\n",
       "ENSG00000000419 0.5773503 0.5773503 0.5773503 0.5773503\n",
       "ENSG00000000457 0.5773503 0.5773503 0.5773503 0.5773503\n",
       "ENSG00000000460 0.5773503 0.5773503 0.5773503 0.5773503\n",
       "20167 more rows ...\n",
       "\n",
       "$pivot\n",
       "[1] 1 2 3 4\n",
       "\n",
       "$Amean\n",
       "ENSG00000000003 ENSG00000000005 ENSG00000000419 ENSG00000000457 ENSG00000000460 \n",
       "       3.595778        3.382697        9.349707        5.605955        4.276029 \n",
       "20167 more elements ...\n",
       "\n",
       "$method\n",
       "[1] \"ls\"\n",
       "\n",
       "$design\n",
       "  ALL AML CLL NoL\n",
       "1   1   0   0   0\n",
       "2   1   0   0   0\n",
       "3   1   0   0   0\n",
       "4   0   1   0   0\n",
       "5   0   1   0   0\n",
       "7 more rows ...\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A matrix: 4 × 3 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>NoL - ALL</th><th scope=col>NoL - AML</th><th scope=col>NoL - CLL</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ALL</th><td>-1</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>AML</th><td> 0</td><td>-1</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>CLL</th><td> 0</td><td> 0</td><td>-1</td></tr>\n",
       "\t<tr><th scope=row>NoL</th><td> 1</td><td> 1</td><td> 1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 4 × 3 of type dbl\n",
       "\\begin{tabular}{r|lll}\n",
       "  & NoL - ALL & NoL - AML & NoL - CLL\\\\\n",
       "\\hline\n",
       "\tALL & -1 &  0 &  0\\\\\n",
       "\tAML &  0 & -1 &  0\\\\\n",
       "\tCLL &  0 &  0 & -1\\\\\n",
       "\tNoL &  1 &  1 &  1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 4 × 3 of type dbl\n",
       "\n",
       "| <!--/--> | NoL - ALL | NoL - AML | NoL - CLL |\n",
       "|---|---|---|---|\n",
       "| ALL | -1 |  0 |  0 |\n",
       "| AML |  0 | -1 |  0 |\n",
       "| CLL |  0 |  0 | -1 |\n",
       "| NoL |  1 |  1 |  1 |\n",
       "\n"
      ],
      "text/plain": [
       "      Contrasts\n",
       "Levels NoL - ALL NoL - AML NoL - CLL\n",
       "   ALL -1         0         0       \n",
       "   AML  0        -1         0       \n",
       "   CLL  0         0        -1       \n",
       "   NoL  1         1         1       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fit <- lmFit(mydata, design)\n",
    "fit\n",
    "\n",
    "## Create a contrast matrix for pairwise comparisons.\n",
    "contrast.matrix <- makeContrasts(NoL- ALL, NoL- AML, NoL- CLL, levels = design)\n",
    "contrast.matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Fit the linear model using this contrast matrix.\n",
    "\n",
    "fit2 <- contrasts.fit(fit, contrast.matrix) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Perform the empirical Bayes analysis of the model.\n",
    "\n",
    "fit2 <- eBayes(fit2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 20172 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>logFC</th><th scope=col>AveExpr</th><th scope=col>t</th><th scope=col>P.Value</th><th scope=col>adj.P.Val</th><th scope=col>B</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000152078</th><td> 4.510507</td><td>4.856523</td><td> 28.13988</td><td>4.463747e-11</td><td>9.004270e-07</td><td>14.014724</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000117519</th><td>-4.185175</td><td>4.791585</td><td>-22.73888</td><td>3.878292e-10</td><td>3.911645e-06</td><td>12.697381</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000145850</th><td> 4.142236</td><td>4.507655</td><td> 17.38636</td><td>5.759942e-09</td><td>2.925048e-05</td><td>10.727823</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000170180</th><td> 5.681327</td><td>5.734169</td><td> 17.37423</td><td>5.800214e-09</td><td>2.925048e-05</td><td>10.722311</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000087586</th><td> 3.952183</td><td>5.720789</td><td> 16.45393</td><td>9.977396e-09</td><td>3.111188e-05</td><td>10.287051</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000047597</th><td> 5.362419</td><td>5.108415</td><td> 16.32474</td><td>1.079114e-08</td><td>3.111188e-05</td><td>10.223153</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000175449</th><td> 3.954293</td><td>4.667288</td><td> 16.32395</td><td>1.079631e-08</td><td>3.111188e-05</td><td>10.222762</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000104043</th><td> 4.594551</td><td>5.877417</td><td> 15.91546</td><td>1.388748e-08</td><td>3.501728e-05</td><td>10.015920</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000024526</th><td> 3.048430</td><td>4.657230</td><td> 15.70978</td><td>1.580050e-08</td><td>3.541419e-05</td><td> 9.908949</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000115641</th><td> 3.660332</td><td>5.460531</td><td> 15.49887</td><td>1.806560e-08</td><td>3.644192e-05</td><td> 9.797237</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000188672</th><td> 4.614139</td><td>5.380265</td><td> 15.26785</td><td>2.096196e-08</td><td>3.844043e-05</td><td> 9.672452</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000133063</th><td> 4.857495</td><td>5.630711</td><td> 14.76852</td><td>2.911671e-08</td><td>4.894519e-05</td><td> 9.393827</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000164330</th><td>-4.548742</td><td>5.139590</td><td>-14.54086</td><td>3.393791e-08</td><td>4.941928e-05</td><td> 9.262590</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000173372</th><td> 2.784815</td><td>5.523598</td><td> 14.52528</td><td>3.429853e-08</td><td>4.941928e-05</td><td> 9.253507</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000234955</th><td>-3.181503</td><td>4.290691</td><td>-14.21654</td><td>4.237846e-08</td><td>5.699055e-05</td><td> 9.070902</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000102935</th><td>-5.341545</td><td>4.829078</td><td>-13.98406</td><td>4.983512e-08</td><td>6.282962e-05</td><td> 8.929961</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000159189</th><td> 3.512535</td><td>4.917536</td><td> 13.86537</td><td>5.418509e-08</td><td>6.429540e-05</td><td> 8.856847</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000107562</th><td> 3.359836</td><td>4.984015</td><td> 13.50747</td><td>7.001837e-08</td><td>7.434224e-05</td><td> 8.631454</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000086506</th><td> 4.011569</td><td>5.859682</td><td> 13.50738</td><td>7.002293e-08</td><td>7.434224e-05</td><td> 8.631396</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000130635</th><td>-4.395195</td><td>5.362425</td><td>-13.27887</td><td>8.274167e-08</td><td>8.345325e-05</td><td> 8.483522</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000066923</th><td>-4.520295</td><td>6.494213</td><td>-12.93093</td><td>1.072199e-07</td><td>1.029924e-04</td><td> 8.252195</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000162692</th><td> 4.844911</td><td>4.630924</td><td> 12.65173</td><td>1.326070e-07</td><td>1.215885e-04</td><td> 8.060998</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000198336</th><td> 3.883021</td><td>6.768568</td><td> 12.52486</td><td>1.462526e-07</td><td>1.232467e-04</td><td> 7.972428</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000109501</th><td>-4.056787</td><td>5.692457</td><td>-12.51679</td><td>1.471700e-07</td><td>1.232467e-04</td><td> 7.966765</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000168497</th><td> 3.108121</td><td>4.741989</td><td> 12.40452</td><td>1.606225e-07</td><td>1.232467e-04</td><td> 7.887420</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000080819</th><td> 4.365147</td><td>5.626793</td><td> 12.34964</td><td>1.676802e-07</td><td>1.232467e-04</td><td> 7.848332</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000130821</th><td> 3.111452</td><td>6.040674</td><td> 12.31654</td><td>1.721015e-07</td><td>1.232467e-04</td><td> 7.824650</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000162599</th><td> 2.402039</td><td>4.391735</td><td> 12.29547</td><td>1.749813e-07</td><td>1.232467e-04</td><td> 7.809539</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000173391</th><td> 4.511073</td><td>4.289218</td><td> 12.27961</td><td>1.771840e-07</td><td>1.232467e-04</td><td> 7.798144</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000071967</th><td> 3.474916</td><td>5.842135</td><td> 11.93721</td><td>2.329409e-07</td><td>1.524925e-04</td><td> 7.547824</td></tr>\n",
       "\t<tr><th scope=row>...</th><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000184898</th><td>-6.679917e-04</td><td> 4.385725</td><td>-2.006986e-03</td><td>0.9984370</td><td>0.9997165</td><td>-6.787415</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000162526</th><td> 5.333565e-04</td><td> 6.845486</td><td> 2.006777e-03</td><td>0.9984372</td><td>0.9997165</td><td>-6.787415</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000117174</th><td> 1.246917e-03</td><td> 7.133992</td><td> 1.978443e-03</td><td>0.9984592</td><td>0.9997165</td><td>-6.787415</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000205038</th><td> 4.642149e-04</td><td> 3.707259</td><td> 1.977952e-03</td><td>0.9984596</td><td>0.9997165</td><td>-6.787415</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000245927</th><td> 3.255004e-04</td><td> 3.583726</td><td> 1.954928e-03</td><td>0.9984776</td><td>0.9997165</td><td>-6.787415</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000113327</th><td>-2.485544e-04</td><td> 3.238793</td><td>-1.678736e-03</td><td>0.9986926</td><td>0.9997335</td><td>-6.787416</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000163479</th><td> 5.506234e-04</td><td>10.494864</td><td> 1.663374e-03</td><td>0.9987046</td><td>0.9997335</td><td>-6.787416</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000011260</th><td>-5.297753e-04</td><td> 9.563794</td><td>-1.649236e-03</td><td>0.9987156</td><td>0.9997335</td><td>-6.787416</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000121743</th><td> 2.468503e-04</td><td> 3.792622</td><td> 1.588047e-03</td><td>0.9987633</td><td>0.9997335</td><td>-6.787416</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000148335</th><td>-3.661422e-04</td><td> 6.499836</td><td>-1.554274e-03</td><td>0.9987896</td><td>0.9997335</td><td>-6.787416</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000092345</th><td>-2.874116e-04</td><td> 2.950834</td><td>-1.551385e-03</td><td>0.9987918</td><td>0.9997335</td><td>-6.787416</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000166348</th><td> 1.128627e-03</td><td> 5.727922</td><td> 1.482353e-03</td><td>0.9988456</td><td>0.9997377</td><td>-6.787416</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000018236</th><td> 3.532907e-04</td><td> 3.782055</td><td> 1.403421e-03</td><td>0.9989071</td><td>0.9997426</td><td>-6.787416</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000164604</th><td>-2.339540e-04</td><td> 3.328498</td><td>-1.327779e-03</td><td>0.9989660</td><td>0.9997426</td><td>-6.787416</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000118596</th><td> 5.541809e-04</td><td> 4.865807</td><td> 1.285061e-03</td><td>0.9989992</td><td>0.9997426</td><td>-6.787416</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000230743</th><td> 3.698132e-04</td><td> 9.729000</td><td> 1.052519e-03</td><td>0.9991803</td><td>0.9997845</td><td>-6.787417</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000101350</th><td> 5.139717e-04</td><td> 6.506761</td><td> 1.017790e-03</td><td>0.9992074</td><td>0.9997845</td><td>-6.787417</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000168591</th><td>-3.643930e-04</td><td> 7.228878</td><td>-9.431005e-04</td><td>0.9992655</td><td>0.9997845</td><td>-6.787417</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000101441</th><td>-2.494937e-04</td><td> 4.053601</td><td>-8.461279e-04</td><td>0.9993411</td><td>0.9997845</td><td>-6.787417</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000197472</th><td> 1.479224e-04</td><td> 2.947163</td><td> 8.299368e-04</td><td>0.9993537</td><td>0.9997845</td><td>-6.787417</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000139656</th><td>-1.622774e-04</td><td> 3.339458</td><td>-8.232701e-04</td><td>0.9993589</td><td>0.9997845</td><td>-6.787417</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000147439</th><td>-3.089195e-04</td><td> 8.237133</td><td>-7.858004e-04</td><td>0.9993880</td><td>0.9997845</td><td>-6.787417</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000002745</th><td> 1.185971e-04</td><td> 3.225855</td><td> 5.995940e-04</td><td>0.9995331</td><td>0.9998287</td><td>-6.787417</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000134138</th><td>-2.667634e-04</td><td> 3.451379</td><td>-5.995502e-04</td><td>0.9995331</td><td>0.9998287</td><td>-6.787417</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000122884</th><td> 2.610773e-04</td><td> 7.224659</td><td> 5.358619e-04</td><td>0.9995827</td><td>0.9998287</td><td>-6.787417</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000182187</th><td>-7.211313e-05</td><td> 5.232893</td><td>-4.745563e-04</td><td>0.9996304</td><td>0.9998287</td><td>-6.787417</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000239961</th><td> 2.724331e-04</td><td> 6.020185</td><td> 3.815179e-04</td><td>0.9997029</td><td>0.9998516</td><td>-6.787417</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000163449</th><td> 5.805249e-05</td><td> 3.751309</td><td> 3.007044e-04</td><td>0.9997658</td><td>0.9998650</td><td>-6.787417</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000151718</th><td> 8.595550e-06</td><td> 3.492013</td><td> 5.490059e-05</td><td>0.9999572</td><td>0.9999758</td><td>-6.787417</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000138768</th><td>-1.614100e-05</td><td> 7.786855</td><td>-3.104010e-05</td><td>0.9999758</td><td>0.9999758</td><td>-6.787417</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 20172 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & logFC & AveExpr & t & P.Value & adj.P.Val & B\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tENSG00000152078 &  4.510507 & 4.856523 &  28.13988 & 4.463747e-11 & 9.004270e-07 & 14.014724\\\\\n",
       "\tENSG00000117519 & -4.185175 & 4.791585 & -22.73888 & 3.878292e-10 & 3.911645e-06 & 12.697381\\\\\n",
       "\tENSG00000145850 &  4.142236 & 4.507655 &  17.38636 & 5.759942e-09 & 2.925048e-05 & 10.727823\\\\\n",
       "\tENSG00000170180 &  5.681327 & 5.734169 &  17.37423 & 5.800214e-09 & 2.925048e-05 & 10.722311\\\\\n",
       "\tENSG00000087586 &  3.952183 & 5.720789 &  16.45393 & 9.977396e-09 & 3.111188e-05 & 10.287051\\\\\n",
       "\tENSG00000047597 &  5.362419 & 5.108415 &  16.32474 & 1.079114e-08 & 3.111188e-05 & 10.223153\\\\\n",
       "\tENSG00000175449 &  3.954293 & 4.667288 &  16.32395 & 1.079631e-08 & 3.111188e-05 & 10.222762\\\\\n",
       "\tENSG00000104043 &  4.594551 & 5.877417 &  15.91546 & 1.388748e-08 & 3.501728e-05 & 10.015920\\\\\n",
       "\tENSG00000024526 &  3.048430 & 4.657230 &  15.70978 & 1.580050e-08 & 3.541419e-05 &  9.908949\\\\\n",
       "\tENSG00000115641 &  3.660332 & 5.460531 &  15.49887 & 1.806560e-08 & 3.644192e-05 &  9.797237\\\\\n",
       "\tENSG00000188672 &  4.614139 & 5.380265 &  15.26785 & 2.096196e-08 & 3.844043e-05 &  9.672452\\\\\n",
       "\tENSG00000133063 &  4.857495 & 5.630711 &  14.76852 & 2.911671e-08 & 4.894519e-05 &  9.393827\\\\\n",
       "\tENSG00000164330 & -4.548742 & 5.139590 & -14.54086 & 3.393791e-08 & 4.941928e-05 &  9.262590\\\\\n",
       "\tENSG00000173372 &  2.784815 & 5.523598 &  14.52528 & 3.429853e-08 & 4.941928e-05 &  9.253507\\\\\n",
       "\tENSG00000234955 & -3.181503 & 4.290691 & -14.21654 & 4.237846e-08 & 5.699055e-05 &  9.070902\\\\\n",
       "\tENSG00000102935 & -5.341545 & 4.829078 & -13.98406 & 4.983512e-08 & 6.282962e-05 &  8.929961\\\\\n",
       "\tENSG00000159189 &  3.512535 & 4.917536 &  13.86537 & 5.418509e-08 & 6.429540e-05 &  8.856847\\\\\n",
       "\tENSG00000107562 &  3.359836 & 4.984015 &  13.50747 & 7.001837e-08 & 7.434224e-05 &  8.631454\\\\\n",
       "\tENSG00000086506 &  4.011569 & 5.859682 &  13.50738 & 7.002293e-08 & 7.434224e-05 &  8.631396\\\\\n",
       "\tENSG00000130635 & -4.395195 & 5.362425 & -13.27887 & 8.274167e-08 & 8.345325e-05 &  8.483522\\\\\n",
       "\tENSG00000066923 & -4.520295 & 6.494213 & -12.93093 & 1.072199e-07 & 1.029924e-04 &  8.252195\\\\\n",
       "\tENSG00000162692 &  4.844911 & 4.630924 &  12.65173 & 1.326070e-07 & 1.215885e-04 &  8.060998\\\\\n",
       "\tENSG00000198336 &  3.883021 & 6.768568 &  12.52486 & 1.462526e-07 & 1.232467e-04 &  7.972428\\\\\n",
       "\tENSG00000109501 & -4.056787 & 5.692457 & -12.51679 & 1.471700e-07 & 1.232467e-04 &  7.966765\\\\\n",
       "\tENSG00000168497 &  3.108121 & 4.741989 &  12.40452 & 1.606225e-07 & 1.232467e-04 &  7.887420\\\\\n",
       "\tENSG00000080819 &  4.365147 & 5.626793 &  12.34964 & 1.676802e-07 & 1.232467e-04 &  7.848332\\\\\n",
       "\tENSG00000130821 &  3.111452 & 6.040674 &  12.31654 & 1.721015e-07 & 1.232467e-04 &  7.824650\\\\\n",
       "\tENSG00000162599 &  2.402039 & 4.391735 &  12.29547 & 1.749813e-07 & 1.232467e-04 &  7.809539\\\\\n",
       "\tENSG00000173391 &  4.511073 & 4.289218 &  12.27961 & 1.771840e-07 & 1.232467e-04 &  7.798144\\\\\n",
       "\tENSG00000071967 &  3.474916 & 5.842135 &  11.93721 & 2.329409e-07 & 1.524925e-04 &  7.547824\\\\\n",
       "\t... & ... & ... & ... & ... & ... & ...\\\\\n",
       "\tENSG00000184898 & -6.679917e-04 &  4.385725 & -2.006986e-03 & 0.9984370 & 0.9997165 & -6.787415\\\\\n",
       "\tENSG00000162526 &  5.333565e-04 &  6.845486 &  2.006777e-03 & 0.9984372 & 0.9997165 & -6.787415\\\\\n",
       "\tENSG00000117174 &  1.246917e-03 &  7.133992 &  1.978443e-03 & 0.9984592 & 0.9997165 & -6.787415\\\\\n",
       "\tENSG00000205038 &  4.642149e-04 &  3.707259 &  1.977952e-03 & 0.9984596 & 0.9997165 & -6.787415\\\\\n",
       "\tENSG00000245927 &  3.255004e-04 &  3.583726 &  1.954928e-03 & 0.9984776 & 0.9997165 & -6.787415\\\\\n",
       "\tENSG00000113327 & -2.485544e-04 &  3.238793 & -1.678736e-03 & 0.9986926 & 0.9997335 & -6.787416\\\\\n",
       "\tENSG00000163479 &  5.506234e-04 & 10.494864 &  1.663374e-03 & 0.9987046 & 0.9997335 & -6.787416\\\\\n",
       "\tENSG00000011260 & -5.297753e-04 &  9.563794 & -1.649236e-03 & 0.9987156 & 0.9997335 & -6.787416\\\\\n",
       "\tENSG00000121743 &  2.468503e-04 &  3.792622 &  1.588047e-03 & 0.9987633 & 0.9997335 & -6.787416\\\\\n",
       "\tENSG00000148335 & -3.661422e-04 &  6.499836 & -1.554274e-03 & 0.9987896 & 0.9997335 & -6.787416\\\\\n",
       "\tENSG00000092345 & -2.874116e-04 &  2.950834 & -1.551385e-03 & 0.9987918 & 0.9997335 & -6.787416\\\\\n",
       "\tENSG00000166348 &  1.128627e-03 &  5.727922 &  1.482353e-03 & 0.9988456 & 0.9997377 & -6.787416\\\\\n",
       "\tENSG00000018236 &  3.532907e-04 &  3.782055 &  1.403421e-03 & 0.9989071 & 0.9997426 & -6.787416\\\\\n",
       "\tENSG00000164604 & -2.339540e-04 &  3.328498 & -1.327779e-03 & 0.9989660 & 0.9997426 & -6.787416\\\\\n",
       "\tENSG00000118596 &  5.541809e-04 &  4.865807 &  1.285061e-03 & 0.9989992 & 0.9997426 & -6.787416\\\\\n",
       "\tENSG00000230743 &  3.698132e-04 &  9.729000 &  1.052519e-03 & 0.9991803 & 0.9997845 & -6.787417\\\\\n",
       "\tENSG00000101350 &  5.139717e-04 &  6.506761 &  1.017790e-03 & 0.9992074 & 0.9997845 & -6.787417\\\\\n",
       "\tENSG00000168591 & -3.643930e-04 &  7.228878 & -9.431005e-04 & 0.9992655 & 0.9997845 & -6.787417\\\\\n",
       "\tENSG00000101441 & -2.494937e-04 &  4.053601 & -8.461279e-04 & 0.9993411 & 0.9997845 & -6.787417\\\\\n",
       "\tENSG00000197472 &  1.479224e-04 &  2.947163 &  8.299368e-04 & 0.9993537 & 0.9997845 & -6.787417\\\\\n",
       "\tENSG00000139656 & -1.622774e-04 &  3.339458 & -8.232701e-04 & 0.9993589 & 0.9997845 & -6.787417\\\\\n",
       "\tENSG00000147439 & -3.089195e-04 &  8.237133 & -7.858004e-04 & 0.9993880 & 0.9997845 & -6.787417\\\\\n",
       "\tENSG00000002745 &  1.185971e-04 &  3.225855 &  5.995940e-04 & 0.9995331 & 0.9998287 & -6.787417\\\\\n",
       "\tENSG00000134138 & -2.667634e-04 &  3.451379 & -5.995502e-04 & 0.9995331 & 0.9998287 & -6.787417\\\\\n",
       "\tENSG00000122884 &  2.610773e-04 &  7.224659 &  5.358619e-04 & 0.9995827 & 0.9998287 & -6.787417\\\\\n",
       "\tENSG00000182187 & -7.211313e-05 &  5.232893 & -4.745563e-04 & 0.9996304 & 0.9998287 & -6.787417\\\\\n",
       "\tENSG00000239961 &  2.724331e-04 &  6.020185 &  3.815179e-04 & 0.9997029 & 0.9998516 & -6.787417\\\\\n",
       "\tENSG00000163449 &  5.805249e-05 &  3.751309 &  3.007044e-04 & 0.9997658 & 0.9998650 & -6.787417\\\\\n",
       "\tENSG00000151718 &  8.595550e-06 &  3.492013 &  5.490059e-05 & 0.9999572 & 0.9999758 & -6.787417\\\\\n",
       "\tENSG00000138768 & -1.614100e-05 &  7.786855 & -3.104010e-05 & 0.9999758 & 0.9999758 & -6.787417\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 20172 × 6\n",
       "\n",
       "| <!--/--> | logFC &lt;dbl&gt; | AveExpr &lt;dbl&gt; | t &lt;dbl&gt; | P.Value &lt;dbl&gt; | adj.P.Val &lt;dbl&gt; | B &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| ENSG00000152078 |  4.510507 | 4.856523 |  28.13988 | 4.463747e-11 | 9.004270e-07 | 14.014724 |\n",
       "| ENSG00000117519 | -4.185175 | 4.791585 | -22.73888 | 3.878292e-10 | 3.911645e-06 | 12.697381 |\n",
       "| ENSG00000145850 |  4.142236 | 4.507655 |  17.38636 | 5.759942e-09 | 2.925048e-05 | 10.727823 |\n",
       "| ENSG00000170180 |  5.681327 | 5.734169 |  17.37423 | 5.800214e-09 | 2.925048e-05 | 10.722311 |\n",
       "| ENSG00000087586 |  3.952183 | 5.720789 |  16.45393 | 9.977396e-09 | 3.111188e-05 | 10.287051 |\n",
       "| ENSG00000047597 |  5.362419 | 5.108415 |  16.32474 | 1.079114e-08 | 3.111188e-05 | 10.223153 |\n",
       "| ENSG00000175449 |  3.954293 | 4.667288 |  16.32395 | 1.079631e-08 | 3.111188e-05 | 10.222762 |\n",
       "| ENSG00000104043 |  4.594551 | 5.877417 |  15.91546 | 1.388748e-08 | 3.501728e-05 | 10.015920 |\n",
       "| ENSG00000024526 |  3.048430 | 4.657230 |  15.70978 | 1.580050e-08 | 3.541419e-05 |  9.908949 |\n",
       "| ENSG00000115641 |  3.660332 | 5.460531 |  15.49887 | 1.806560e-08 | 3.644192e-05 |  9.797237 |\n",
       "| ENSG00000188672 |  4.614139 | 5.380265 |  15.26785 | 2.096196e-08 | 3.844043e-05 |  9.672452 |\n",
       "| ENSG00000133063 |  4.857495 | 5.630711 |  14.76852 | 2.911671e-08 | 4.894519e-05 |  9.393827 |\n",
       "| ENSG00000164330 | -4.548742 | 5.139590 | -14.54086 | 3.393791e-08 | 4.941928e-05 |  9.262590 |\n",
       "| ENSG00000173372 |  2.784815 | 5.523598 |  14.52528 | 3.429853e-08 | 4.941928e-05 |  9.253507 |\n",
       "| ENSG00000234955 | -3.181503 | 4.290691 | -14.21654 | 4.237846e-08 | 5.699055e-05 |  9.070902 |\n",
       "| ENSG00000102935 | -5.341545 | 4.829078 | -13.98406 | 4.983512e-08 | 6.282962e-05 |  8.929961 |\n",
       "| ENSG00000159189 |  3.512535 | 4.917536 |  13.86537 | 5.418509e-08 | 6.429540e-05 |  8.856847 |\n",
       "| ENSG00000107562 |  3.359836 | 4.984015 |  13.50747 | 7.001837e-08 | 7.434224e-05 |  8.631454 |\n",
       "| ENSG00000086506 |  4.011569 | 5.859682 |  13.50738 | 7.002293e-08 | 7.434224e-05 |  8.631396 |\n",
       "| ENSG00000130635 | -4.395195 | 5.362425 | -13.27887 | 8.274167e-08 | 8.345325e-05 |  8.483522 |\n",
       "| ENSG00000066923 | -4.520295 | 6.494213 | -12.93093 | 1.072199e-07 | 1.029924e-04 |  8.252195 |\n",
       "| ENSG00000162692 |  4.844911 | 4.630924 |  12.65173 | 1.326070e-07 | 1.215885e-04 |  8.060998 |\n",
       "| ENSG00000198336 |  3.883021 | 6.768568 |  12.52486 | 1.462526e-07 | 1.232467e-04 |  7.972428 |\n",
       "| ENSG00000109501 | -4.056787 | 5.692457 | -12.51679 | 1.471700e-07 | 1.232467e-04 |  7.966765 |\n",
       "| ENSG00000168497 |  3.108121 | 4.741989 |  12.40452 | 1.606225e-07 | 1.232467e-04 |  7.887420 |\n",
       "| ENSG00000080819 |  4.365147 | 5.626793 |  12.34964 | 1.676802e-07 | 1.232467e-04 |  7.848332 |\n",
       "| ENSG00000130821 |  3.111452 | 6.040674 |  12.31654 | 1.721015e-07 | 1.232467e-04 |  7.824650 |\n",
       "| ENSG00000162599 |  2.402039 | 4.391735 |  12.29547 | 1.749813e-07 | 1.232467e-04 |  7.809539 |\n",
       "| ENSG00000173391 |  4.511073 | 4.289218 |  12.27961 | 1.771840e-07 | 1.232467e-04 |  7.798144 |\n",
       "| ENSG00000071967 |  3.474916 | 5.842135 |  11.93721 | 2.329409e-07 | 1.524925e-04 |  7.547824 |\n",
       "| ... | ... | ... | ... | ... | ... | ... |\n",
       "| ENSG00000184898 | -6.679917e-04 |  4.385725 | -2.006986e-03 | 0.9984370 | 0.9997165 | -6.787415 |\n",
       "| ENSG00000162526 |  5.333565e-04 |  6.845486 |  2.006777e-03 | 0.9984372 | 0.9997165 | -6.787415 |\n",
       "| ENSG00000117174 |  1.246917e-03 |  7.133992 |  1.978443e-03 | 0.9984592 | 0.9997165 | -6.787415 |\n",
       "| ENSG00000205038 |  4.642149e-04 |  3.707259 |  1.977952e-03 | 0.9984596 | 0.9997165 | -6.787415 |\n",
       "| ENSG00000245927 |  3.255004e-04 |  3.583726 |  1.954928e-03 | 0.9984776 | 0.9997165 | -6.787415 |\n",
       "| ENSG00000113327 | -2.485544e-04 |  3.238793 | -1.678736e-03 | 0.9986926 | 0.9997335 | -6.787416 |\n",
       "| ENSG00000163479 |  5.506234e-04 | 10.494864 |  1.663374e-03 | 0.9987046 | 0.9997335 | -6.787416 |\n",
       "| ENSG00000011260 | -5.297753e-04 |  9.563794 | -1.649236e-03 | 0.9987156 | 0.9997335 | -6.787416 |\n",
       "| ENSG00000121743 |  2.468503e-04 |  3.792622 |  1.588047e-03 | 0.9987633 | 0.9997335 | -6.787416 |\n",
       "| ENSG00000148335 | -3.661422e-04 |  6.499836 | -1.554274e-03 | 0.9987896 | 0.9997335 | -6.787416 |\n",
       "| ENSG00000092345 | -2.874116e-04 |  2.950834 | -1.551385e-03 | 0.9987918 | 0.9997335 | -6.787416 |\n",
       "| ENSG00000166348 |  1.128627e-03 |  5.727922 |  1.482353e-03 | 0.9988456 | 0.9997377 | -6.787416 |\n",
       "| ENSG00000018236 |  3.532907e-04 |  3.782055 |  1.403421e-03 | 0.9989071 | 0.9997426 | -6.787416 |\n",
       "| ENSG00000164604 | -2.339540e-04 |  3.328498 | -1.327779e-03 | 0.9989660 | 0.9997426 | -6.787416 |\n",
       "| ENSG00000118596 |  5.541809e-04 |  4.865807 |  1.285061e-03 | 0.9989992 | 0.9997426 | -6.787416 |\n",
       "| ENSG00000230743 |  3.698132e-04 |  9.729000 |  1.052519e-03 | 0.9991803 | 0.9997845 | -6.787417 |\n",
       "| ENSG00000101350 |  5.139717e-04 |  6.506761 |  1.017790e-03 | 0.9992074 | 0.9997845 | -6.787417 |\n",
       "| ENSG00000168591 | -3.643930e-04 |  7.228878 | -9.431005e-04 | 0.9992655 | 0.9997845 | -6.787417 |\n",
       "| ENSG00000101441 | -2.494937e-04 |  4.053601 | -8.461279e-04 | 0.9993411 | 0.9997845 | -6.787417 |\n",
       "| ENSG00000197472 |  1.479224e-04 |  2.947163 |  8.299368e-04 | 0.9993537 | 0.9997845 | -6.787417 |\n",
       "| ENSG00000139656 | -1.622774e-04 |  3.339458 | -8.232701e-04 | 0.9993589 | 0.9997845 | -6.787417 |\n",
       "| ENSG00000147439 | -3.089195e-04 |  8.237133 | -7.858004e-04 | 0.9993880 | 0.9997845 | -6.787417 |\n",
       "| ENSG00000002745 |  1.185971e-04 |  3.225855 |  5.995940e-04 | 0.9995331 | 0.9998287 | -6.787417 |\n",
       "| ENSG00000134138 | -2.667634e-04 |  3.451379 | -5.995502e-04 | 0.9995331 | 0.9998287 | -6.787417 |\n",
       "| ENSG00000122884 |  2.610773e-04 |  7.224659 |  5.358619e-04 | 0.9995827 | 0.9998287 | -6.787417 |\n",
       "| ENSG00000182187 | -7.211313e-05 |  5.232893 | -4.745563e-04 | 0.9996304 | 0.9998287 | -6.787417 |\n",
       "| ENSG00000239961 |  2.724331e-04 |  6.020185 |  3.815179e-04 | 0.9997029 | 0.9998516 | -6.787417 |\n",
       "| ENSG00000163449 |  5.805249e-05 |  3.751309 |  3.007044e-04 | 0.9997658 | 0.9998650 | -6.787417 |\n",
       "| ENSG00000151718 |  8.595550e-06 |  3.492013 |  5.490059e-05 | 0.9999572 | 0.9999758 | -6.787417 |\n",
       "| ENSG00000138768 | -1.614100e-05 |  7.786855 | -3.104010e-05 | 0.9999758 | 0.9999758 | -6.787417 |\n",
       "\n"
      ],
      "text/plain": [
       "                logFC         AveExpr   t             P.Value      adj.P.Val   \n",
       "ENSG00000152078  4.510507     4.856523   28.13988     4.463747e-11 9.004270e-07\n",
       "ENSG00000117519 -4.185175     4.791585  -22.73888     3.878292e-10 3.911645e-06\n",
       "ENSG00000145850  4.142236     4.507655   17.38636     5.759942e-09 2.925048e-05\n",
       "ENSG00000170180  5.681327     5.734169   17.37423     5.800214e-09 2.925048e-05\n",
       "ENSG00000087586  3.952183     5.720789   16.45393     9.977396e-09 3.111188e-05\n",
       "ENSG00000047597  5.362419     5.108415   16.32474     1.079114e-08 3.111188e-05\n",
       "ENSG00000175449  3.954293     4.667288   16.32395     1.079631e-08 3.111188e-05\n",
       "ENSG00000104043  4.594551     5.877417   15.91546     1.388748e-08 3.501728e-05\n",
       "ENSG00000024526  3.048430     4.657230   15.70978     1.580050e-08 3.541419e-05\n",
       "ENSG00000115641  3.660332     5.460531   15.49887     1.806560e-08 3.644192e-05\n",
       "ENSG00000188672  4.614139     5.380265   15.26785     2.096196e-08 3.844043e-05\n",
       "ENSG00000133063  4.857495     5.630711   14.76852     2.911671e-08 4.894519e-05\n",
       "ENSG00000164330 -4.548742     5.139590  -14.54086     3.393791e-08 4.941928e-05\n",
       "ENSG00000173372  2.784815     5.523598   14.52528     3.429853e-08 4.941928e-05\n",
       "ENSG00000234955 -3.181503     4.290691  -14.21654     4.237846e-08 5.699055e-05\n",
       "ENSG00000102935 -5.341545     4.829078  -13.98406     4.983512e-08 6.282962e-05\n",
       "ENSG00000159189  3.512535     4.917536   13.86537     5.418509e-08 6.429540e-05\n",
       "ENSG00000107562  3.359836     4.984015   13.50747     7.001837e-08 7.434224e-05\n",
       "ENSG00000086506  4.011569     5.859682   13.50738     7.002293e-08 7.434224e-05\n",
       "ENSG00000130635 -4.395195     5.362425  -13.27887     8.274167e-08 8.345325e-05\n",
       "ENSG00000066923 -4.520295     6.494213  -12.93093     1.072199e-07 1.029924e-04\n",
       "ENSG00000162692  4.844911     4.630924   12.65173     1.326070e-07 1.215885e-04\n",
       "ENSG00000198336  3.883021     6.768568   12.52486     1.462526e-07 1.232467e-04\n",
       "ENSG00000109501 -4.056787     5.692457  -12.51679     1.471700e-07 1.232467e-04\n",
       "ENSG00000168497  3.108121     4.741989   12.40452     1.606225e-07 1.232467e-04\n",
       "ENSG00000080819  4.365147     5.626793   12.34964     1.676802e-07 1.232467e-04\n",
       "ENSG00000130821  3.111452     6.040674   12.31654     1.721015e-07 1.232467e-04\n",
       "ENSG00000162599  2.402039     4.391735   12.29547     1.749813e-07 1.232467e-04\n",
       "ENSG00000173391  4.511073     4.289218   12.27961     1.771840e-07 1.232467e-04\n",
       "ENSG00000071967  3.474916     5.842135   11.93721     2.329409e-07 1.524925e-04\n",
       "...             ...           ...       ...           ...          ...         \n",
       "ENSG00000184898 -6.679917e-04  4.385725 -2.006986e-03 0.9984370    0.9997165   \n",
       "ENSG00000162526  5.333565e-04  6.845486  2.006777e-03 0.9984372    0.9997165   \n",
       "ENSG00000117174  1.246917e-03  7.133992  1.978443e-03 0.9984592    0.9997165   \n",
       "ENSG00000205038  4.642149e-04  3.707259  1.977952e-03 0.9984596    0.9997165   \n",
       "ENSG00000245927  3.255004e-04  3.583726  1.954928e-03 0.9984776    0.9997165   \n",
       "ENSG00000113327 -2.485544e-04  3.238793 -1.678736e-03 0.9986926    0.9997335   \n",
       "ENSG00000163479  5.506234e-04 10.494864  1.663374e-03 0.9987046    0.9997335   \n",
       "ENSG00000011260 -5.297753e-04  9.563794 -1.649236e-03 0.9987156    0.9997335   \n",
       "ENSG00000121743  2.468503e-04  3.792622  1.588047e-03 0.9987633    0.9997335   \n",
       "ENSG00000148335 -3.661422e-04  6.499836 -1.554274e-03 0.9987896    0.9997335   \n",
       "ENSG00000092345 -2.874116e-04  2.950834 -1.551385e-03 0.9987918    0.9997335   \n",
       "ENSG00000166348  1.128627e-03  5.727922  1.482353e-03 0.9988456    0.9997377   \n",
       "ENSG00000018236  3.532907e-04  3.782055  1.403421e-03 0.9989071    0.9997426   \n",
       "ENSG00000164604 -2.339540e-04  3.328498 -1.327779e-03 0.9989660    0.9997426   \n",
       "ENSG00000118596  5.541809e-04  4.865807  1.285061e-03 0.9989992    0.9997426   \n",
       "ENSG00000230743  3.698132e-04  9.729000  1.052519e-03 0.9991803    0.9997845   \n",
       "ENSG00000101350  5.139717e-04  6.506761  1.017790e-03 0.9992074    0.9997845   \n",
       "ENSG00000168591 -3.643930e-04  7.228878 -9.431005e-04 0.9992655    0.9997845   \n",
       "ENSG00000101441 -2.494937e-04  4.053601 -8.461279e-04 0.9993411    0.9997845   \n",
       "ENSG00000197472  1.479224e-04  2.947163  8.299368e-04 0.9993537    0.9997845   \n",
       "ENSG00000139656 -1.622774e-04  3.339458 -8.232701e-04 0.9993589    0.9997845   \n",
       "ENSG00000147439 -3.089195e-04  8.237133 -7.858004e-04 0.9993880    0.9997845   \n",
       "ENSG00000002745  1.185971e-04  3.225855  5.995940e-04 0.9995331    0.9998287   \n",
       "ENSG00000134138 -2.667634e-04  3.451379 -5.995502e-04 0.9995331    0.9998287   \n",
       "ENSG00000122884  2.610773e-04  7.224659  5.358619e-04 0.9995827    0.9998287   \n",
       "ENSG00000182187 -7.211313e-05  5.232893 -4.745563e-04 0.9996304    0.9998287   \n",
       "ENSG00000239961  2.724331e-04  6.020185  3.815179e-04 0.9997029    0.9998516   \n",
       "ENSG00000163449  5.805249e-05  3.751309  3.007044e-04 0.9997658    0.9998650   \n",
       "ENSG00000151718  8.595550e-06  3.492013  5.490059e-05 0.9999572    0.9999758   \n",
       "ENSG00000138768 -1.614100e-05  7.786855 -3.104010e-05 0.9999758    0.9999758   \n",
       "                B        \n",
       "ENSG00000152078 14.014724\n",
       "ENSG00000117519 12.697381\n",
       "ENSG00000145850 10.727823\n",
       "ENSG00000170180 10.722311\n",
       "ENSG00000087586 10.287051\n",
       "ENSG00000047597 10.223153\n",
       "ENSG00000175449 10.222762\n",
       "ENSG00000104043 10.015920\n",
       "ENSG00000024526  9.908949\n",
       "ENSG00000115641  9.797237\n",
       "ENSG00000188672  9.672452\n",
       "ENSG00000133063  9.393827\n",
       "ENSG00000164330  9.262590\n",
       "ENSG00000173372  9.253507\n",
       "ENSG00000234955  9.070902\n",
       "ENSG00000102935  8.929961\n",
       "ENSG00000159189  8.856847\n",
       "ENSG00000107562  8.631454\n",
       "ENSG00000086506  8.631396\n",
       "ENSG00000130635  8.483522\n",
       "ENSG00000066923  8.252195\n",
       "ENSG00000162692  8.060998\n",
       "ENSG00000198336  7.972428\n",
       "ENSG00000109501  7.966765\n",
       "ENSG00000168497  7.887420\n",
       "ENSG00000080819  7.848332\n",
       "ENSG00000130821  7.824650\n",
       "ENSG00000162599  7.809539\n",
       "ENSG00000173391  7.798144\n",
       "ENSG00000071967  7.547824\n",
       "...             ...      \n",
       "ENSG00000184898 -6.787415\n",
       "ENSG00000162526 -6.787415\n",
       "ENSG00000117174 -6.787415\n",
       "ENSG00000205038 -6.787415\n",
       "ENSG00000245927 -6.787415\n",
       "ENSG00000113327 -6.787416\n",
       "ENSG00000163479 -6.787416\n",
       "ENSG00000011260 -6.787416\n",
       "ENSG00000121743 -6.787416\n",
       "ENSG00000148335 -6.787416\n",
       "ENSG00000092345 -6.787416\n",
       "ENSG00000166348 -6.787416\n",
       "ENSG00000018236 -6.787416\n",
       "ENSG00000164604 -6.787416\n",
       "ENSG00000118596 -6.787416\n",
       "ENSG00000230743 -6.787417\n",
       "ENSG00000101350 -6.787417\n",
       "ENSG00000168591 -6.787417\n",
       "ENSG00000101441 -6.787417\n",
       "ENSG00000197472 -6.787417\n",
       "ENSG00000139656 -6.787417\n",
       "ENSG00000147439 -6.787417\n",
       "ENSG00000002745 -6.787417\n",
       "ENSG00000134138 -6.787417\n",
       "ENSG00000122884 -6.787417\n",
       "ENSG00000182187 -6.787417\n",
       "ENSG00000239961 -6.787417\n",
       "ENSG00000163449 -6.787417\n",
       "ENSG00000151718 -6.787417\n",
       "ENSG00000138768 -6.787417"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Extract the differentially expressed gene for each of the pairwise comparisons using the `coef` argument \n",
    "## in the `topTable` function. For the first pairwise comparison, \n",
    "## set `coef=1` to compare non-lukemia control with Acute Lymphoblastic Leukemia (ALL) leukemia.\n",
    "\n",
    "tested2 <- topTable(fit2,adjust=\"fdr\",sort.by=\"B\",number=Inf, coef=1)\n",
    "tested2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>252</li>\n",
       "\t<li>6</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 252\n",
       "\\item 6\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 252\n",
       "2. 6\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 252   6"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "DE2 <- tested2[tested2$adj.P.Val < 0.01,]\n",
    "dim(DE2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The major difference compared to the two conditions(Finding the differentially expressed genes) is the use of the contrast matrix and the `coef` argument in the `topTable` function. The contrast matrix enables the pairwise comparison for the computation of p-values. The model fitted returns a set of p-values for each comparison. The desired set of DE genes for a specific comparison can then be extracted using the proper `coef` value for comparisons.<br>\n",
    "\n",
    "与两种条件（寻找差异表达基因）相比，主要区别在于使用对照矩阵和`topTable`函数中的`coef`参数。对比矩阵可以用于计算p值的成对比较。拟合的模型会为每次比较返回一组p值。然后可以使用适当的“coef”值来提取用于特定比较的期望的一组DE基因进行比较。\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 3.6.2",
   "language": "R",
   "name": "ir362"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
